{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "eb9b0e84",
   "metadata": {},
   "source": [
    "# Pangenomes graph database construction\n",
    "\n",
    "## Authors\n",
    "- Jérôme Arnoux, Genoscope/LABGeM - CEA, CNRS, Paris Saclay University\n",
    "- Angela Bonifati, Liris CNRS, Lyon 1 University\n",
    "- Alexandra Calteau, Genoscope/LABGeM - CEA, CNRS, Paris Saclay University\n",
    "- Stefania Dumbrava, SAMOVAR/Inst. Poltech de Paris, ENSIIE\n",
    "- Guillaume Gautreau, MetaGenoPolis, Université Paris-Saclay, INRAE, MGP\n",
    "\n",
    "## Purpose\n",
    "This notebook allows to reproduce the construction of a Neo4J graph database with 10 ESKAPE pangenome. The code behind was developped in the ICDE conference paper submisson.\n",
    "\n",
    "## Notebook organisation\n",
    "The structure of the notebook is as follows:\n",
    "\n",
    "- Basic setup and dependencies.\n",
    "- Pangenomic data import in the Neo4j database.\n",
    "- Running the experimental analysis: query workload and scripts.\n",
    "- Additional resources.\n",
    "\n",
    "## Get our original data\n",
    "To reproduce our result you can download this [folder](https://drive.google.com/drive/folders/1eZ7GQgU5tAgfryK31EPV6OP2wVRrj79B?usp=share_link). If you prefer to begin from scratch go to the supplementary code section. It will guide you to generate pangenomes and compute their similarities."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3428c268",
   "metadata": {},
   "source": [
    "# Basic setup and dependencies\n",
    "## Dependencies\n",
    "Below is listed all dependencies. To install it, you can follow the step in the next section 'Environment configuration'. If you prefer you can construct your own environment by installing the dependencies listed below.\n",
    "\n",
    "Use pip to install:\n",
    "- dict2graph==2.0.0  # pip3 install git+https://git.connect.dzd-ev.de/dzdpythonmodules/dict2graph.git\n",
    "- graphio==0.4.0\n",
    "\n",
    "Use conda to install:\n",
    "- ppanggolin==1.2.74  # You can only install dependencies because you will need to install the development version (1.2.105) after\n",
    "- pyhmmer==0.6.3\n",
    "- py2neo==2021.2.3\n",
    "- rgi==6.0.1  # WARNING rgi is only compatible with python3.6 and all other dependencies with python3.9. It's necessary to install card in another environment (Refer to rgi [github](https://github.com/arpcard/rgi)).\n",
    "- genome_updater==0.5.1\n",
    "\n",
    "Neo4J:\n",
    "- Add local DBMS with a Neo4j version of 4.4.11\n",
    "- APOC 4.4.0.10 or more\n",
    "- Optional : Neo4J Desktop 1.5.0\n",
    "\n",
    "## Environment configuration\n",
    "To begin, note that you must have an empty Neo4J DMBS (version 4.4.11) open and available with the APOC plugin install (version 4.4.0.10).\n",
    "\n",
    "To execute the following script, you will need to install some packages. They're listed in the following conda environment file. The *in development* version of PPanGGOLiN is required to satisfy some feature and pangenomes compatibility.\n",
    "\n",
    "To install the conda environment in jupyter kernel, please copy and paste the following code in your terminal:\n",
    "```\n",
    "conda update -n base -c defaults conda -y\n",
    "conda env create --file conda-env.yml\n",
    "conda init bash\n",
    "conda activate pangraph\n",
    "pip install --user ipykernel\n",
    "python -m ipykernel install --user --name=pangraph\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fatal : le chemin de destination 'PPanGGOLiN' existe déjà et n'est pas un répertoire vide.\r\n",
      "Processing ./PPanGGOLiN\r\n",
      "  Preparing metadata (setup.py) ... \u001B[?25ldone\r\n",
      "\u001B[?25hBuilding wheels for collected packages: ppanggolin\r\n",
      "  Building wheel for ppanggolin (setup.py) ... \u001B[?25ldone\r\n",
      "\u001B[?25h  Created wheel for ppanggolin: filename=ppanggolin-1.2.105-cp39-cp39-linux_x86_64.whl size=3537950 sha256=8535ad73f66b6a80e1e9f2bf2f069a3f6683e20ac213696680649a88a452557e\r\n",
      "  Stored in directory: /tmp/pip-ephem-wheel-cache-9wzla4e6/wheels/b8/98/cb/1dca3b3248b69f64c7949960a16d580eb98d8bf958044c04ed\r\n",
      "Successfully built ppanggolin\r\n",
      "Installing collected packages: ppanggolin\r\n",
      "  Attempting uninstall: ppanggolin\r\n",
      "    Found existing installation: ppanggolin 1.2.105\r\n",
      "    Uninstalling ppanggolin-1.2.105:\r\n",
      "      Successfully uninstalled ppanggolin-1.2.105\r\n",
      "Successfully installed ppanggolin-1.2.105\r\n"
     ]
    }
   ],
   "source": [
    "!git clone -b release1.3 https://github.com/labgem/PPanGGOLiN.git\n",
    "!pip install PPanGGOLiN/."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Library import\n",
    "We import all the common required Python libraries to execute all the script"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "47eab96a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# default libraries\n",
    "import logging\n",
    "import os\n",
    "from tqdm import tqdm\n",
    "from pathlib import Path\n",
    "\n",
    "# installed libraries\n",
    "from py2neo import Graph\n",
    "\n",
    "# local libraries\n",
    "from script.python.utils import check_tsv_sanity"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ce7c84f",
   "metadata": {},
   "source": [
    "# Parameter definition\n",
    "We set all relevant parameters for our notebook. By convention, parameters are uppercase, while all the \n",
    "other variables follow Python's guidelines."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "outputs": [],
   "source": [
    "GBFF=Path(\"data/GBFF\")  # Genome used to construct pangenomes\n",
    "os.environ[\"GBFF\"] = GBFF.as_posix()\n",
    "PANGENOMES=Path(\"data/pangenomes\")  # Directory with our pangenomes\n",
    "os.environ[\"PANGENOMES\"] = PANGENOMES.as_posix()\n",
    "GF=Path(\"data/GF_fasta\")  # Directory with for all pangenomes the DNA sequences of all gene families\n",
    "os.environ[\"GF\"] = GF.as_posix()\n",
    "CARD=Path(\"data/CARD\")\n",
    "os.environ[\"CARD\"] = CARD.as_posix()\n",
    "SIMILARITY=Path(\"data/similarity\")\n",
    "os.environ[\"SIMILARITY\"] = SIMILARITY.as_posix()\n",
    "\n",
    "PANGENOMES_TSV=Path(f\"{PANGENOMES}/organism_eskape.list\")\n",
    "SIMILARITIES=Path(f\"{SIMILARITY}/similarity.tsv\")\n",
    "\n",
    "#Neo4J paramaeters\n",
    "URI=\"bolt://localhost:7687\"\n",
    "USER=\"neo4j\"\n",
    "PWD=\"PANORAMA2022\"\n",
    "\n",
    "#Exec parameters\n",
    "CPU=6\n",
    "os.environ[\"CPU\"]=str(CPU)\n",
    "BATCH_SIZE=1000\n",
    "#\n",
    "pangenomes = check_tsv_sanity(PANGENOMES_TSV)  # Create a dictionnary with path file and information to pangenomes.\n",
    "graph = Graph(uri=URI, user=USER, password=PWD)\n",
    "graph.delete_all()  # make sure the graph is empty"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Pangenomic data import in the Neo4j database\n",
    "## Translation pangenome in dictionnary to export in Graph database\n",
    "\n",
    "The first step is to translate our pangenomes in a data structure adapted to export into our Graph database."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [],
   "source": [
    "from script.python import Pangenome\n",
    "from script.python.translate import write_families, write_organisms, write_spot, write_rgp, write_modules\n",
    "\n",
    "def create_dict(pangenome: Pangenome) -> dict:\n",
    "    \"\"\"Create a dictionary with the pangenome content to export one pangenome into a graph database\n",
    "    :param pangenome: A pangenome construct thanks to PPanGGOLiN\n",
    "\n",
    "    :return: Compatible dictionary to export in graph database, corresponding to a pangenome\n",
    "    \"\"\"\n",
    "\n",
    "    translate_dict = {\"Pangenome\": {\"name\": pangenome.name, \"taxid\": pangenome.taxid,\n",
    "                                  \"Family\": [],\n",
    "                                  \"Partition\": [],\n",
    "                                  \"Module\": [],\n",
    "                                  \"RGP\": write_rgp(parent=pangenome),\n",
    "                                  \"Spot\": [], \"Genome\": []}}\n",
    "    write_families(pangenome, translate_dict)\n",
    "    write_organisms(pangenome, translate_dict)\n",
    "    write_spot(pangenome, translate_dict)\n",
    "    write_modules(pangenome, translate_dict)\n",
    "    return translate_dict"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Loader configuration\n",
    "To export pangenomes into our graph database we are using the package employed by the CovidGraph framework."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [],
   "source": [
    "from multiprocessing import Lock\n",
    "\n",
    "from py2neo import Node\n",
    "# installed librairies\n",
    "from dict2graph import Dict2graph\n",
    "\n",
    "\n",
    "def custom_post_func(node: Node):\n",
    "    if node is not None and node.__primarylabel__ == \"Gene\":\n",
    "        del node[\"tmp_id\"]\n",
    "    return node\n",
    "\n",
    "\n",
    "class PangenomeLoader:\n",
    "\n",
    "    def __init__(self, pangenome_name: str, pangenome_data: dict, lock: Lock, batch_size: int = 1000):\n",
    "        self.name = pangenome_name\n",
    "        self.lock = lock\n",
    "        self.data = pangenome_data\n",
    "        self.batch_size = batch_size\n",
    "        self._build_loader()\n",
    "\n",
    "    def load(self, graph: Graph):\n",
    "        assert self.lock is not None, \"Lock not Initialized\"\n",
    "        try:\n",
    "            with self.lock:\n",
    "                logging.getLogger().debug(\"parse\")\n",
    "                self.loader.parse(self.data)\n",
    "                logging.getLogger().debug(\"index\")\n",
    "                self.loader.create_indexes(graph)\n",
    "                logging.getLogger().debug(\"merge\")\n",
    "                self.loader.merge(graph)\n",
    "        except Exception as error:\n",
    "            raise Exception(f\"Load to Neo4j failed because : {error}\")\n",
    "\n",
    "    def _build_loader(self):\n",
    "        d2g = Dict2graph()\n",
    "        d2g.config_dict_primarykey_generated_hashed_attrs_by_label = {\n",
    "            \"Pangenome\": 'AllAttributes',  # Random id\n",
    "            \"Family\": [\"name\"],\n",
    "            \"Partition\": 'AllAttributes',\n",
    "            \"Gene\": 'AllAttributes',\n",
    "            \"Module\": \"InnerContent\",\n",
    "            \"Spot\": \"AllContent\",\n",
    "            \"RGP\": \"InnerContent\",\n",
    "            \"Genome\": \"InnerContent\",\n",
    "            \"Contig\": \"AllContent\"\n",
    "        }\n",
    "        d2g.config_str_primarykey_generated_attr_name = \"hash_id\"\n",
    "        d2g.config_list_blocklist_collection_hubs = [\n",
    "            \"PangenomeCollection\",\n",
    "            \"FamilyCollection\",\n",
    "            \"PartitionCollection\",\n",
    "            \"GeneCollection\",\n",
    "            \"NeighborCollection\",\n",
    "            \"ModuleCollection\",\n",
    "            \"SpotCollection\",\n",
    "            \"RGPCollection\",\n",
    "            \"GenomeCollection\",\n",
    "            \"ContigCollection\",\n",
    "        ]\n",
    "        d2g.config_dict_node_prop_to_rel_prop = {\"Family\": {\"weight\": [\"NEIGHBOR\"]}} #,\n",
    "                                                 # \"Shell\": {\"weight\": [\"NEIGHBOR\"]},\n",
    "                                                 # \"Cloud\": {\"weight\": [\"NEIGHBOR\"]}}  # ,  \"partition\": [\"IN_MODULE\"]}}\n",
    "        d2g.config_dict_primarykey_attr_by_label = {\"Family\": [\"name\"],\n",
    "                                                    \"Gene\": [\"name\"],\n",
    "                                                    \"Partition\": [\"partition\"]}\n",
    "        d2g.config_dict_reltype_override = {\"PANGENOME_HAS_FAMILY\": \"IS_IN_PANGENOME\",\n",
    "                                            \"FAMILY_HAS_GENE\": \"IS_IN_FAMILY\",\n",
    "                                            \"FAMILY_HAS_PARTITION\": \"HAS_PARTITION\",\n",
    "                                            \"FAMILY_HAS_FAMILY\": \"NEIGHBOR\",\n",
    "                                            \"MODULE_HAS_FAMILY\": \"IS_IN_MODULE\",\n",
    "                                            \"SPOT_HAS_RGP\": \"IS_IN_SPOT\",\n",
    "                                            \"RGP_HAS_GENE\": \"IS_IN_RGP\",\n",
    "                                            \"GENOME_HAS_CONTIG\": \"IS_IN_GENOME\",\n",
    "                                            \"CONTIG_HAS_GENE\": \"IS_IN_CONTIG\"}\n",
    "        d2g.config_list_blocklist_reltypes = [\"PANGENOME_HAS_MODULE\",\n",
    "                                              \"PANGENOME_HAS_RGP\",\n",
    "                                              \"PANGENOME_HAS_SPOT\",\n",
    "                                              \"PANGENOME_HAS_GENOME\"]\n",
    "        d2g.config_bool_capitalize_labels = False\n",
    "        d2g.config_func_node_post_modifier = custom_post_func\n",
    "        d2g.config_graphio_batch_size = self.batch_size\n",
    "        self.loader = d2g"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Export to Graph Database\n",
    "### Export pangenomes"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Begin pangenomes load\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████| 591490/591490 [00:01<00:00, 321010.58gene/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 1070798/1070798 [00:03<00:00, 324808.58gene/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 124/124 [00:09<00:00, 12.57organism/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████| 581629/581629 [00:02<00:00, 209995.57gene family/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████| 19918/19918 [00:00<00:00, 69774.81gene family/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████| 572409/572409 [00:03<00:00, 160633.42contig adjacency/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 112136/112136 [00:00<00:00, 419234.12region/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 4499/4499 [00:00<00:00, 14032.60spot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 5017/5017 [00:00<00:00, 307216.92module/s]\n",
      "100%|██████████████████████████████████████████████████████████████████████████████████████████████| 117/117 [00:00<00:00, 123052.55gene family/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 284/284 [00:17<00:00, 16.35organism/s]\n",
      "100%|██████████████████████████████████████████████████████████████████████████████████████| 1045048/1045048 [00:05<00:00, 183655.01gene family/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████| 14427/14427 [00:00<00:00, 65635.16gene family/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████| 1043916/1043916 [00:04<00:00, 223771.96contig adjacency/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 257053/257053 [00:00<00:00, 423998.41region/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 8933/8933 [00:02<00:00, 3794.61spot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 3789/3789 [00:00<00:00, 200994.31module/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████| 97/97 [00:00<00:00, 152377.34gene family/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UNWIND $rels AS rel\n",
      "MATCH (a:Family), (b:Partition)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:HAS_PARTITION]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Family), (b:Gene)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:IS_IN_FAMILY]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Family), (b:Family)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:NEIGHBOR]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Pangenome), (b:Family)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:IS_IN_PANGENOME]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Module), (b:Family)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:IS_IN_MODULE]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:RGP), (b:Gene)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:IS_IN_RGP]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Spot), (b:RGP)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:IS_IN_SPOT]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Contig), (b:Gene)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:IS_IN_CONTIG]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Genome), (b:Contig)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:IS_IN_GENOME]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Family), (b:Partition)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:HAS_PARTITION]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Family), (b:Gene)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:IS_IN_FAMILY]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Family), (b:Family)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:NEIGHBOR]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Pangenome), (b:Family)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:IS_IN_PANGENOME]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Module), (b:Family)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:IS_IN_MODULE]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:RGP), (b:Gene)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:IS_IN_RGP]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Spot), (b:RGP)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:IS_IN_SPOT]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Contig), (b:Gene)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:IS_IN_CONTIG]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n",
      "UNWIND $rels AS rel\n",
      "MATCH (a:Genome), (b:Contig)\n",
      "WHERE a.hash_id = rel.start_hash_id AND b.hash_id = rel.end_hash_id\n",
      "MERGE (a)-[r:IS_IN_GENOME]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [19:54<00:00, 597.43s/pangenome]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All pangenomes loaded in : 0:19:55.591789\n"
     ]
    }
   ],
   "source": [
    "# default libraries\n",
    "from concurrent.futures import ProcessPoolExecutor\n",
    "from multiprocessing import Manager, Lock\n",
    "from time import time\n",
    "from datetime import timedelta\n",
    "\n",
    "# local librairies\n",
    "from script.python.utils import check_pangenome_info\n",
    "from script.python.export import give_gene_tmp_id\n",
    "\n",
    "\n",
    "\n",
    "db_loading_lock: Lock = None\n",
    "\n",
    "\n",
    "def init_db_lock(lock: Lock):\n",
    "    global db_loading_lock\n",
    "    if db_loading_lock is None:\n",
    "        db_loading_lock = lock\n",
    "\n",
    "\n",
    "def load_pangenome(pangenome_name, pangenome_info, batch_size: int = 1000):\n",
    "    \"\"\"\n",
    "\n",
    "    :param pangenome_name:\n",
    "    :param pangenome_info:\n",
    "    :param batch_size:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    logging.getLogger(f\"Add {pangenome_name} to load list\")\n",
    "    pangenome = Pangenome(name=pangenome_name, taxid=pangenome_info[\"taxid\"])\n",
    "    pangenome.add_file(pangenome_info[\"path\"])\n",
    "    check_pangenome_info(pangenome, need_annotations=True, need_families=True, need_graph=True,\n",
    "                         need_rgp=True, need_spots=True, need_modules=True, need_anntation_fam=True,\n",
    "                         disable_bar=False)\n",
    "    give_gene_tmp_id(pangenome)\n",
    "    data = create_dict(pangenome)\n",
    "    loader = PangenomeLoader(pangenome_name, data, db_loading_lock, batch_size=batch_size)\n",
    "    loader.load(graph)\n",
    "\n",
    "\n",
    "def load_pangenome_mp(pangenomes: dict, cpu: int = 1, batch_size: int = 1000):\n",
    "    \"\"\"\n",
    "\n",
    "    :param pangenomes:\n",
    "    :param cpu:\n",
    "    :param batch_size:\n",
    "    :return:\n",
    "    \"\"\"\n",
    "    manager = Manager()\n",
    "    lock = manager.Lock()\n",
    "    with ProcessPoolExecutor(max_workers=cpu, initializer=init_db_lock, initargs=(lock,)) as executor:\n",
    "        list(tqdm(executor.map(load_pangenome, pangenomes.keys(), pangenomes.values(),\n",
    "                               [batch_size] * len(pangenomes)),\n",
    "                  total=len(pangenomes), unit='pangenome'))\n",
    "\n",
    "print(\"Begin pangenomes load\")\n",
    "begin_load_time = time()\n",
    "load_pangenome_mp(pangenomes, CPU, BATCH_SIZE)\n",
    "load_time = time() - begin_load_time\n",
    "print(f\"All pangenomes loaded in : {timedelta(seconds=load_time)}\")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Export similarities"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Bengin load of similarities...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|                                                                                                       | 0/1 [00:00<?, ?similarities_batch/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "UNWIND $rels AS rel\n",
      "MATCH (a:Family), (b:Family)\n",
      "WHERE a.name = rel.start_name AND b.name = rel.end_name\n",
      "MERGE (a)-[r:IS_SIMILAR]->(b)\n",
      "SET r = rel.properties RETURN count(r)\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [01:12<00:00, 72.42s/similarities_batch]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All similarities loaded in : 0:01:12.611399\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# installed librairies\n",
    "from graphio import RelationshipSet\n",
    "import pandas as pd\n",
    "\n",
    "def load_similarities(tsv: Path, batch_size: int = 1000):\n",
    "    df = pd.read_csv(filepath_or_buffer=tsv, sep=\"\\t\", header=None,\n",
    "                     names=[\"Family_1\", \"Family_2\", \"identity\", \"covery\"])\n",
    "\n",
    "    is_similar_list = []\n",
    "    is_similar_to = RelationshipSet('IS_SIMILAR', ['Family'], ['Family'], ['name'], ['name'])\n",
    "    chunk_size = batch_size * 10\n",
    "    for row in df.iterrows():\n",
    "        if len(is_similar_to.relationships) >= chunk_size:\n",
    "            is_similar_list.append(is_similar_to)\n",
    "            is_similar_to = RelationshipSet('IS_SIMILAR', ['Family'], ['Family'], ['name'], ['name'])\n",
    "        is_similar_to.add_relationship(start_node_properties={\"name\": row[1]['Family_1']},\n",
    "                                       end_node_properties={\"name\": row[1]['Family_2']},\n",
    "                                       properties={\"identity\": row[1]['identity'],\n",
    "                                                   \"coverage\": row[1]['covery']})\n",
    "    is_similar_list.append(is_similar_to)\n",
    "    for sim in tqdm(is_similar_list, unit=\"similarities_batch\", total=len(is_similar_list)):\n",
    "        sim.merge(graph=graph, batch_size=batch_size)\n",
    "\n",
    "print(\"Bengin load of similarities...\")\n",
    "begin_sim_time = time()\n",
    "load_similarities(SIMILARITIES, BATCH_SIZE)\n",
    "sim_time = time() - begin_sim_time\n",
    "print(f\"All similarities loaded in : {timedelta(seconds=sim_time)}\")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Invert edges"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Invert edges...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████| 7/7 [02:51<00:00, 24.57s/label]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All edges inverted in : 0:02:51.981150\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "from script.python.export import invert_edges_query\n",
    "\n",
    "\n",
    "def invert_edges(edge_label: str):\n",
    "    query = invert_edges_query(edge_label)\n",
    "    try:\n",
    "        graph.run(query)\n",
    "    except Exception as errror:\n",
    "        raise Exception(f\"Invert edges failed because : {errror}\")\n",
    "\n",
    "print(\"Invert edges...\")\n",
    "begin_invert_time = time()\n",
    "labels2invert = [\"IS_IN_PANGENOME\", \"IS_IN_MODULE\", \"IS_IN_FAMILY\", \"IS_IN_CONTIG\",\n",
    "                 \"IS_IN_GENOME\", \"IS_IN_SPOT\", \"IS_IN_RGP\"]\n",
    "for edge_label in tqdm(labels2invert, unit='label'):\n",
    "    logging.getLogger().debug(f\"Invert: {edge_label}\")\n",
    "    invert_edges(edge_label)\n",
    "invert_time = time() - begin_invert_time\n",
    "print(f\"All edges inverted in : {timedelta(seconds=invert_time)}\")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Running the experimental analysis: query workload and scripts"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family) WHERE f.annotation IS NOT NULL RETURN p.name, count(f)\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(:Gene)-[:IS_IN_RGP]->(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WHERE f.annotation IS NOT NULL WITH f, p, s, count(DISTINCT r) AS cnt ORDER BY cnt DESC RETURN f.name, p.name, cnt, s.name\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(:Gene)-[:IS_IN_RGP]->(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WHERE f.annotation IS NOT NULL WITH f, p, s, count(DISTINCT r) AS cnt ORDER BY cnt DESC RETURN f.name, p.name, cnt, s.name\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(:Gene)-[:IS_IN_RGP]->(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WHERE f.annotation IS NOT NULL WITH f, p, s, count(DISTINCT r) AS cnt ORDER BY cnt DESC RETURN f.name, p.name, cnt, s.name\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(:Gene)-[:IS_IN_RGP]->(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WHERE f.annotation IS NOT NULL WITH f, p, s, count(DISTINCT r) AS cnt ORDER BY cnt DESC RETURN f.name, p.name, cnt, s.name\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(:Gene)-[:IS_IN_RGP]->(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WHERE f.annotation IS NOT NULL WITH f, p, s, count(DISTINCT r) AS cnt ORDER BY cnt DESC RETURN f.name, p.name, cnt, s.name\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(:Gene)-[:IS_IN_RGP]->(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WHERE f.annotation IS NOT NULL WITH f, p, s, count(DISTINCT r) AS cnt ORDER BY cnt DESC RETURN f.name, p.name, cnt, s.name\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(:Gene)-[:IS_IN_RGP]->(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WHERE f.annotation IS NOT NULL WITH f, p, s, count(DISTINCT r) AS cnt ORDER BY cnt DESC RETURN f.name, p.name, cnt, s.name\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(:Gene)-[:IS_IN_RGP]->(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WHERE f.annotation IS NOT NULL WITH f, p, s, count(DISTINCT r) AS cnt ORDER BY cnt DESC RETURN f.name, p.name, cnt, s.name\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(:Gene)-[:IS_IN_RGP]->(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WHERE f.annotation IS NOT NULL WITH f, p, s, count(DISTINCT r) AS cnt ORDER BY cnt DESC RETURN f.name, p.name, cnt, s.name\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(:Gene)-[:IS_IN_RGP]->(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WHERE f.annotation IS NOT NULL WITH f, p, s, count(DISTINCT r) AS cnt ORDER BY cnt DESC RETURN f.name, p.name, cnt, s.name\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]->(r:RGP) WITH p, count(DISTINCT r) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]->(r:RGP) WITH p, count(DISTINCT r) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]->(r:RGP) WITH p, count(DISTINCT r) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]->(r:RGP) WITH p, count(DISTINCT r) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]->(r:RGP) WITH p, count(DISTINCT r) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]->(r:RGP) WITH p, count(DISTINCT r) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]->(r:RGP) WITH p, count(DISTINCT r) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]->(r:RGP) WITH p, count(DISTINCT r) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]->(r:RGP) WITH p, count(DISTINCT r) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)<-[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]->(r:RGP) WITH p, count(DISTINCT r) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT f) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT f) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT f) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT f) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT f) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT f) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT f) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT f) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT f) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT f) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT m) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT m) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT m) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT m) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT m) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT m) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT m) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT m) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT m) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p:Pangenome)<-[:IS_IN_PANGENOME]-(f:Family)-[:IS_IN_MODULE]->(m:Module) WITH p, count(DISTINCT m) AS cnt ORDER BY cnt DESC LIMIT 10 RETURN p.name, cnt\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p1:Pangenome)<-[:IS_IN_PANGENOME]-(f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH p1, f1, s1 MATCH (f1)-[IS_SIMILAR]-(f2:Family) -[:HAS_PARTITION]->(s2) WITH p1, f1, s1, s2, f2 MATCH (p2:Pangenome)<-[IS_IN_PANGENOME]-(f2) WHERE p1.name <> p2.name RETURN p1.name, f1.name, s1.partition, p2.name, f2.name, s2.partition\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p1:Pangenome)<-[:IS_IN_PANGENOME]-(f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH p1, f1, s1 MATCH (f1)-[IS_SIMILAR]-(f2:Family) -[:HAS_PARTITION]->(s2) WITH p1, f1, s1, s2, f2 MATCH (p2:Pangenome)<-[IS_IN_PANGENOME]-(f2) WHERE p1.name <> p2.name RETURN p1.name, f1.name, s1.partition, p2.name, f2.name, s2.partition\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p1:Pangenome)<-[:IS_IN_PANGENOME]-(f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH p1, f1, s1 MATCH (f1)-[IS_SIMILAR]-(f2:Family) -[:HAS_PARTITION]->(s2) WITH p1, f1, s1, s2, f2 MATCH (p2:Pangenome)<-[IS_IN_PANGENOME]-(f2) WHERE p1.name <> p2.name RETURN p1.name, f1.name, s1.partition, p2.name, f2.name, s2.partition\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p1:Pangenome)<-[:IS_IN_PANGENOME]-(f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH p1, f1, s1 MATCH (f1)-[IS_SIMILAR]-(f2:Family) -[:HAS_PARTITION]->(s2) WITH p1, f1, s1, s2, f2 MATCH (p2:Pangenome)<-[IS_IN_PANGENOME]-(f2) WHERE p1.name <> p2.name RETURN p1.name, f1.name, s1.partition, p2.name, f2.name, s2.partition\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p1:Pangenome)<-[:IS_IN_PANGENOME]-(f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH p1, f1, s1 MATCH (f1)-[IS_SIMILAR]-(f2:Family) -[:HAS_PARTITION]->(s2) WITH p1, f1, s1, s2, f2 MATCH (p2:Pangenome)<-[IS_IN_PANGENOME]-(f2) WHERE p1.name <> p2.name RETURN p1.name, f1.name, s1.partition, p2.name, f2.name, s2.partition\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p1:Pangenome)<-[:IS_IN_PANGENOME]-(f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH p1, f1, s1 MATCH (f1)-[IS_SIMILAR]-(f2:Family) -[:HAS_PARTITION]->(s2) WITH p1, f1, s1, s2, f2 MATCH (p2:Pangenome)<-[IS_IN_PANGENOME]-(f2) WHERE p1.name <> p2.name RETURN p1.name, f1.name, s1.partition, p2.name, f2.name, s2.partition\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p1:Pangenome)<-[:IS_IN_PANGENOME]-(f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH p1, f1, s1 MATCH (f1)-[IS_SIMILAR]-(f2:Family) -[:HAS_PARTITION]->(s2) WITH p1, f1, s1, s2, f2 MATCH (p2:Pangenome)<-[IS_IN_PANGENOME]-(f2) WHERE p1.name <> p2.name RETURN p1.name, f1.name, s1.partition, p2.name, f2.name, s2.partition\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p1:Pangenome)<-[:IS_IN_PANGENOME]-(f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH p1, f1, s1 MATCH (f1)-[IS_SIMILAR]-(f2:Family) -[:HAS_PARTITION]->(s2) WITH p1, f1, s1, s2, f2 MATCH (p2:Pangenome)<-[IS_IN_PANGENOME]-(f2) WHERE p1.name <> p2.name RETURN p1.name, f1.name, s1.partition, p2.name, f2.name, s2.partition\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p1:Pangenome)<-[:IS_IN_PANGENOME]-(f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH p1, f1, s1 MATCH (f1)-[IS_SIMILAR]-(f2:Family) -[:HAS_PARTITION]->(s2) WITH p1, f1, s1, s2, f2 MATCH (p2:Pangenome)<-[IS_IN_PANGENOME]-(f2) WHERE p1.name <> p2.name RETURN p1.name, f1.name, s1.partition, p2.name, f2.name, s2.partition\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (p1:Pangenome)<-[:IS_IN_PANGENOME]-(f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH p1, f1, s1 MATCH (f1)-[IS_SIMILAR]-(f2:Family) -[:HAS_PARTITION]->(s2) WITH p1, f1, s1, s2, f2 MATCH (p2:Pangenome)<-[IS_IN_PANGENOME]-(f2) WHERE p1.name <> p2.name RETURN p1.name, f1.name, s1.partition, p2.name, f2.name, s2.partition\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH a=(p:Partition)<-[HAS_PARTITION]-(f1:Family) -[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]-(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WITH a,p,f1,g,r,s MATCH b=(z1:Pangenome)-[:IS_IN_PANGENOME]-(f1)-[:IS_SIMILAR]-(f2:Family)-[:IS_IN_PANGENOME]-(z2:Pangenome) WHERE f1.annotation IS NOT NULL AND z1<>z2 WITH a,b,p,f1,g,r,s,z1,z2,f2 MATCH c=(f2)-[:HAS_PARTITION]-(p2)RETURN a,b,c\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH a=(p:Partition)<-[HAS_PARTITION]-(f1:Family) -[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]-(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WITH a,p,f1,g,r,s MATCH b=(z1:Pangenome)-[:IS_IN_PANGENOME]-(f1)-[:IS_SIMILAR]-(f2:Family)-[:IS_IN_PANGENOME]-(z2:Pangenome) WHERE f1.annotation IS NOT NULL AND z1<>z2 WITH a,b,p,f1,g,r,s,z1,z2,f2 MATCH c=(f2)-[:HAS_PARTITION]-(p2)RETURN a,b,c\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH a=(p:Partition)<-[HAS_PARTITION]-(f1:Family) -[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]-(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WITH a,p,f1,g,r,s MATCH b=(z1:Pangenome)-[:IS_IN_PANGENOME]-(f1)-[:IS_SIMILAR]-(f2:Family)-[:IS_IN_PANGENOME]-(z2:Pangenome) WHERE f1.annotation IS NOT NULL AND z1<>z2 WITH a,b,p,f1,g,r,s,z1,z2,f2 MATCH c=(f2)-[:HAS_PARTITION]-(p2)RETURN a,b,c\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH a=(p:Partition)<-[HAS_PARTITION]-(f1:Family) -[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]-(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WITH a,p,f1,g,r,s MATCH b=(z1:Pangenome)-[:IS_IN_PANGENOME]-(f1)-[:IS_SIMILAR]-(f2:Family)-[:IS_IN_PANGENOME]-(z2:Pangenome) WHERE f1.annotation IS NOT NULL AND z1<>z2 WITH a,b,p,f1,g,r,s,z1,z2,f2 MATCH c=(f2)-[:HAS_PARTITION]-(p2)RETURN a,b,c\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH a=(p:Partition)<-[HAS_PARTITION]-(f1:Family) -[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]-(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WITH a,p,f1,g,r,s MATCH b=(z1:Pangenome)-[:IS_IN_PANGENOME]-(f1)-[:IS_SIMILAR]-(f2:Family)-[:IS_IN_PANGENOME]-(z2:Pangenome) WHERE f1.annotation IS NOT NULL AND z1<>z2 WITH a,b,p,f1,g,r,s,z1,z2,f2 MATCH c=(f2)-[:HAS_PARTITION]-(p2)RETURN a,b,c\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH a=(p:Partition)<-[HAS_PARTITION]-(f1:Family) -[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]-(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WITH a,p,f1,g,r,s MATCH b=(z1:Pangenome)-[:IS_IN_PANGENOME]-(f1)-[:IS_SIMILAR]-(f2:Family)-[:IS_IN_PANGENOME]-(z2:Pangenome) WHERE f1.annotation IS NOT NULL AND z1<>z2 WITH a,b,p,f1,g,r,s,z1,z2,f2 MATCH c=(f2)-[:HAS_PARTITION]-(p2)RETURN a,b,c\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH a=(p:Partition)<-[HAS_PARTITION]-(f1:Family) -[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]-(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WITH a,p,f1,g,r,s MATCH b=(z1:Pangenome)-[:IS_IN_PANGENOME]-(f1)-[:IS_SIMILAR]-(f2:Family)-[:IS_IN_PANGENOME]-(z2:Pangenome) WHERE f1.annotation IS NOT NULL AND z1<>z2 WITH a,b,p,f1,g,r,s,z1,z2,f2 MATCH c=(f2)-[:HAS_PARTITION]-(p2)RETURN a,b,c\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH a=(p:Partition)<-[HAS_PARTITION]-(f1:Family) -[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]-(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WITH a,p,f1,g,r,s MATCH b=(z1:Pangenome)-[:IS_IN_PANGENOME]-(f1)-[:IS_SIMILAR]-(f2:Family)-[:IS_IN_PANGENOME]-(z2:Pangenome) WHERE f1.annotation IS NOT NULL AND z1<>z2 WITH a,b,p,f1,g,r,s,z1,z2,f2 MATCH c=(f2)-[:HAS_PARTITION]-(p2)RETURN a,b,c\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH a=(p:Partition)<-[HAS_PARTITION]-(f1:Family) -[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]-(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WITH a,p,f1,g,r,s MATCH b=(z1:Pangenome)-[:IS_IN_PANGENOME]-(f1)-[:IS_SIMILAR]-(f2:Family)-[:IS_IN_PANGENOME]-(z2:Pangenome) WHERE f1.annotation IS NOT NULL AND z1<>z2 WITH a,b,p,f1,g,r,s,z1,z2,f2 MATCH c=(f2)-[:HAS_PARTITION]-(p2)RETURN a,b,c\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH a=(p:Partition)<-[HAS_PARTITION]-(f1:Family) -[:IS_IN_FAMILY]-(g:Gene)-[:IS_IN_RGP]-(r:RGP)-[:IS_IN_SPOT]-(s:Spot) WITH a,p,f1,g,r,s MATCH b=(z1:Pangenome)-[:IS_IN_PANGENOME]-(f1)-[:IS_SIMILAR]-(f2:Family)-[:IS_IN_PANGENOME]-(z2:Pangenome) WHERE f1.annotation IS NOT NULL AND z1<>z2 WITH a,b,p,f1,g,r,s,z1,z2,f2 MATCH c=(f2)-[:HAS_PARTITION]-(p2)RETURN a,b,c\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (m1:Module)<-[:IS_IN_MODULE]-(f1:Family)-[:IS_IN_PANGENOME]->(p1:Pangenome) WITH f1, m1, p1 MATCH (p2:Pangenome)<-[:IS_IN_PANGENOME]-(f2:Family)-[:IS_SIMILAR]-(f1) WITH f1, m1, p1, f2, p2 MATCH (f2)-[:IS_IN_MODULE]-(m2:Module) WHERE f1.annotation IS NOT NULL AND p1 <> p2 RETURN p1, p2, m1, m2, f1, f2\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (m1:Module)<-[:IS_IN_MODULE]-(f1:Family)-[:IS_IN_PANGENOME]->(p1:Pangenome) WITH f1, m1, p1 MATCH (p2:Pangenome)<-[:IS_IN_PANGENOME]-(f2:Family)-[:IS_SIMILAR]-(f1) WITH f1, m1, p1, f2, p2 MATCH (f2)-[:IS_IN_MODULE]-(m2:Module) WHERE f1.annotation IS NOT NULL AND p1 <> p2 RETURN p1, p2, m1, m2, f1, f2\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (m1:Module)<-[:IS_IN_MODULE]-(f1:Family)-[:IS_IN_PANGENOME]->(p1:Pangenome) WITH f1, m1, p1 MATCH (p2:Pangenome)<-[:IS_IN_PANGENOME]-(f2:Family)-[:IS_SIMILAR]-(f1) WITH f1, m1, p1, f2, p2 MATCH (f2)-[:IS_IN_MODULE]-(m2:Module) WHERE f1.annotation IS NOT NULL AND p1 <> p2 RETURN p1, p2, m1, m2, f1, f2\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (m1:Module)<-[:IS_IN_MODULE]-(f1:Family)-[:IS_IN_PANGENOME]->(p1:Pangenome) WITH f1, m1, p1 MATCH (p2:Pangenome)<-[:IS_IN_PANGENOME]-(f2:Family)-[:IS_SIMILAR]-(f1) WITH f1, m1, p1, f2, p2 MATCH (f2)-[:IS_IN_MODULE]-(m2:Module) WHERE f1.annotation IS NOT NULL AND p1 <> p2 RETURN p1, p2, m1, m2, f1, f2\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (m1:Module)<-[:IS_IN_MODULE]-(f1:Family)-[:IS_IN_PANGENOME]->(p1:Pangenome) WITH f1, m1, p1 MATCH (p2:Pangenome)<-[:IS_IN_PANGENOME]-(f2:Family)-[:IS_SIMILAR]-(f1) WITH f1, m1, p1, f2, p2 MATCH (f2)-[:IS_IN_MODULE]-(m2:Module) WHERE f1.annotation IS NOT NULL AND p1 <> p2 RETURN p1, p2, m1, m2, f1, f2\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (m1:Module)<-[:IS_IN_MODULE]-(f1:Family)-[:IS_IN_PANGENOME]->(p1:Pangenome) WITH f1, m1, p1 MATCH (p2:Pangenome)<-[:IS_IN_PANGENOME]-(f2:Family)-[:IS_SIMILAR]-(f1) WITH f1, m1, p1, f2, p2 MATCH (f2)-[:IS_IN_MODULE]-(m2:Module) WHERE f1.annotation IS NOT NULL AND p1 <> p2 RETURN p1, p2, m1, m2, f1, f2\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (m1:Module)<-[:IS_IN_MODULE]-(f1:Family)-[:IS_IN_PANGENOME]->(p1:Pangenome) WITH f1, m1, p1 MATCH (p2:Pangenome)<-[:IS_IN_PANGENOME]-(f2:Family)-[:IS_SIMILAR]-(f1) WITH f1, m1, p1, f2, p2 MATCH (f2)-[:IS_IN_MODULE]-(m2:Module) WHERE f1.annotation IS NOT NULL AND p1 <> p2 RETURN p1, p2, m1, m2, f1, f2\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (m1:Module)<-[:IS_IN_MODULE]-(f1:Family)-[:IS_IN_PANGENOME]->(p1:Pangenome) WITH f1, m1, p1 MATCH (p2:Pangenome)<-[:IS_IN_PANGENOME]-(f2:Family)-[:IS_SIMILAR]-(f1) WITH f1, m1, p1, f2, p2 MATCH (f2)-[:IS_IN_MODULE]-(m2:Module) WHERE f1.annotation IS NOT NULL AND p1 <> p2 RETURN p1, p2, m1, m2, f1, f2\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (m1:Module)<-[:IS_IN_MODULE]-(f1:Family)-[:IS_IN_PANGENOME]->(p1:Pangenome) WITH f1, m1, p1 MATCH (p2:Pangenome)<-[:IS_IN_PANGENOME]-(f2:Family)-[:IS_SIMILAR]-(f1) WITH f1, m1, p1, f2, p2 MATCH (f2)-[:IS_IN_MODULE]-(m2:Module) WHERE f1.annotation IS NOT NULL AND p1 <> p2 RETURN p1, p2, m1, m2, f1, f2\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (m1:Module)<-[:IS_IN_MODULE]-(f1:Family)-[:IS_IN_PANGENOME]->(p1:Pangenome) WITH f1, m1, p1 MATCH (p2:Pangenome)<-[:IS_IN_PANGENOME]-(f2:Family)-[:IS_SIMILAR]-(f1) WITH f1, m1, p1, f2, p2 MATCH (f2)-[:IS_IN_MODULE]-(m2:Module) WHERE f1.annotation IS NOT NULL AND p1 <> p2 RETURN p1, p2, m1, m2, f1, f2\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH f1, s1 MATCH (f1)-[r1:IS_SIMILAR]->(f2) -[:HAS_PARTITION]->(s2:Partition) WHERE r1.identity >= 0.8 AND r1.coverage >= 0.8 RETURN f1.name, s1.partition, f2.name, s2.partition, r1.identity, r1.coverage,f1.annotation, f2.annotation ORDER BY r1.identity LIMIT 10\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH f1, s1 MATCH (f1)-[r1:IS_SIMILAR]->(f2) -[:HAS_PARTITION]->(s2:Partition) WHERE r1.identity >= 0.8 AND r1.coverage >= 0.8 RETURN f1.name, s1.partition, f2.name, s2.partition, r1.identity, r1.coverage,f1.annotation, f2.annotation ORDER BY r1.identity LIMIT 10\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH f1, s1 MATCH (f1)-[r1:IS_SIMILAR]->(f2) -[:HAS_PARTITION]->(s2:Partition) WHERE r1.identity >= 0.8 AND r1.coverage >= 0.8 RETURN f1.name, s1.partition, f2.name, s2.partition, r1.identity, r1.coverage,f1.annotation, f2.annotation ORDER BY r1.identity LIMIT 10\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH f1, s1 MATCH (f1)-[r1:IS_SIMILAR]->(f2) -[:HAS_PARTITION]->(s2:Partition) WHERE r1.identity >= 0.8 AND r1.coverage >= 0.8 RETURN f1.name, s1.partition, f2.name, s2.partition, r1.identity, r1.coverage,f1.annotation, f2.annotation ORDER BY r1.identity LIMIT 10\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH f1, s1 MATCH (f1)-[r1:IS_SIMILAR]->(f2) -[:HAS_PARTITION]->(s2:Partition) WHERE r1.identity >= 0.8 AND r1.coverage >= 0.8 RETURN f1.name, s1.partition, f2.name, s2.partition, r1.identity, r1.coverage,f1.annotation, f2.annotation ORDER BY r1.identity LIMIT 10\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH f1, s1 MATCH (f1)-[r1:IS_SIMILAR]->(f2) -[:HAS_PARTITION]->(s2:Partition) WHERE r1.identity >= 0.8 AND r1.coverage >= 0.8 RETURN f1.name, s1.partition, f2.name, s2.partition, r1.identity, r1.coverage,f1.annotation, f2.annotation ORDER BY r1.identity LIMIT 10\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH f1, s1 MATCH (f1)-[r1:IS_SIMILAR]->(f2) -[:HAS_PARTITION]->(s2:Partition) WHERE r1.identity >= 0.8 AND r1.coverage >= 0.8 RETURN f1.name, s1.partition, f2.name, s2.partition, r1.identity, r1.coverage,f1.annotation, f2.annotation ORDER BY r1.identity LIMIT 10\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH f1, s1 MATCH (f1)-[r1:IS_SIMILAR]->(f2) -[:HAS_PARTITION]->(s2:Partition) WHERE r1.identity >= 0.8 AND r1.coverage >= 0.8 RETURN f1.name, s1.partition, f2.name, s2.partition, r1.identity, r1.coverage,f1.annotation, f2.annotation ORDER BY r1.identity LIMIT 10\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH f1, s1 MATCH (f1)-[r1:IS_SIMILAR]->(f2) -[:HAS_PARTITION]->(s2:Partition) WHERE r1.identity >= 0.8 AND r1.coverage >= 0.8 RETURN f1.name, s1.partition, f2.name, s2.partition, r1.identity, r1.coverage,f1.annotation, f2.annotation ORDER BY r1.identity LIMIT 10\n",
      "CALL db.clearQueryCaches()\n",
      "CALL apoc.warmup.run()\n",
      "MATCH (f1:Family)-[:HAS_PARTITION]->(s1:Partition) WITH f1, s1 MATCH (f1)-[r1:IS_SIMILAR]->(f2) -[:HAS_PARTITION]->(s2:Partition) WHERE r1.identity >= 0.8 AND r1.coverage >= 0.8 RETURN f1.name, s1.partition, f2.name, s2.partition, r1.identity, r1.coverage,f1.annotation, f2.annotation ORDER BY r1.identity LIMIT 10\n",
      "CALL db.clearQueryCaches()\n"
     ]
    },
    {
     "data": {
      "text/plain": "         Mean     Stdev   Mediane       Min       Max\nQ1   0.083915  0.046694  0.069744  0.048335  0.204392\nQ2   0.056861  0.007986  0.055250  0.045735  0.074058\nQ3   0.747413  0.113417  0.724246  0.657622  1.053809\nQ4   1.490831  0.132928  1.450817  1.327790  1.781401\nQ5   0.046255  0.005752  0.044578  0.042577  0.062443\nQ6   0.044386  0.003446  0.043214  0.039686  0.050909\nQ7   2.436580  0.067438  2.420400  2.357164  2.563146\nQ8   1.271319  0.172032  1.198555  1.097136  1.658775\nQ9   0.074360  0.015719  0.064597  0.061571  0.106834\nQ10  0.071507  0.035591  0.058783  0.051264  0.170405",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>Mean</th>\n      <th>Stdev</th>\n      <th>Mediane</th>\n      <th>Min</th>\n      <th>Max</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>Q1</th>\n      <td>0.083915</td>\n      <td>0.046694</td>\n      <td>0.069744</td>\n      <td>0.048335</td>\n      <td>0.204392</td>\n    </tr>\n    <tr>\n      <th>Q2</th>\n      <td>0.056861</td>\n      <td>0.007986</td>\n      <td>0.055250</td>\n      <td>0.045735</td>\n      <td>0.074058</td>\n    </tr>\n    <tr>\n      <th>Q3</th>\n      <td>0.747413</td>\n      <td>0.113417</td>\n      <td>0.724246</td>\n      <td>0.657622</td>\n      <td>1.053809</td>\n    </tr>\n    <tr>\n      <th>Q4</th>\n      <td>1.490831</td>\n      <td>0.132928</td>\n      <td>1.450817</td>\n      <td>1.327790</td>\n      <td>1.781401</td>\n    </tr>\n    <tr>\n      <th>Q5</th>\n      <td>0.046255</td>\n      <td>0.005752</td>\n      <td>0.044578</td>\n      <td>0.042577</td>\n      <td>0.062443</td>\n    </tr>\n    <tr>\n      <th>Q6</th>\n      <td>0.044386</td>\n      <td>0.003446</td>\n      <td>0.043214</td>\n      <td>0.039686</td>\n      <td>0.050909</td>\n    </tr>\n    <tr>\n      <th>Q7</th>\n      <td>2.436580</td>\n      <td>0.067438</td>\n      <td>2.420400</td>\n      <td>2.357164</td>\n      <td>2.563146</td>\n    </tr>\n    <tr>\n      <th>Q8</th>\n      <td>1.271319</td>\n      <td>0.172032</td>\n      <td>1.198555</td>\n      <td>1.097136</td>\n      <td>1.658775</td>\n    </tr>\n    <tr>\n      <th>Q9</th>\n      <td>0.074360</td>\n      <td>0.015719</td>\n      <td>0.064597</td>\n      <td>0.061571</td>\n      <td>0.106834</td>\n    </tr>\n    <tr>\n      <th>Q10</th>\n      <td>0.071507</td>\n      <td>0.035591</td>\n      <td>0.058783</td>\n      <td>0.051264</td>\n      <td>0.170405</td>\n    </tr>\n  </tbody>\n</table>\n</div>"
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from time import time\n",
    "\n",
    "def launch_query(query):\n",
    "    print(query)\n",
    "    try:\n",
    "        res = graph.run(query)\n",
    "    except Exception as errror:\n",
    "        raise Exception(f\"Query : '{query}' failed because of the following errror\\n{errror}\")\n",
    "    else:\n",
    "        return  res\n",
    "\n",
    "def launch_WF():\n",
    "    from script.python.wf import WF\n",
    "    from statistics import mean, median, stdev\n",
    "    import pandas as pd\n",
    "\n",
    "    nb_rep = 10\n",
    "    stat_dict = {}\n",
    "    for q_id, query in WF.items():\n",
    "        list_q_time = []\n",
    "        for i in range(nb_rep):\n",
    "            launch_query(\"CALL apoc.warmup.run()\")\n",
    "            begin_time = time()\n",
    "            res = launch_query(query)\n",
    "            q_time = time() - begin_time\n",
    "            list_q_time.append(q_time)\n",
    "            launch_query(\"CALL db.clearQueryCaches()\")\n",
    "        stat_dict[q_id] = [mean(list_q_time), stdev(list_q_time), median(list_q_time),\n",
    "                           min(list_q_time), max(list_q_time)]\n",
    "    return pd.DataFrame.from_dict(stat_dict, orient='index', columns=[\"Mean\", \"Stdev\", \"Mediane\", \"Min\", \"Max\"])\n",
    "\n",
    "launch_WF()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "id": "bedc22b2",
   "metadata": {},
   "source": [
    "# Suplementary code\n",
    "## Genomes downloading\n",
    "Step do download genomes necessary to consctruct pangenomes. Note that genomes database are in constant evolution and your pangenome could be different than our."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-------------------------------------------\r\n",
      "┌─┐┌─┐┌┐┌┌─┐┌┬┐┌─┐    ┬ ┬┌─┐┌┬┐┌─┐┌┬┐┌─┐┬─┐\r\n",
      "│ ┬├┤ ││││ ││││├┤     │ │├─┘ ││├─┤ │ ├┤ ├┬┘\r\n",
      "└─┘└─┘┘└┘└─┘┴ ┴└─┘────└─┘┴  ─┴┘┴ ┴ ┴ └─┘┴└─\r\n",
      "                                     v0.5.1 \r\n",
      "-------------------------------------------\r\n",
      "Mode: NEW \r\n",
      "Args: -T 's__Acinetobacter baumannii' -M 'gtdb' -t '6' -o 'data/GBFF/Acinetobacter.baumannii' -l 'complete genome' -f 'assembly_report.txt,genomic.gbff.gz' -d 'genbank,refseq'\r\n",
      "Outp: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Acinetobacter.baumannii/\r\n",
      "-------------------------------------\r\n",
      "Downloading assembly summary [2022-11-21_10-51-39]\r\n",
      " - Database [genbank,refseq]\r\n",
      " - 1821534 assembly entries available\r\n",
      "\r\n",
      "Filtering assembly summary [2022-11-21_10-51-39]\r\n",
      " - Downloading taxonomy (gtdb)\r\n",
      " - 1503993 assemblies removed not in GTDB\r\n",
      " - Could not retrieve 1 GTDB assemblies\r\n",
      " - 312124 assemblies removed not in taxids [s__Acinetobacter baumannii]\r\n",
      " - 5132 assemblies removed based on filters:\r\n",
      "   valid URLs\r\n",
      "   assembly level=complete genome\r\n",
      " - 285 assembly entries to download\r\n",
      "\r\n",
      "Downloading 570 files with 6 threads\r\n",
      " - 570/570 files successfully downloaded\r\n",
      "\r\n",
      "-------------------------------------------\r\n",
      "# 570/570 files in the current version\r\n",
      "# Current version: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Acinetobacter.baumannii/2022-11-21_10-51-39\r\n",
      "# Log file       : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Acinetobacter.baumannii/2022-11-21_10-51-39/2022-11-21_10-51-39.log\r\n",
      "# History        : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Acinetobacter.baumannii/history.tsv\r\n",
      "-------------------------------------------\r\n",
      "-------------------------------------------\r\n",
      "┌─┐┌─┐┌┐┌┌─┐┌┬┐┌─┐    ┬ ┬┌─┐┌┬┐┌─┐┌┬┐┌─┐┬─┐\r\n",
      "│ ┬├┤ ││││ ││││├┤     │ │├─┘ ││├─┤ │ ├┤ ├┬┘\r\n",
      "└─┘└─┘┘└┘└─┘┴ ┴└─┘────└─┘┴  ─┴┘┴ ┴ ┴ └─┘┴└─\r\n",
      "                                     v0.5.1 \r\n",
      "-------------------------------------------\r\n",
      "Mode: NEW \r\n",
      "Args: -T 's__Enterobacter bugandensis' -M 'gtdb' -t '6' -o 'data/GBFF/Enterobacter.bugandensis' -f 'assembly_report.txt,genomic.gbff.gz' -d 'genbank,refseq'\r\n",
      "Outp: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.bugandensis/\r\n",
      "-------------------------------------\r\n",
      "Downloading assembly summary [2022-11-21_10-54-09]\r\n",
      " - Database [genbank,refseq]\r\n",
      " - 1821534 assembly entries available\r\n",
      "\r\n",
      "Filtering assembly summary [2022-11-21_10-54-09]\r\n",
      " - Downloading taxonomy (gtdb)\r\n",
      " - 1503993 assemblies removed not in GTDB\r\n",
      " - Could not retrieve 1 GTDB assemblies\r\n",
      " - 317426 assemblies removed not in taxids [s__Enterobacter bugandensis]\r\n",
      " - 115 assembly entries to download\r\n",
      "\r\n",
      "Downloading 230 files with 6 threads\r\n",
      " - 230/230 files successfully downloaded\r\n",
      "\r\n",
      "-------------------------------------------\r\n",
      "# 230/230 files in the current version\r\n",
      "# Current version: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.bugandensis/2022-11-21_10-54-09\r\n",
      "# Log file       : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.bugandensis/2022-11-21_10-54-09/2022-11-21_10-54-09.log\r\n",
      "# History        : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.bugandensis/history.tsv\r\n",
      "-------------------------------------------\r\n",
      "-------------------------------------------\r\n",
      "┌─┐┌─┐┌┐┌┌─┐┌┬┐┌─┐    ┬ ┬┌─┐┌┬┐┌─┐┌┬┐┌─┐┬─┐\r\n",
      "│ ┬├┤ ││││ ││││├┤     │ │├─┘ ││├─┤ │ ├┤ ├┬┘\r\n",
      "└─┘└─┘┘└┘└─┘┴ ┴└─┘────└─┘┴  ─┴┘┴ ┴ ┴ └─┘┴└─\r\n",
      "                                     v0.5.1 \r\n",
      "-------------------------------------------\r\n",
      "Mode: NEW \r\n",
      "Args: -T 's__Enterobacter cloacae' -M 'gtdb' -t '6' -o 'data/GBFF/Enterobacter.cloacae' -f 'assembly_report.txt,genomic.gbff.gz' -d 'genbank,refseq'\r\n",
      "Outp: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.cloacae/\r\n",
      "-------------------------------------\r\n",
      "Downloading assembly summary [2022-11-21_10-56-01]\r\n",
      " - Database [genbank,refseq]\r\n",
      " - 1821534 assembly entries available\r\n",
      "\r\n",
      "Filtering assembly summary [2022-11-21_10-56-01]\r\n",
      " - Downloading taxonomy (gtdb)\r\n",
      " - 1503993 assemblies removed not in GTDB\r\n",
      " - Could not retrieve 1 GTDB assemblies\r\n",
      " - 317412 assemblies removed not in taxids [s__Enterobacter cloacae]\r\n",
      " - 129 assembly entries to download\r\n",
      "\r\n",
      "Downloading 258 files with 6 threads\r\n",
      " - 258/258 files successfully downloaded\r\n",
      "\r\n",
      "-------------------------------------------\r\n",
      "# 258/258 files in the current version\r\n",
      "# Current version: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.cloacae/2022-11-21_10-56-01\r\n",
      "# Log file       : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.cloacae/2022-11-21_10-56-01/2022-11-21_10-56-01.log\r\n",
      "# History        : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.cloacae/history.tsv\r\n",
      "-------------------------------------------\r\n",
      "-------------------------------------------\r\n",
      "┌─┐┌─┐┌┐┌┌─┐┌┬┐┌─┐    ┬ ┬┌─┐┌┬┐┌─┐┌┬┐┌─┐┬─┐\r\n",
      "│ ┬├┤ ││││ ││││├┤     │ │├─┘ ││├─┤ │ ├┤ ├┬┘\r\n",
      "└─┘└─┘┘└┘└─┘┴ ┴└─┘────└─┘┴  ─┴┘┴ ┴ ┴ └─┘┴└─\r\n",
      "                                     v0.5.1 \r\n",
      "-------------------------------------------\r\n",
      "Mode: NEW \r\n",
      "Args: -T 's__Enterobacter hormaechei_A' -M 'gtdb' -t '6' -o 'data/GBFF/Enterobacter.hormaechei_A' -l 'complete genome' -f 'assembly_report.txt,genomic.gbff.gz' -d 'genbank,refseq'\r\n",
      "Outp: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.hormaechei_A/\r\n",
      "-------------------------------------\r\n",
      "Downloading assembly summary [2022-11-21_10-57-56]\r\n",
      " - Database [genbank,refseq]\r\n",
      " - 1821534 assembly entries available\r\n",
      "\r\n",
      "Filtering assembly summary [2022-11-21_10-57-56]\r\n",
      " - Downloading taxonomy (gtdb)\r\n",
      " - 1503993 assemblies removed not in GTDB\r\n",
      " - Could not retrieve 1 GTDB assemblies\r\n",
      " - 315674 assemblies removed not in taxids [s__Enterobacter hormaechei_A]\r\n",
      " - 1708 assemblies removed based on filters:\r\n",
      "   valid URLs\r\n",
      "   assembly level=complete genome\r\n",
      " - 159 assembly entries to download\r\n",
      "\r\n",
      "Downloading 318 files with 6 threads\r\n",
      " - 318/318 files successfully downloaded\r\n",
      "\r\n",
      "-------------------------------------------\r\n",
      "# 318/318 files in the current version\r\n",
      "# Current version: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.hormaechei_A/2022-11-21_10-57-56\r\n",
      "# Log file       : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.hormaechei_A/2022-11-21_10-57-56/2022-11-21_10-57-56.log\r\n",
      "# History        : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.hormaechei_A/history.tsv\r\n",
      "-------------------------------------------\r\n",
      "-------------------------------------------\r\n",
      "┌─┐┌─┐┌┐┌┌─┐┌┬┐┌─┐    ┬ ┬┌─┐┌┬┐┌─┐┌┬┐┌─┐┬─┐\r\n",
      "│ ┬├┤ ││││ ││││├┤     │ │├─┘ ││├─┤ │ ├┤ ├┬┘\r\n",
      "└─┘└─┘┘└┘└─┘┴ ┴└─┘────└─┘┴  ─┴┘┴ ┴ ┴ └─┘┴└─\r\n",
      "                                     v0.5.1 \r\n",
      "-------------------------------------------\r\n",
      "Mode: NEW \r\n",
      "Args: -T 's__Enterobacter kobei' -M 'gtdb' -t '6' -o 'data/GBFF/Enterobacter.kobei' -f 'assembly_report.txt,genomic.gbff.gz' -d 'genbank,refseq'\r\n",
      "Outp: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.kobei/\r\n",
      "-------------------------------------\r\n",
      "Downloading assembly summary [2022-11-21_10-59-57]\r\n",
      " - Database [genbank,refseq]\r\n",
      " - 1821534 assembly entries available\r\n",
      "\r\n",
      "Filtering assembly summary [2022-11-21_10-59-57]\r\n",
      " - Downloading taxonomy (gtdb)\r\n",
      " - 1503993 assemblies removed not in GTDB\r\n",
      " - Could not retrieve 1 GTDB assemblies\r\n",
      " - 317400 assemblies removed not in taxids [s__Enterobacter kobei]\r\n",
      " - 141 assembly entries to download\r\n",
      "\r\n",
      "Downloading 282 files with 6 threads\r\n",
      " - 282/282 files successfully downloaded\r\n",
      "\r\n",
      "-------------------------------------------\r\n",
      "# 282/282 files in the current version\r\n",
      "# Current version: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.kobei/2022-11-21_10-59-57\r\n",
      "# Log file       : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.kobei/2022-11-21_10-59-57/2022-11-21_10-59-57.log\r\n",
      "# History        : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.kobei/history.tsv\r\n",
      "-------------------------------------------\r\n",
      "-------------------------------------------\r\n",
      "┌─┐┌─┐┌┐┌┌─┐┌┬┐┌─┐    ┬ ┬┌─┐┌┬┐┌─┐┌┬┐┌─┐┬─┐\r\n",
      "│ ┬├┤ ││││ ││││├┤     │ │├─┘ ││├─┤ │ ├┤ ├┬┘\r\n",
      "└─┘└─┘┘└┘└─┘┴ ┴└─┘────└─┘┴  ─┴┘┴ ┴ ┴ └─┘┴└─\r\n",
      "                                     v0.5.1 \r\n",
      "-------------------------------------------\r\n",
      "Mode: NEW \r\n",
      "Args: -T 's__Enterobacter roggenkampii' -M 'gtdb' -t '6' -o 'data/GBFF/Enterobacter.roggenkampii' -f 'assembly_report.txt,genomic.gbff.gz' -d 'genbank,refseq'\r\n",
      "Outp: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.roggenkampii/\r\n",
      "-------------------------------------\r\n",
      "Downloading assembly summary [2022-11-21_11-01-57]\r\n",
      " - Database [genbank,refseq]\r\n",
      " - 1821534 assembly entries available\r\n",
      "\r\n",
      "Filtering assembly summary [2022-11-21_11-01-57]\r\n",
      " - Downloading taxonomy (gtdb)\r\n",
      " - 1503993 assemblies removed not in GTDB\r\n",
      " - Could not retrieve 1 GTDB assemblies\r\n",
      " - 317344 assemblies removed not in taxids [s__Enterobacter roggenkampii]\r\n",
      " - 197 assembly entries to download\r\n",
      "\r\n",
      "Downloading 394 files with 6 threads\r\n",
      " - 394/394 files successfully downloaded\r\n",
      "\r\n",
      "-------------------------------------------\r\n",
      "# 394/394 files in the current version\r\n",
      "# Current version: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.roggenkampii/2022-11-21_11-01-57\r\n",
      "# Log file       : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.roggenkampii/2022-11-21_11-01-57/2022-11-21_11-01-57.log\r\n",
      "# History        : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterobacter.roggenkampii/history.tsv\r\n",
      "-------------------------------------------\r\n",
      "-------------------------------------------\r\n",
      "┌─┐┌─┐┌┐┌┌─┐┌┬┐┌─┐    ┬ ┬┌─┐┌┬┐┌─┐┌┬┐┌─┐┬─┐\r\n",
      "│ ┬├┤ ││││ ││││├┤     │ │├─┘ ││├─┤ │ ├┤ ├┬┘\r\n",
      "└─┘└─┘┘└┘└─┘┴ ┴└─┘────└─┘┴  ─┴┘┴ ┴ ┴ └─┘┴└─\r\n",
      "                                     v0.5.1 \r\n",
      "-------------------------------------------\r\n",
      "Mode: NEW \r\n",
      "Args: -T 's__Enterococcus_B faecium' -M 'gtdb' -t '6' -o 'data/GBFF/Enterococcus_B.faecium' -l 'complete genome' -f 'assembly_report.txt,genomic.gbff.gz' -d 'genbank,refseq'\r\n",
      "Outp: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterococcus_B.faecium/\r\n",
      "-------------------------------------\r\n",
      "Downloading assembly summary [2022-11-21_11-04-10]\r\n",
      " - Database [genbank,refseq]\r\n",
      " - 1821534 assembly entries available\r\n",
      "\r\n",
      "Filtering assembly summary [2022-11-21_11-04-10]\r\n",
      " - Downloading taxonomy (gtdb)\r\n",
      " - 1503993 assemblies removed not in GTDB\r\n",
      " - Could not retrieve 1 GTDB assemblies\r\n",
      " - 315364 assemblies removed not in taxids [s__Enterococcus_B faecium]\r\n",
      " - 1970 assemblies removed based on filters:\r\n",
      "   valid URLs\r\n",
      "   assembly level=complete genome\r\n",
      " - 207 assembly entries to download\r\n",
      "\r\n",
      "Downloading 414 files with 6 threads\r\n",
      " - 414/414 files successfully downloaded\r\n",
      "\r\n",
      "-------------------------------------------\r\n",
      "# 414/414 files in the current version\r\n",
      "# Current version: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterococcus_B.faecium/2022-11-21_11-04-10\r\n",
      "# Log file       : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterococcus_B.faecium/2022-11-21_11-04-10/2022-11-21_11-04-10.log\r\n",
      "# History        : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Enterococcus_B.faecium/history.tsv\r\n",
      "-------------------------------------------\r\n",
      "-------------------------------------------\r\n",
      "┌─┐┌─┐┌┐┌┌─┐┌┬┐┌─┐    ┬ ┬┌─┐┌┬┐┌─┐┌┬┐┌─┐┬─┐\r\n",
      "│ ┬├┤ ││││ ││││├┤     │ │├─┘ ││├─┤ │ ├┤ ├┬┘\r\n",
      "└─┘└─┘┘└┘└─┘┴ ┴└─┘────└─┘┴  ─┴┘┴ ┴ ┴ └─┘┴└─\r\n",
      "                                     v0.5.1 \r\n",
      "-------------------------------------------\r\n",
      "Mode: NEW \r\n",
      "Args: -T 's__Klebsiella pneumoniae' -M 'gtdb' -A '600' -t '6' -o 'data/GBFF/Klebsiella.pneumoniae' -l 'complete genome' -f 'assembly_report.txt,genomic.gbff.gz' -d 'genbank,refseq'\r\n",
      "Outp: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Klebsiella.pneumoniae/\r\n",
      "-------------------------------------\r\n",
      "Downloading assembly summary [2022-11-21_11-06-26]\r\n",
      " - Database [genbank,refseq]\r\n",
      " - 1821534 assembly entries available\r\n",
      "\r\n",
      "Filtering assembly summary [2022-11-21_11-06-26]\r\n",
      " - Downloading taxonomy (gtdb)\r\n",
      " - 1503993 assemblies removed not in GTDB\r\n",
      " - Could not retrieve 1 GTDB assemblies\r\n",
      " - 306247 assemblies removed not in taxids [s__Klebsiella pneumoniae]\r\n",
      " - 10415 assemblies removed based on filters:\r\n",
      "   valid URLs\r\n",
      "   assembly level=complete genome\r\n",
      " - 279 entries removed with top 600\r\n",
      " - 600 assembly entries to download\r\n",
      "\r\n",
      "Downloading 1200 files with 6 threads\r\n",
      " - 1200/1200 files successfully downloaded\r\n",
      "\r\n",
      "-------------------------------------------\r\n",
      "# 1200/1200 files in the current version\r\n",
      "# Current version: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Klebsiella.pneumoniae/2022-11-21_11-06-26\r\n",
      "# Log file       : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Klebsiella.pneumoniae/2022-11-21_11-06-26/2022-11-21_11-06-26.log\r\n",
      "# History        : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Klebsiella.pneumoniae/history.tsv\r\n",
      "-------------------------------------------\r\n",
      "-------------------------------------------\r\n",
      "┌─┐┌─┐┌┐┌┌─┐┌┬┐┌─┐    ┬ ┬┌─┐┌┬┐┌─┐┌┬┐┌─┐┬─┐\r\n",
      "│ ┬├┤ ││││ ││││├┤     │ │├─┘ ││├─┤ │ ├┤ ├┬┘\r\n",
      "└─┘└─┘┘└┘└─┘┴ ┴└─┘────└─┘┴  ─┴┘┴ ┴ ┴ └─┘┴└─\r\n",
      "                                     v0.5.1 \r\n",
      "-------------------------------------------\r\n",
      "Mode: NEW \r\n",
      "Args: -T 's__Pseudomonas aeruginosa' -M 'gtdb' -t '6' -o 'data/GBFF/Pseudomonas.aeruginosa' -l 'complete genome' -f 'assembly_report.txt,genomic.gbff.gz' -d 'genbank,refseq'\r\n",
      "Outp: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Pseudomonas.aeruginosa/\r\n",
      "-------------------------------------\r\n",
      "Downloading assembly summary [2022-11-21_11-10-55]\r\n",
      " - Database [genbank,refseq]\r\n",
      " - 1821534 assembly entries available\r\n",
      "\r\n",
      "Filtering assembly summary [2022-11-21_11-10-55]\r\n",
      " - Downloading taxonomy (gtdb)\r\n",
      " - 1503993 assemblies removed not in GTDB\r\n",
      " - Could not retrieve 1 GTDB assemblies\r\n",
      " - 311918 assemblies removed not in taxids [s__Pseudomonas aeruginosa]\r\n",
      " - 5310 assemblies removed based on filters:\r\n",
      "   valid URLs\r\n",
      "   assembly level=complete genome\r\n",
      " - 313 assembly entries to download\r\n",
      "\r\n",
      "Downloading 626 files with 6 threads\r\n",
      " - 626/626 files successfully downloaded\r\n",
      "\r\n",
      "-------------------------------------------\r\n",
      "# 626/626 files in the current version\r\n",
      "# Current version: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Pseudomonas.aeruginosa/2022-11-21_11-10-55\r\n",
      "# Log file       : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Pseudomonas.aeruginosa/2022-11-21_11-10-55/2022-11-21_11-10-55.log\r\n",
      "# History        : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Pseudomonas.aeruginosa/history.tsv\r\n",
      "-------------------------------------------\r\n",
      "-------------------------------------------\r\n",
      "┌─┐┌─┐┌┐┌┌─┐┌┬┐┌─┐    ┬ ┬┌─┐┌┬┐┌─┐┌┬┐┌─┐┬─┐\r\n",
      "│ ┬├┤ ││││ ││││├┤     │ │├─┘ ││├─┤ │ ├┤ ├┬┘\r\n",
      "└─┘└─┘┘└┘└─┘┴ ┴└─┘────└─┘┴  ─┴┘┴ ┴ ┴ └─┘┴└─\r\n",
      "                                     v0.5.1 \r\n",
      "-------------------------------------------\r\n",
      "Mode: NEW \r\n",
      "Args: -T 's__Staphylococcus aureus' -M 'gtdb' -A '600' -t '6' -o 'data/GBFF/Staphylococcus.aureus' -l 'complete genome' -f 'assembly_report.txt,genomic.gbff.gz' -d 'genbank,refseq'\r\n",
      "Outp: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Staphylococcus.aureus/\r\n",
      "-------------------------------------\r\n",
      "Downloading assembly summary [2022-11-21_11-13-48]\r\n",
      " - Database [genbank,refseq]\r\n",
      " - 1821534 assembly entries available\r\n",
      "\r\n",
      "Filtering assembly summary [2022-11-21_11-13-48]\r\n",
      " - Downloading taxonomy (gtdb)\r\n",
      " - 1503993 assemblies removed not in GTDB\r\n",
      " - Could not retrieve 1 GTDB assemblies\r\n",
      " - 304483 assemblies removed not in taxids [s__Staphylococcus aureus]\r\n",
      " - 12421 assemblies removed based on filters:\r\n",
      "   valid URLs\r\n",
      "   assembly level=complete genome\r\n",
      " - 37 entries removed with top 600\r\n",
      " - 600 assembly entries to download\r\n",
      "\r\n",
      "Downloading 1200 files with 6 threads\r\n",
      " - 1200/1200 files successfully downloaded\r\n",
      "\r\n",
      "-------------------------------------------\r\n",
      "# 1200/1200 files in the current version\r\n",
      "# Current version: /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Staphylococcus.aureus/2022-11-21_11-13-48\r\n",
      "# Log file       : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Staphylococcus.aureus/2022-11-21_11-13-48/2022-11-21_11-13-48.log\r\n",
      "# History        : /home/jarnoux/Projects/PanGraph-DB/data/GBFF/Staphylococcus.aureus/history.tsv\r\n",
      "-------------------------------------------\r\n"
     ]
    }
   ],
   "source": [
    "!mkdir $GBFF\n",
    "!genome_updater.sh -d genbank,refseq -M gtdb -T \"s__Acinetobacter baumannii\" -f assembly_report.txt,genomic.gbff.gz -l \"complete genome\" -o $GBFF/Acinetobacter.baumannii -t $CPU\n",
    "!genome_updater.sh -d genbank,refseq -M gtdb -T \"s__Enterobacter bugandensis\" -f assembly_report.txt,genomic.gbff.gz -o $GBFF/Enterobacter.bugandensis -t $CPU\n",
    "!genome_updater.sh -d genbank,refseq -M gtdb -T \"s__Enterobacter cloacae\" -f assembly_report.txt,genomic.gbff.gz -o $GBFF/Enterobacter.cloacae -t $CPU\n",
    "!genome_updater.sh -d genbank,refseq -M gtdb -T \"s__Enterobacter hormaechei_A\" -f assembly_report.txt,genomic.gbff.gz -l \"complete genome\" -o $GBFF/Enterobacter.hormaechei_A -t $CPU\n",
    "!genome_updater.sh -d genbank,refseq -M gtdb -T \"s__Enterobacter kobei\" -f assembly_report.txt,genomic.gbff.gz -o $GBFF/Enterobacter.kobei -t $CPU\n",
    "!genome_updater.sh -d genbank,refseq -M gtdb -T \"s__Enterobacter roggenkampii\" -f assembly_report.txt,genomic.gbff.gz -o $GBFF/Enterobacter.roggenkampii -t $CPU\n",
    "!genome_updater.sh -d genbank,refseq -M gtdb -T \"s__Enterococcus_B faecium\" -f assembly_report.txt,genomic.gbff.gz -l \"complete genome\" -o $GBFF/Enterococcus_B.faecium -t $CPU\n",
    "!genome_updater.sh -d genbank,refseq -M gtdb -T \"s__Klebsiella pneumoniae\" -f assembly_report.txt,genomic.gbff.gz -l \"complete genome\" -A 600 -o $GBFF/Klebsiella.pneumoniae -t $CPU\n",
    "!genome_updater.sh -d genbank,refseq -M gtdb -T \"s__Pseudomonas aeruginosa\" -f assembly_report.txt,genomic.gbff.gz -l \"complete genome\" -o $GBFF/Pseudomonas.aeruginosa -t $CPU\n",
    "!genome_updater.sh -d genbank,refseq -M gtdb -T \"s__Staphylococcus aureus\" -f assembly_report.txt,genomic.gbff.gz -l \"complete genome\" -A 600 -o $GBFF/Staphylococcus.aureus -t $CPU"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Pangenome construction with PPanGGOLiN\n",
    "### Generate list of genomes"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "outputs": [],
   "source": [
    "! for sp in $(ls $GBFF); do for path in $(ls $GBFF/$sp/*/files/*.gbff.gz);do genome=$(echo $path | cut -d'/' -f6 | cut -d. -f1,2); echo $genome $(pwd)/$path | sed 's/\\s/\\t/'; done > $PANGENOMES/$sp.list; done"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Generate pangenomes"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-11-21 12:00:06 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin all --anno data/pangenomes/Acinetobacter.baumannii.list -o data/pangenomes/Acinetobacter.baumannii -c 6 -f\r\n",
      "2022-11-21 12:00:06 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 12:00:06 annotate.py:l448 INFO\tReading data/pangenomes/Acinetobacter.baumannii.list the list of organism files ...\r\n",
      "100%|███████████████████████████████████████| 285/285 [00:34<00:00,  8.20file/s]\r\n",
      "2022-11-21 12:00:43 annotate.py:l469 INFO\tgene identifiers used in the provided annotation files were unique, PPanGGOLiN will use them.\r\n",
      "2022-11-21 12:00:43 writeBinaries.py:l932 INFO\tWriting genome annotations...\r\n",
      "100%|█████████████████████████████████████| 285/285 [00:02<00:00, 96.23genome/s]\r\n",
      "2022-11-21 12:00:47 writeBinaries.py:l950 INFO\twriting the protein coding gene dna sequences\r\n",
      "100%|███████████████████████████| 1045048/1045048 [00:01<00:00, 651535.64gene/s]\r\n",
      "2022-11-21 12:00:53 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Acinetobacter.baumannii/pangenome.h5\r\n",
      "2022-11-21 12:00:53 writeSequences.py:l39 INFO\tWriting all of the CDS sequences...\r\n",
      "100%|███████████████████████████| 1045048/1045048 [00:01<00:00, 641263.04gene/s]\r\n",
      "2022-11-21 12:00:55 cluster.py:l297 INFO\tClustering all of the genes sequences...\r\n",
      "2022-11-21 12:00:55 cluster.py:l86 INFO\tCreating sequence database...\r\n",
      "2022-11-21 12:01:01 cluster.py:l93 INFO\tClustering sequences...\r\n",
      "2022-11-21 12:01:45 cluster.py:l99 INFO\tExtracting cluster representatives...\r\n",
      "2022-11-21 12:01:46 cluster.py:l108 INFO\tWriting gene to family informations\r\n",
      "2022-11-21 12:01:46 cluster.py:l305 INFO\tAssociating fragments to their original gene family...\r\n",
      "2022-11-21 12:01:46 cluster.py:l153 INFO\tAligning cluster representatives...\r\n",
      "2022-11-21 12:02:13 cluster.py:l159 INFO\tExtracting alignments...\r\n",
      "2022-11-21 12:02:15 cluster.py:l197 INFO\tStarting with 17329 families\r\n",
      "2022-11-21 12:02:15 cluster.py:l231 INFO\tEnding with 14427 gene families\r\n",
      "2022-11-21 12:02:16 cluster.py:l242 INFO\tAdding protein sequences to the gene families\r\n",
      "2022-11-21 12:02:16 cluster.py:l256 INFO\tAdding 1045048 genes to the gene families\r\n",
      "100%|███████████████████████████| 1045048/1045048 [00:03<00:00, 273761.53gene/s]\r\n",
      "2022-11-21 12:02:20 writeBinaries.py:l955 INFO\tWriting gene families and gene associations...\r\n",
      "100%|█████████████████████████| 14427/14427 [00:00<00:00, 15601.49gene family/s]\r\n",
      "2022-11-21 12:02:21 writeBinaries.py:l957 INFO\tWriting gene families information...\r\n",
      "100%|█████████████████████████| 14427/14427 [00:00<00:00, 97409.43gene family/s]\r\n",
      "2022-11-21 12:02:21 writeBinaries.py:l823 INFO\tUpdating annotations with fragment information\r\n",
      "100%|███████████████████████████| 1070798/1070798 [00:03<00:00, 299296.27gene/s]\r\n",
      "2022-11-21 12:02:25 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Acinetobacter.baumannii/pangenome.h5\r\n",
      "2022-11-21 12:02:25 makeGraph.py:l87 INFO\tComputing the neighbors graph...\r\n",
      "Processing GCF_900478145.1_34555_D01_genomic: 100%|█| 285/285 [00:04<00:00, 57.1\r\n",
      "2022-11-21 12:02:30 makeGraph.py:l108 INFO\tDone making the neighbors graph.\r\n",
      "2022-11-21 12:02:30 partition.py:l482 INFO\tEstimating the optimal number of partitions...\r\n",
      "100%|█████████████████████████| 19/19 [01:39<00:00,  5.24s/Number of partitions]\r\n",
      "2022-11-21 12:04:10 partition.py:l485 INFO\tThe number of partitions has been evaluated at 17\r\n",
      "2022-11-21 12:04:10 partition.py:l503 INFO\tPartitioning...\r\n",
      "2022-11-21 12:05:05 partition.py:l578 INFO\tPartitioned 285 genomes in 54.53 seconds.\r\n",
      "2022-11-21 12:05:05 writeBinaries.py:l966 INFO\tWriting the edges...\r\n",
      "100%|████████████████████████████████| 30168/30168 [00:00<00:00, 30586.33edge/s]\r\n",
      "2022-11-21 12:05:06 writeBinaries.py:l808 INFO\tUpdating gene families with partition information\r\n",
      "100%|█████████████████████████| 14427/14427 [00:00<00:00, 60984.59gene family/s]\r\n",
      "2022-11-21 12:05:06 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Acinetobacter.baumannii/pangenome.h5\r\n",
      "2022-11-21 12:05:06 genomicIsland.py:l250 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 12:05:07 genomicIsland.py:l252 INFO\tCompute Regions of Genomic Plasticity ...\r\n",
      "100%|████████████████████████████████████| 285/285 [00:03<00:00, 77.74genomes/s]\r\n",
      "2022-11-21 12:05:11 genomicIsland.py:l257 INFO\tPredicted 9365 RGP\r\n",
      "2022-11-21 12:05:11 spot.py:l191 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 12:05:11 spot.py:l194 INFO\tDetecting hotspots in the pan...\r\n",
      "2022-11-21 12:05:11 spot.py:l118 INFO\t432 RGPs were not used as they are on a contig border (or have less than 3 persistent gene families until the contig border)\r\n",
      "2022-11-21 12:05:11 spot.py:l120 INFO\t8933 RGPs are being used to predict spots of insertion\r\n",
      "2022-11-21 12:05:11 spot.py:l122 INFO\t648 number of different pairs of flanking gene families\r\n",
      "2022-11-21 12:05:12 spot.py:l203 INFO\t364 spots were detected\r\n",
      "2022-11-21 12:05:12 module.py:l121 INFO\tBuilding the graph...\r\n",
      "100%|█████████████████████████████████████| 285/285 [00:14<00:00, 19.37genome/s]\r\n",
      "2022-11-21 12:05:27 module.py:l123 INFO\tTook 14.71 seconds to build the graph to find modules in\r\n",
      "2022-11-21 12:05:27 module.py:l124 INFO\tThere are 14427 nodes and 166919 edges\r\n",
      "2022-11-21 12:05:28 module.py:l137 INFO\tThere are 3789 families among 611 modules\r\n",
      "2022-11-21 12:05:28 module.py:l138 INFO\tComputing modules took 1.01 seconds\r\n",
      "2022-11-21 12:05:28 writeBinaries.py:l975 INFO\tWriting Regions of Genomic Plasticity...\r\n",
      "100%|████████████████████████████████| 9365/9365 [00:00<00:00, 74639.26region/s]\r\n",
      "2022-11-21 12:05:28 writeBinaries.py:l980 INFO\tWriting Spots of Insertion...\r\n",
      "100%|███████████████████████████████████| 364/364 [00:00<00:00, 118204.29spot/s]\r\n",
      "2022-11-21 12:05:28 writeBinaries.py:l985 INFO\tWriting Modules...\r\n",
      "100%|████████████████████████████████| 611/611 [00:00<00:00, 285126.81modules/s]\r\n",
      "2022-11-21 12:05:28 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Acinetobacter.baumannii/pangenome.h5\r\n",
      "2022-11-21 12:05:31 draw_spot.py:l650 INFO\tDrawing 146 spots\r\n",
      "2022-11-21 12:05:31 draw_spot.py:l559 INFO\tOrdering genes among regions, and drawing spots...\r\n",
      "  0%|                                                 | 0/146 [00:00<?, ?spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_122.html' already exists, will be overwritten.\r\n",
      "  1%|▎                                        | 1/146 [00:00<00:20,  7.05spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_123.html' already exists, will be overwritten.\r\n",
      "  1%|▌                                        | 2/146 [00:00<00:17,  8.27spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_124.html' already exists, will be overwritten.\r\n",
      "  2%|▊                                        | 3/146 [00:00<00:18,  7.54spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_125.html' already exists, will be overwritten.\r\n",
      "  3%|█                                        | 4/146 [00:00<00:17,  8.12spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_126.html' already exists, will be overwritten.\r\n",
      "  3%|█▍                                       | 5/146 [00:00<00:16,  8.31spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_127.html' already exists, will be overwritten.\r\n",
      "  4%|█▋                                       | 6/146 [00:00<00:19,  7.29spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_128.html' already exists, will be overwritten.\r\n",
      "  8%|███                                     | 11/146 [00:01<00:15,  8.93spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_143.html' already exists, will be overwritten.\r\n",
      "  8%|███▎                                    | 12/146 [00:01<00:15,  8.48spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_145.html' already exists, will be overwritten.\r\n",
      "  9%|███▌                                    | 13/146 [00:01<00:15,  8.76spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_142.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_146.html' already exists, will be overwritten.\r\n",
      " 14%|█████▍                                  | 20/146 [00:02<00:12,  9.73spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_356.html' already exists, will be overwritten.\r\n",
      " 14%|█████▊                                  | 21/146 [00:02<00:12,  9.63spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_236.html' already exists, will be overwritten.\r\n",
      " 16%|██████▎                                 | 23/146 [00:02<00:12,  9.73spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_239.html' already exists, will be overwritten.\r\n",
      " 18%|███████▍                                | 27/146 [00:03<00:12,  9.41spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_29.html' already exists, will be overwritten.\r\n",
      " 19%|███████▋                                | 28/146 [00:03<00:31,  3.74spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_31.html' already exists, will be overwritten.\r\n",
      " 20%|███████▉                                | 29/146 [00:03<00:25,  4.50spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_28.html' already exists, will be overwritten.\r\n",
      " 21%|████████▏                               | 30/146 [00:03<00:23,  4.95spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_30.html' already exists, will be overwritten.\r\n",
      " 21%|████████▍                               | 31/146 [00:04<00:24,  4.73spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_27.html' already exists, will be overwritten.\r\n",
      " 22%|████████▊                               | 32/146 [00:04<00:22,  5.17spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_26.html' already exists, will be overwritten.\r\n",
      " 23%|█████████                               | 33/146 [00:04<00:18,  6.03spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_25.html' already exists, will be overwritten.\r\n",
      " 23%|█████████▎                              | 34/146 [00:04<00:31,  3.61spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_24.html' already exists, will be overwritten.\r\n",
      " 24%|█████████▌                              | 35/146 [00:05<00:27,  3.99spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_23.html' already exists, will be overwritten.\r\n",
      " 25%|█████████▊                              | 36/146 [00:05<00:22,  4.83spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_6.html' already exists, will be overwritten.\r\n",
      " 25%|██████████▏                             | 37/146 [00:05<00:20,  5.32spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_22.html' already exists, will be overwritten.\r\n",
      " 26%|██████████▍                             | 38/146 [00:05<00:21,  5.07spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_21.html' already exists, will be overwritten.\r\n",
      " 27%|██████████▉                             | 40/146 [00:05<00:18,  5.84spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_19.html' already exists, will be overwritten.\r\n",
      " 28%|███████████▏                            | 41/146 [00:06<00:27,  3.86spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_18.html' already exists, will be overwritten.\r\n",
      " 29%|███████████▌                            | 42/146 [00:06<00:23,  4.49spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_16.html' already exists, will be overwritten.\r\n",
      " 29%|███████████▊                            | 43/146 [00:06<00:19,  5.19spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_15.html' already exists, will be overwritten.\r\n",
      " 30%|████████████                            | 44/146 [00:06<00:17,  5.87spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_14.html' already exists, will be overwritten.\r\n",
      " 31%|████████████▎                           | 45/146 [00:06<00:16,  6.19spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_13.html' already exists, will be overwritten.\r\n",
      " 32%|████████████▌                           | 46/146 [00:07<00:18,  5.52spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_12.html' already exists, will be overwritten.\r\n",
      " 32%|████████████▉                           | 47/146 [00:07<00:17,  5.66spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_170.html' already exists, will be overwritten.\r\n",
      " 33%|█████████████▏                          | 48/146 [00:07<00:15,  6.32spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_11.html' already exists, will be overwritten.\r\n",
      " 34%|█████████████▍                          | 49/146 [00:07<00:16,  5.76spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_172.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_10.html' already exists, will be overwritten.\r\n",
      " 37%|██████████████▊                         | 54/146 [00:08<00:12,  7.58spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_8.html' already exists, will be overwritten.\r\n",
      " 38%|███████████████                         | 55/146 [00:08<00:21,  4.25spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_178.html' already exists, will be overwritten.\r\n",
      " 38%|███████████████▎                        | 56/146 [00:08<00:17,  5.06spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_180.html' already exists, will be overwritten.\r\n",
      " 40%|███████████████▉                        | 58/146 [00:09<00:13,  6.50spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_32.html' already exists, will be overwritten.\r\n",
      " 40%|████████████████▏                       | 59/146 [00:09<00:17,  5.04spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_33.html' already exists, will be overwritten.\r\n",
      " 41%|████████████████▍                       | 60/146 [00:09<00:16,  5.18spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_179.html' already exists, will be overwritten.\r\n",
      " 42%|████████████████▋                       | 61/146 [00:09<00:14,  5.94spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_34.html' already exists, will be overwritten.\r\n",
      " 42%|████████████████▉                       | 62/146 [00:09<00:14,  5.82spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_181.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_35.html' already exists, will be overwritten.\r\n",
      " 44%|█████████████████▌                      | 64/146 [00:10<00:12,  6.35spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_36.html' already exists, will be overwritten.\r\n",
      " 45%|█████████████████▊                      | 65/146 [00:10<00:12,  6.48spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_183.html' already exists, will be overwritten.\r\n",
      " 45%|██████████████████                      | 66/146 [00:10<00:11,  7.02spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_37.html' already exists, will be overwritten.\r\n",
      " 46%|██████████████████▎                     | 67/146 [00:10<00:10,  7.49spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_38.html' already exists, will be overwritten.\r\n",
      " 47%|██████████████████▋                     | 68/146 [00:10<00:09,  7.93spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_39.html' already exists, will be overwritten.\r\n",
      " 48%|███████████████████▏                    | 70/146 [00:11<00:13,  5.52spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_40.html' already exists, will be overwritten.\r\n",
      " 49%|███████████████████▍                    | 71/146 [00:11<00:12,  5.99spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_190.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_41.html' already exists, will be overwritten.\r\n",
      " 50%|████████████████████                    | 73/146 [00:11<00:14,  5.12spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_4.html' already exists, will be overwritten.\r\n",
      " 51%|████████████████████▎                   | 74/146 [00:11<00:13,  5.35spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_42.html' already exists, will be overwritten.\r\n",
      " 51%|████████████████████▌                   | 75/146 [00:12<00:17,  4.10spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_43.html' already exists, will be overwritten.\r\n",
      " 52%|████████████████████▊                   | 76/146 [00:12<00:19,  3.62spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_192.html' already exists, will be overwritten.\r\n",
      " 53%|█████████████████████                   | 77/146 [00:12<00:15,  4.34spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_44.html' already exists, will be overwritten.\r\n",
      " 53%|█████████████████████▎                  | 78/146 [00:12<00:13,  5.10spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_191.html' already exists, will be overwritten.\r\n",
      " 54%|█████████████████████▋                  | 79/146 [00:12<00:11,  5.80spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_194.html' already exists, will be overwritten.\r\n",
      " 55%|█████████████████████▉                  | 80/146 [00:13<00:10,  6.45spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_195.html' already exists, will be overwritten.\r\n",
      " 55%|██████████████████████▏                 | 81/146 [00:13<00:09,  7.16spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_47.html' already exists, will be overwritten.\r\n",
      " 57%|██████████████████████▋                 | 83/146 [00:13<00:08,  7.75spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_197.html' already exists, will be overwritten.\r\n",
      " 58%|███████████████████████                 | 84/146 [00:13<00:08,  7.19spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_50.html' already exists, will be overwritten.\r\n",
      " 58%|███████████████████████▎                | 85/146 [00:13<00:07,  7.72spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_51.html' already exists, will be overwritten.\r\n",
      " 59%|███████████████████████▌                | 86/146 [00:13<00:07,  7.95spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_200.html' already exists, will be overwritten.\r\n",
      " 60%|███████████████████████▊                | 87/146 [00:13<00:07,  8.14spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_54.html' already exists, will be overwritten.\r\n",
      " 60%|████████████████████████                | 88/146 [00:14<00:07,  7.94spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_55.html' already exists, will be overwritten.\r\n",
      " 61%|████████████████████████▍               | 89/146 [00:14<00:06,  8.40spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_56.html' already exists, will be overwritten.\r\n",
      " 62%|████████████████████████▉               | 91/146 [00:14<00:05,  9.21spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_60.html' already exists, will be overwritten.\r\n",
      " 63%|█████████████████████████▏              | 92/146 [00:14<00:05,  9.29spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_208.html' already exists, will be overwritten.\r\n",
      " 64%|█████████████████████████▍              | 93/146 [00:14<00:05,  9.40spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_61.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_270.html' already exists, will be overwritten.\r\n",
      " 66%|██████████████████████████▎             | 96/146 [00:14<00:05,  9.72spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_211.html' already exists, will be overwritten.\r\n",
      " 67%|██████████████████████████▊             | 98/146 [00:15<00:04,  9.79spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_2.html' already exists, will be overwritten.\r\n",
      " 68%|██████████████████████████▋            | 100/146 [00:15<00:05,  8.70spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_1.html' already exists, will be overwritten.\r\n",
      " 71%|███████████████████████████▌           | 103/146 [00:15<00:06,  6.52spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_215.html' already exists, will be overwritten.\r\n",
      " 71%|███████████████████████████▊           | 104/146 [00:16<00:05,  7.14spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_216.html' already exists, will be overwritten.\r\n",
      " 72%|████████████████████████████           | 105/146 [00:16<00:05,  7.24spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_217.html' already exists, will be overwritten.\r\n",
      " 73%|████████████████████████████▌          | 107/146 [00:16<00:04,  7.86spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_218.html' already exists, will be overwritten.\r\n",
      " 75%|█████████████████████████████          | 109/146 [00:16<00:04,  8.59spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_260.html' already exists, will be overwritten.\r\n",
      " 75%|█████████████████████████████▍         | 110/146 [00:16<00:04,  8.82spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_221.html' already exists, will be overwritten.\r\n",
      " 76%|█████████████████████████████▋         | 111/146 [00:16<00:03,  9.05spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_0.html' already exists, will be overwritten.\r\n",
      " 77%|██████████████████████████████▏        | 113/146 [00:17<00:05,  5.77spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_3.html' already exists, will be overwritten.\r\n",
      " 79%|██████████████████████████████▋        | 115/146 [00:17<00:05,  5.82spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_268.html' already exists, will be overwritten.\r\n",
      " 81%|███████████████████████████████▌       | 118/146 [00:18<00:03,  7.12spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_271.html' already exists, will be overwritten.\r\n",
      " 82%|███████████████████████████████▊       | 119/146 [00:18<00:03,  7.53spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_272.html' already exists, will be overwritten.\r\n",
      " 82%|████████████████████████████████       | 120/146 [00:18<00:03,  7.95spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_273.html' already exists, will be overwritten.\r\n",
      " 83%|████████████████████████████████▎      | 121/146 [00:18<00:03,  8.26spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_121.html' already exists, will be overwritten.\r\n",
      " 86%|█████████████████████████████████▋     | 126/146 [00:18<00:02,  9.13spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_292.html' already exists, will be overwritten.\r\n",
      " 88%|██████████████████████████████████▏    | 128/146 [00:19<00:01,  9.41spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_303.html' already exists, will be overwritten.\r\n",
      " 89%|██████████████████████████████████▋    | 130/146 [00:19<00:01,  9.80spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_119.html' already exists, will be overwritten.\r\n",
      " 90%|██████████████████████████████████▉    | 131/146 [00:19<00:01,  8.00spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_120.html' already exists, will be overwritten.\r\n",
      " 95%|████████████████████████████████████▊  | 138/146 [00:20<00:01,  7.31spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_103.html' already exists, will be overwritten.\r\n",
      " 97%|█████████████████████████████████████▉ | 142/146 [00:21<00:00,  6.00spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_87.html' already exists, will be overwritten.\r\n",
      " 99%|██████████████████████████████████████▍| 144/146 [00:21<00:00,  7.40spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Acinetobacter.baumannii/spot_figures/spot_5.html' already exists, will be overwritten.\r\n",
      "100%|███████████████████████████████████████| 146/146 [00:21<00:00,  6.75spot/s]\r\n",
      "2022-11-21 12:05:54 draw_spot.py:l617 INFO\tDone drawing spot(s), they can be found in the directory: 'data/pangenomes/Acinetobacter.baumannii/spot_figures'\r\n",
      "2022-11-21 12:05:54 tile_plot.py:l38 INFO\tDrawing the tile plot...\r\n",
      "2022-11-21 12:05:54 tile_plot.py:l56 INFO\tstart with matrice\r\n",
      "2022-11-21 12:05:55 tile_plot.py:l72 INFO\tdone with making the dendrogram to order the organisms on the plot\r\n",
      "2022-11-21 12:05:55 tile_plot.py:l106 INFO\tGetting the gene name(s) and the number for each tile of the plot ...\r\n",
      "2022-11-21 12:05:57 tile_plot.py:l116 INFO\tDone extracting names and numbers. Making the heatmap ...\r\n",
      "2022-11-21 12:06:05 tile_plot.py:l173 INFO\tDrawing the figure itself...\r\n",
      "2022-11-21 12:06:45 tile_plot.py:l175 INFO\tDone with the tile plot : 'data/pangenomes/Acinetobacter.baumannii/tile_plot.html' \r\n",
      "2022-11-21 12:06:45 ucurve.py:l23 INFO\tDrawing the U-shaped curve...\r\n",
      "2022-11-21 12:06:45 ucurve.py:l77 INFO\tDone drawing the U-shaped curve : 'data/pangenomes/Acinetobacter.baumannii/Ushaped_plot.html'\r\n",
      "2022-11-21 12:06:46 writeFlat.py:l328 INFO\tWriting the .csv file ...\r\n",
      "2022-11-21 12:06:46 writeFlat.py:l399 INFO\tWriting the gene presence absence file ...\r\n",
      "2022-11-21 12:06:46 writeFlat.py:l305 INFO\tWriting the light gexf file for the pangenome graph...\r\n",
      "2022-11-21 12:06:46 writeFlat.py:l305 INFO\tWriting the gexf file for the pangenome graph...\r\n",
      "2022-11-21 12:06:46 writeFlat.py:l582 INFO\tWriting the projection files...\r\n",
      "2022-11-21 12:06:46 writeFlat.py:l431 INFO\tWriting pangenome statistics...\r\n",
      "2022-11-21 12:06:46 writeFlat.py:l432 INFO\tWriting statistics on persistent duplication...\r\n",
      "2022-11-21 12:06:46 writeFlat.py:l454 INFO\tDone writing stats on persistent duplication\r\n",
      "2022-11-21 12:06:46 writeFlat.py:l455 INFO\tWriting genome per genome statistics (completeness and counts)...\r\n",
      "2022-11-21 12:06:46 writeFlat.py:l419 INFO\tDone writing the gene presence absence file : 'data/pangenomes/Acinetobacter.baumannii/gene_presence_absence.Rtab'\r\n",
      "2022-11-21 12:06:46 writeFlat.py:l157 INFO\tWriting the json file for the pangenome graph...\r\n",
      "2022-11-21 12:06:49 writeFlat.py:l520 INFO\tDone writing genome per genome statistics\r\n",
      "2022-11-21 12:06:49 writeFlat.py:l598 INFO\tWriting the list of gene families for each partition ...\r\n",
      "2022-11-21 12:06:49 writeFlat.py:l625 INFO\tDone writing the list of gene families for each partition\r\n",
      "2022-11-21 12:06:50 writeFlat.py:l389 INFO\tDone writing the matrix : 'data/pangenomes/Acinetobacter.baumannii/matrix.csv'\r\n",
      "2022-11-21 12:06:50 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Acinetobacter.baumannii/pangenomeGraph_light.gexf'\r\n",
      "2022-11-21 12:06:50 writeFlat.py:l776 INFO\tWriting functional modules...\r\n",
      "2022-11-21 12:06:50 writeFlat.py:l784 INFO\tDone writing functional modules to: 'data/pangenomes/Acinetobacter.baumannii/functional_modules.tsv'\r\n",
      "2022-11-21 12:06:50 writeFlat.py:l752 INFO\tWriting functional modules summary...\r\n",
      "2022-11-21 12:06:50 writeFlat.py:l767 INFO\tDone writing module summary: 'data/pangenomes/Acinetobacter.baumannii/modules_summary.tsv'\r\n",
      "2022-11-21 12:06:50 writeFlat.py:l793 INFO\tWriting modules to organisms associations...\r\n",
      "2022-11-21 12:06:51 writeFlat.py:l814 INFO\tWriting modules to spot associations...\r\n",
      "2022-11-21 12:06:53 writeFlat.py:l691 INFO\tDone writing spots in : 'data/pangenomes/Acinetobacter.baumannii/summarize_spots.tsv'\r\n",
      "2022-11-21 12:06:53 writeFlat.py:l846 INFO\tClustering RGPs based on module content...\r\n",
      "2022-11-21 12:06:54 writeFlat.py:l881 INFO\tRGP and associated modules are listed in : data/pangenomes/Acinetobacter.baumannii/modules_RGP_lists.tsv\r\n",
      "2022-11-21 12:06:55 writeFlat.py:l837 INFO\tDone writing module to spot associations to: data/pangenomes/Acinetobacter.baumannii/modules_spots.tsv\r\n",
      "2022-11-21 12:06:56 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Acinetobacter.baumannii/pangenomeGraph.gexf'\r\n",
      "2022-11-21 12:06:56 writeFlat.py:l164 INFO\tDone writing the json file : 'data/pangenomes/Acinetobacter.baumannii/pangenomeGraph.json'\r\n",
      "2022-11-21 12:06:59 writeFlat.py:l588 INFO\tDone writing the projection files\r\n",
      "2022-11-21 12:07:07 writeFlat.py:l804 INFO\tDone writing modules to organisms associations to: 'data/pangenomes/Acinetobacter.baumannii/modules_in_organisms.tsv'\r\n",
      "2022-11-21 12:07:07 all.py:l125 INFO\tAnnotation took : 36.86 seconds\r\n",
      "2022-11-21 12:07:07 all.py:l126 INFO\tClustering took : 86.11 seconds\r\n",
      "2022-11-21 12:07:07 all.py:l127 INFO\tBuilding the graph took : 4.99 seconds\r\n",
      "2022-11-21 12:07:07 all.py:l128 INFO\tPartitioning the pangenome took : 154.72 seconds\r\n",
      "2022-11-21 12:07:07 all.py:l129 INFO\tPredicting RGP took : 4.25 seconds\r\n",
      "2022-11-21 12:07:07 all.py:l130 INFO\tGathering RGP into spots took : 27.01 seconds\r\n",
      "2022-11-21 12:07:07 all.py:l131 INFO\tPredicting modules took : 15.73 seconds\r\n",
      "2022-11-21 12:07:07 all.py:l132 INFO\tWriting the pangenome data in HDF5 took : 12.18 seconds\r\n",
      "2022-11-21 12:07:07 all.py:l134 INFO\tWriting descriptive files for the pangenome took : 22.0 seconds\r\n",
      "Genes: 1045048\r\n",
      "Organisms: 285\r\n",
      "Families: 14427\r\n",
      "Edges: 30168\r\n",
      "Persistent (min:0.91, max:1.0, sd:0.01, mean:0.98 ): 2686\r\n",
      "Shell ( min:0.02, max:0.96, sd:0.26, mean:0.21 ): 4140\r\n",
      "Cloud ( min:0.0, max:0.02, sd:0.0, mean:0.01 ): 7601\r\n",
      "Number of partitions: 17\r\n",
      "Shell S14 : 955\r\n",
      "Shell S15 : 1953\r\n",
      "Shell S9 : 83\r\n",
      "Shell S1 : 254\r\n",
      "Shell S3 : 25\r\n",
      "Shell S8 : 137\r\n",
      "Shell S2 : 83\r\n",
      "Shell S13 : 155\r\n",
      "Shell S7 : 87\r\n",
      "Shell S6 : 63\r\n",
      "Shell S4 : 77\r\n",
      "Shell S11 : 62\r\n",
      "Shell S5 : 45\r\n",
      "Shell S12 : 81\r\n",
      "Shell S10 : 77\r\n",
      "Shell S_ : 3\r\n",
      "RGPs: 9365\r\n",
      "Spots: 364\r\n",
      "Modules: 611\r\n",
      "Families in Modules: 3789\r\n",
      "2022-11-21 12:07:18 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin all --anno data/pangenomes/Enterobacter.bugandensis.list -o data/pangenomes/Enterobacter.bugandensis -c 6 -f\r\n",
      "2022-11-21 12:07:18 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 12:07:18 annotate.py:l448 INFO\tReading data/pangenomes/Enterobacter.bugandensis.list the list of organism files ...\r\n",
      "100%|███████████████████████████████████████| 115/115 [00:14<00:00,  7.75file/s]\r\n",
      "2022-11-21 12:07:34 annotate.py:l469 INFO\tgene identifiers used in the provided annotation files were unique, PPanGGOLiN will use them.\r\n",
      "2022-11-21 12:07:34 writeBinaries.py:l932 INFO\tWriting genome annotations...\r\n",
      "100%|█████████████████████████████████████| 115/115 [00:01<00:00, 79.38genome/s]\r\n",
      "2022-11-21 12:07:36 writeBinaries.py:l950 INFO\twriting the protein coding gene dna sequences\r\n",
      "100%|█████████████████████████████| 492746/492746 [00:00<00:00, 635826.64gene/s]\r\n",
      "2022-11-21 12:07:40 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.bugandensis/pangenome.h5\r\n",
      "2022-11-21 12:07:40 writeSequences.py:l39 INFO\tWriting all of the CDS sequences...\r\n",
      "100%|█████████████████████████████| 492746/492746 [00:00<00:00, 615469.21gene/s]\r\n",
      "2022-11-21 12:07:41 cluster.py:l297 INFO\tClustering all of the genes sequences...\r\n",
      "2022-11-21 12:07:41 cluster.py:l86 INFO\tCreating sequence database...\r\n",
      "2022-11-21 12:07:44 cluster.py:l93 INFO\tClustering sequences...\r\n",
      "2022-11-21 12:07:52 cluster.py:l99 INFO\tExtracting cluster representatives...\r\n",
      "2022-11-21 12:07:53 cluster.py:l108 INFO\tWriting gene to family informations\r\n",
      "2022-11-21 12:07:53 cluster.py:l305 INFO\tAssociating fragments to their original gene family...\r\n",
      "2022-11-21 12:07:53 cluster.py:l153 INFO\tAligning cluster representatives...\r\n",
      "2022-11-21 12:08:06 cluster.py:l159 INFO\tExtracting alignments...\r\n",
      "2022-11-21 12:08:07 cluster.py:l197 INFO\tStarting with 11804 families\r\n",
      "2022-11-21 12:08:07 cluster.py:l231 INFO\tEnding with 11542 gene families\r\n",
      "2022-11-21 12:08:07 cluster.py:l242 INFO\tAdding protein sequences to the gene families\r\n",
      "2022-11-21 12:08:07 cluster.py:l256 INFO\tAdding 492746 genes to the gene families\r\n",
      "100%|█████████████████████████████| 492746/492746 [00:01<00:00, 271798.58gene/s]\r\n",
      "2022-11-21 12:08:09 writeBinaries.py:l955 INFO\tWriting gene families and gene associations...\r\n",
      "100%|█████████████████████████| 11542/11542 [00:00<00:00, 27183.02gene family/s]\r\n",
      "2022-11-21 12:08:09 writeBinaries.py:l957 INFO\tWriting gene families information...\r\n",
      "100%|████████████████████████| 11542/11542 [00:00<00:00, 135303.15gene family/s]\r\n",
      "2022-11-21 12:08:09 writeBinaries.py:l823 INFO\tUpdating annotations with fragment information\r\n",
      "100%|█████████████████████████████| 500674/500674 [00:01<00:00, 296029.02gene/s]\r\n",
      "2022-11-21 12:08:11 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.bugandensis/pangenome.h5\r\n",
      "2022-11-21 12:08:11 makeGraph.py:l87 INFO\tComputing the neighbors graph...\r\n",
      "Processing GCF_903970835.1_ASM90397083v1_genomic: 100%|█| 115/115 [00:02<00:00, \r\n",
      "2022-11-21 12:08:13 makeGraph.py:l108 INFO\tDone making the neighbors graph.\r\n",
      "2022-11-21 12:08:13 partition.py:l482 INFO\tEstimating the optimal number of partitions...\r\n",
      "100%|█████████████████████████| 19/19 [00:26<00:00,  1.37s/Number of partitions]\r\n",
      "2022-11-21 12:08:40 partition.py:l485 INFO\tThe number of partitions has been evaluated at 13\r\n",
      "2022-11-21 12:08:40 partition.py:l503 INFO\tPartitioning...\r\n",
      "2022-11-21 12:09:01 partition.py:l578 INFO\tPartitioned 115 genomes in 21.24 seconds.\r\n",
      "2022-11-21 12:09:01 writeBinaries.py:l966 INFO\tWriting the edges...\r\n",
      "100%|████████████████████████████████| 14347/14347 [00:00<00:00, 36920.43edge/s]\r\n",
      "2022-11-21 12:09:02 writeBinaries.py:l808 INFO\tUpdating gene families with partition information\r\n",
      "100%|█████████████████████████| 11542/11542 [00:00<00:00, 72268.30gene family/s]\r\n",
      "2022-11-21 12:09:02 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.bugandensis/pangenome.h5\r\n",
      "2022-11-21 12:09:02 genomicIsland.py:l250 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 12:09:02 genomicIsland.py:l252 INFO\tCompute Regions of Genomic Plasticity ...\r\n",
      "100%|███████████████████████████████████| 115/115 [00:00<00:00, 118.92genomes/s]\r\n",
      "2022-11-21 12:09:03 genomicIsland.py:l257 INFO\tPredicted 3830 RGP\r\n",
      "2022-11-21 12:09:03 spot.py:l191 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 12:09:04 spot.py:l194 INFO\tDetecting hotspots in the pan...\r\n",
      "2022-11-21 12:09:04 spot.py:l118 INFO\t782 RGPs were not used as they are on a contig border (or have less than 3 persistent gene families until the contig border)\r\n",
      "2022-11-21 12:09:04 spot.py:l120 INFO\t3048 RGPs are being used to predict spots of insertion\r\n",
      "2022-11-21 12:09:04 spot.py:l122 INFO\t208 number of different pairs of flanking gene families\r\n",
      "2022-11-21 12:09:04 spot.py:l203 INFO\t130 spots were detected\r\n",
      "2022-11-21 12:09:04 module.py:l121 INFO\tBuilding the graph...\r\n",
      "100%|█████████████████████████████████████| 115/115 [00:06<00:00, 17.01genome/s]\r\n",
      "2022-11-21 12:09:10 module.py:l123 INFO\tTook 6.76 seconds to build the graph to find modules in\r\n",
      "2022-11-21 12:09:10 module.py:l124 INFO\tThere are 11542 nodes and 76943 edges\r\n",
      "2022-11-21 12:09:11 module.py:l137 INFO\tThere are 2635 families among 360 modules\r\n",
      "2022-11-21 12:09:11 module.py:l138 INFO\tComputing modules took 0.44 seconds\r\n",
      "2022-11-21 12:09:11 writeBinaries.py:l975 INFO\tWriting Regions of Genomic Plasticity...\r\n",
      "100%|███████████████████████████████| 3830/3830 [00:00<00:00, 142541.88region/s]\r\n",
      "2022-11-21 12:09:11 writeBinaries.py:l980 INFO\tWriting Spots of Insertion...\r\n",
      "100%|███████████████████████████████████| 130/130 [00:00<00:00, 130382.48spot/s]\r\n",
      "2022-11-21 12:09:11 writeBinaries.py:l985 INFO\tWriting Modules...\r\n",
      "100%|████████████████████████████████| 360/360 [00:00<00:00, 223895.23modules/s]\r\n",
      "2022-11-21 12:09:11 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.bugandensis/pangenome.h5\r\n",
      "2022-11-21 12:09:11 draw_spot.py:l650 INFO\tDrawing 83 spots\r\n",
      "2022-11-21 12:09:11 draw_spot.py:l559 INFO\tOrdering genes among regions, and drawing spots...\r\n",
      "  1%|▌                                         | 1/83 [00:00<00:10,  7.89spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_114.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_116.html' already exists, will be overwritten.\r\n",
      "  4%|█▌                                        | 3/83 [00:00<00:08,  9.62spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_113.html' already exists, will be overwritten.\r\n",
      "  5%|██                                        | 4/83 [00:00<00:08,  9.33spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_102.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_99.html' already exists, will be overwritten.\r\n",
      "  7%|███                                       | 6/83 [00:00<00:07,  9.66spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_104.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_106.html' already exists, will be overwritten.\r\n",
      " 10%|████                                      | 8/83 [00:00<00:07,  9.76spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_108.html' already exists, will be overwritten.\r\n",
      " 12%|████▉                                    | 10/83 [00:01<00:07,  9.86spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_51.html' already exists, will be overwritten.\r\n",
      " 13%|█████▍                                   | 11/83 [00:01<00:07,  9.80spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_105.html' already exists, will be overwritten.\r\n",
      " 14%|█████▉                                   | 12/83 [00:01<00:07,  9.69spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_110.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_111.html' already exists, will be overwritten.\r\n",
      " 18%|███████▍                                 | 15/83 [00:01<00:06,  9.87spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_121.html' already exists, will be overwritten.\r\n",
      " 19%|███████▉                                 | 16/83 [00:01<00:06,  9.83spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_27.html' already exists, will be overwritten.\r\n",
      " 20%|████████▍                                | 17/83 [00:01<00:07,  9.19spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_39.html' already exists, will be overwritten.\r\n",
      " 22%|████████▉                                | 18/83 [00:01<00:07,  9.02spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_65.html' already exists, will be overwritten.\r\n",
      " 23%|█████████▍                               | 19/83 [00:01<00:06,  9.27spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_38.html' already exists, will be overwritten.\r\n",
      " 27%|██████████▊                              | 22/83 [00:02<00:06,  8.71spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_66.html' already exists, will be overwritten.\r\n",
      " 29%|███████████▊                             | 24/83 [00:02<00:06,  9.02spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_29.html' already exists, will be overwritten.\r\n",
      " 30%|████████████▎                            | 25/83 [00:02<00:06,  8.84spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_23.html' already exists, will be overwritten.\r\n",
      " 31%|████████████▊                            | 26/83 [00:02<00:06,  9.12spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_10.html' already exists, will be overwritten.\r\n",
      " 33%|█████████████▎                           | 27/83 [00:02<00:06,  9.14spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_58.html' already exists, will be overwritten.\r\n",
      " 34%|█████████████▊                           | 28/83 [00:02<00:05,  9.37spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_59.html' already exists, will be overwritten.\r\n",
      " 35%|██████████████▎                          | 29/83 [00:03<00:05,  9.16spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_34.html' already exists, will be overwritten.\r\n",
      " 36%|██████████████▊                          | 30/83 [00:03<00:06,  8.62spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_57.html' already exists, will be overwritten.\r\n",
      " 37%|███████████████▎                         | 31/83 [00:03<00:05,  8.93spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_67.html' already exists, will be overwritten.\r\n",
      " 39%|███████████████▊                         | 32/83 [00:03<00:05,  9.11spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_63.html' already exists, will be overwritten.\r\n",
      " 40%|████████████████▎                        | 33/83 [00:03<00:05,  9.21spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_0.html' already exists, will be overwritten.\r\n",
      " 41%|████████████████▊                        | 34/83 [00:03<00:06,  7.82spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_33.html' already exists, will be overwritten.\r\n",
      " 42%|█████████████████▎                       | 35/83 [00:03<00:05,  8.01spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_16.html' already exists, will be overwritten.\r\n",
      " 45%|██████████████████▎                      | 37/83 [00:04<00:05,  7.90spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_37.html' already exists, will be overwritten.\r\n",
      " 46%|██████████████████▊                      | 38/83 [00:04<00:05,  8.31spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_82.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_48.html' already exists, will be overwritten.\r\n",
      " 48%|███████████████████▊                     | 40/83 [00:04<00:04,  8.87spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_22.html' already exists, will be overwritten.\r\n",
      " 51%|████████████████████▋                    | 42/83 [00:04<00:04,  8.68spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_90.html' already exists, will be overwritten.\r\n",
      " 54%|██████████████████████▏                  | 45/83 [00:04<00:04,  8.74spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_4.html' already exists, will be overwritten.\r\n",
      " 57%|███████████████████████▏                 | 47/83 [00:05<00:04,  8.92spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_25.html' already exists, will be overwritten.\r\n",
      " 58%|███████████████████████▋                 | 48/83 [00:05<00:03,  8.83spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_50.html' already exists, will be overwritten.\r\n",
      " 59%|████████████████████████▏                | 49/83 [00:05<00:03,  9.10spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_40.html' already exists, will be overwritten.\r\n",
      " 60%|████████████████████████▋                | 50/83 [00:05<00:03,  9.34spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_35.html' already exists, will be overwritten.\r\n",
      " 61%|█████████████████████████▏               | 51/83 [00:05<00:03,  9.16spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_55.html' already exists, will be overwritten.\r\n",
      " 64%|██████████████████████████▏              | 53/83 [00:05<00:03,  9.27spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_20.html' already exists, will be overwritten.\r\n",
      " 66%|███████████████████████████▏             | 55/83 [00:06<00:02,  9.36spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_54.html' already exists, will be overwritten.\r\n",
      " 67%|███████████████████████████▋             | 56/83 [00:06<00:02,  9.29spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_44.html' already exists, will be overwritten.\r\n",
      " 70%|████████████████████████████▋            | 58/83 [00:06<00:02,  8.63spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_14.html' already exists, will be overwritten.\r\n",
      " 72%|█████████████████████████████▋           | 60/83 [00:06<00:02,  8.40spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_13.html' already exists, will be overwritten.\r\n",
      " 75%|██████████████████████████████▋          | 62/83 [00:06<00:02,  8.72spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_68.html' already exists, will be overwritten.\r\n",
      " 76%|███████████████████████████████          | 63/83 [00:06<00:02,  8.86spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_2.html' already exists, will be overwritten.\r\n",
      " 77%|███████████████████████████████▌         | 64/83 [00:07<00:02,  8.09spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_26.html' already exists, will be overwritten.\r\n",
      " 78%|████████████████████████████████         | 65/83 [00:07<00:02,  8.14spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_83.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_5.html' already exists, will be overwritten.\r\n",
      " 81%|█████████████████████████████████        | 67/83 [00:07<00:01,  8.76spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_42.html' already exists, will be overwritten.\r\n",
      " 82%|█████████████████████████████████▌       | 68/83 [00:07<00:01,  8.96spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_71.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_21.html' already exists, will be overwritten.\r\n",
      " 84%|██████████████████████████████████▌      | 70/83 [00:07<00:01,  9.22spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_24.html' already exists, will be overwritten.\r\n",
      " 86%|███████████████████████████████████      | 71/83 [00:07<00:01,  9.22spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_3.html' already exists, will be overwritten.\r\n",
      " 88%|████████████████████████████████████     | 73/83 [00:08<00:01,  9.66spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_19.html' already exists, will be overwritten.\r\n",
      " 89%|████████████████████████████████████▌    | 74/83 [00:08<00:00,  9.40spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_30.html' already exists, will be overwritten.\r\n",
      " 90%|█████████████████████████████████████    | 75/83 [00:08<00:00,  9.39spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_79.html' already exists, will be overwritten.\r\n",
      " 92%|█████████████████████████████████████▌   | 76/83 [00:08<00:00,  9.23spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_36.html' already exists, will be overwritten.\r\n",
      " 93%|██████████████████████████████████████   | 77/83 [00:08<00:00,  9.14spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_49.html' already exists, will be overwritten.\r\n",
      " 94%|██████████████████████████████████████▌  | 78/83 [00:08<00:00,  8.97spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_52.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_64.html' already exists, will be overwritten.\r\n",
      " 99%|████████████████████████████████████████▌| 82/83 [00:09<00:00,  9.29spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.bugandensis/spot_figures/spot_56.html' already exists, will be overwritten.\r\n",
      "100%|█████████████████████████████████████████| 83/83 [00:09<00:00,  9.05spot/s]\r\n",
      "2022-11-21 12:09:21 draw_spot.py:l617 INFO\tDone drawing spot(s), they can be found in the directory: 'data/pangenomes/Enterobacter.bugandensis/spot_figures'\r\n",
      "2022-11-21 12:09:21 tile_plot.py:l38 INFO\tDrawing the tile plot...\r\n",
      "2022-11-21 12:09:21 tile_plot.py:l56 INFO\tstart with matrice\r\n",
      "2022-11-21 12:09:21 tile_plot.py:l72 INFO\tdone with making the dendrogram to order the organisms on the plot\r\n",
      "2022-11-21 12:09:21 tile_plot.py:l106 INFO\tGetting the gene name(s) and the number for each tile of the plot ...\r\n",
      "2022-11-21 12:09:22 tile_plot.py:l116 INFO\tDone extracting names and numbers. Making the heatmap ...\r\n",
      "2022-11-21 12:09:25 tile_plot.py:l173 INFO\tDrawing the figure itself...\r\n",
      "2022-11-21 12:09:38 tile_plot.py:l175 INFO\tDone with the tile plot : 'data/pangenomes/Enterobacter.bugandensis/tile_plot.html' \r\n",
      "2022-11-21 12:09:38 ucurve.py:l23 INFO\tDrawing the U-shaped curve...\r\n",
      "2022-11-21 12:09:38 ucurve.py:l77 INFO\tDone drawing the U-shaped curve : 'data/pangenomes/Enterobacter.bugandensis/Ushaped_plot.html'\r\n",
      "2022-11-21 12:09:38 writeFlat.py:l328 INFO\tWriting the .csv file ...\r\n",
      "2022-11-21 12:09:38 writeFlat.py:l399 INFO\tWriting the gene presence absence file ...\r\n",
      "2022-11-21 12:09:38 writeFlat.py:l305 INFO\tWriting the gexf file for the pangenome graph...\r\n",
      "2022-11-21 12:09:38 writeFlat.py:l305 INFO\tWriting the light gexf file for the pangenome graph...\r\n",
      "2022-11-21 12:09:38 writeFlat.py:l582 INFO\tWriting the projection files...\r\n",
      "2022-11-21 12:09:38 writeFlat.py:l431 INFO\tWriting pangenome statistics...\r\n",
      "2022-11-21 12:09:38 writeFlat.py:l432 INFO\tWriting statistics on persistent duplication...\r\n",
      "2022-11-21 12:09:38 writeFlat.py:l419 INFO\tDone writing the gene presence absence file : 'data/pangenomes/Enterobacter.bugandensis/gene_presence_absence.Rtab'\r\n",
      "2022-11-21 12:09:38 writeFlat.py:l157 INFO\tWriting the json file for the pangenome graph...\r\n",
      "2022-11-21 12:09:38 writeFlat.py:l454 INFO\tDone writing stats on persistent duplication\r\n",
      "2022-11-21 12:09:38 writeFlat.py:l455 INFO\tWriting genome per genome statistics (completeness and counts)...\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l520 INFO\tDone writing genome per genome statistics\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l598 INFO\tWriting the list of gene families for each partition ...\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l625 INFO\tDone writing the list of gene families for each partition\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l691 INFO\tDone writing spots in : 'data/pangenomes/Enterobacter.bugandensis/summarize_spots.tsv'\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l389 INFO\tDone writing the matrix : 'data/pangenomes/Enterobacter.bugandensis/matrix.csv'\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l776 INFO\tWriting functional modules...\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l784 INFO\tDone writing functional modules to: 'data/pangenomes/Enterobacter.bugandensis/functional_modules.tsv'\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l752 INFO\tWriting functional modules summary...\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l767 INFO\tDone writing module summary: 'data/pangenomes/Enterobacter.bugandensis/modules_summary.tsv'\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l793 INFO\tWriting modules to organisms associations...\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l814 INFO\tWriting modules to spot associations...\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l837 INFO\tDone writing module to spot associations to: data/pangenomes/Enterobacter.bugandensis/modules_spots.tsv\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l846 INFO\tClustering RGPs based on module content...\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l881 INFO\tRGP and associated modules are listed in : data/pangenomes/Enterobacter.bugandensis/modules_RGP_lists.tsv\r\n",
      "2022-11-21 12:09:40 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Enterobacter.bugandensis/pangenomeGraph_light.gexf'\r\n",
      "2022-11-21 12:09:42 writeFlat.py:l588 INFO\tDone writing the projection files\r\n",
      "2022-11-21 12:09:43 writeFlat.py:l164 INFO\tDone writing the json file : 'data/pangenomes/Enterobacter.bugandensis/pangenomeGraph.json'\r\n",
      "2022-11-21 12:09:43 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Enterobacter.bugandensis/pangenomeGraph.gexf'\r\n",
      "2022-11-21 12:09:46 writeFlat.py:l804 INFO\tDone writing modules to organisms associations to: 'data/pangenomes/Enterobacter.bugandensis/modules_in_organisms.tsv'\r\n",
      "2022-11-21 12:09:46 all.py:l125 INFO\tAnnotation took : 15.81 seconds\r\n",
      "2022-11-21 12:09:46 all.py:l126 INFO\tClustering took : 28.52 seconds\r\n",
      "2022-11-21 12:09:46 all.py:l127 INFO\tBuilding the graph took : 2.33 seconds\r\n",
      "2022-11-21 12:09:46 all.py:l128 INFO\tPartitioning the pangenome took : 47.72 seconds\r\n",
      "2022-11-21 12:09:46 all.py:l129 INFO\tPredicting RGP took : 1.26 seconds\r\n",
      "2022-11-21 12:09:46 all.py:l130 INFO\tGathering RGP into spots took : 9.96 seconds\r\n",
      "2022-11-21 12:09:46 all.py:l131 INFO\tPredicting modules took : 7.2 seconds\r\n",
      "2022-11-21 12:09:46 all.py:l132 INFO\tWriting the pangenome data in HDF5 took : 6.8 seconds\r\n",
      "2022-11-21 12:09:46 all.py:l134 INFO\tWriting descriptive files for the pangenome took : 8.63 seconds\r\n",
      "Genes: 492746\r\n",
      "Organisms: 115\r\n",
      "Families: 11542\r\n",
      "Edges: 14347\r\n",
      "Persistent (min:0.89, max:0.99, sd:0.01, mean:0.99 ): 3667\r\n",
      "Shell ( min:0.05, max:0.9, sd:0.25, mean:0.28 ): 2021\r\n",
      "Cloud ( min:0.01, max:0.05, sd:0.01, mean:0.02 ): 5854\r\n",
      "Number of partitions: 13\r\n",
      "Shell S6 : 78\r\n",
      "Shell S7 : 167\r\n",
      "Shell S3 : 86\r\n",
      "Shell S11 : 655\r\n",
      "Shell S2 : 98\r\n",
      "Shell S5 : 106\r\n",
      "Shell S9 : 159\r\n",
      "Shell S10 : 227\r\n",
      "Shell S1 : 43\r\n",
      "Shell S8 : 342\r\n",
      "Shell S4 : 58\r\n",
      "Shell S_ : 2\r\n",
      "RGPs: 3830\r\n",
      "Spots: 130\r\n",
      "Modules: 360\r\n",
      "Families in Modules: 2635\r\n",
      "2022-11-21 12:09:53 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin all --anno data/pangenomes/Enterobacter.cloacae.list -o data/pangenomes/Enterobacter.cloacae -c 6 -f\r\n",
      "2022-11-21 12:09:53 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 12:09:53 annotate.py:l448 INFO\tReading data/pangenomes/Enterobacter.cloacae.list the list of organism files ...\r\n",
      "100%|███████████████████████████████████████| 129/129 [00:17<00:00,  7.34file/s]\r\n",
      "2022-11-21 12:10:11 annotate.py:l469 INFO\tgene identifiers used in the provided annotation files were unique, PPanGGOLiN will use them.\r\n",
      "2022-11-21 12:10:11 writeBinaries.py:l932 INFO\tWriting genome annotations...\r\n",
      "100%|█████████████████████████████████████| 129/129 [00:01<00:00, 77.61genome/s]\r\n",
      "2022-11-21 12:10:13 writeBinaries.py:l950 INFO\twriting the protein coding gene dna sequences\r\n",
      "100%|█████████████████████████████| 581629/581629 [00:00<00:00, 624816.42gene/s]\r\n",
      "2022-11-21 12:10:19 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.cloacae/pangenome.h5\r\n",
      "2022-11-21 12:10:19 writeSequences.py:l39 INFO\tWriting all of the CDS sequences...\r\n",
      "100%|█████████████████████████████| 581629/581629 [00:00<00:00, 647772.87gene/s]\r\n",
      "2022-11-21 12:10:20 cluster.py:l297 INFO\tClustering all of the genes sequences...\r\n",
      "2022-11-21 12:10:20 cluster.py:l86 INFO\tCreating sequence database...\r\n",
      "2022-11-21 12:10:23 cluster.py:l93 INFO\tClustering sequences...\r\n",
      "2022-11-21 12:10:35 cluster.py:l99 INFO\tExtracting cluster representatives...\r\n",
      "2022-11-21 12:10:36 cluster.py:l108 INFO\tWriting gene to family informations\r\n",
      "2022-11-21 12:10:36 cluster.py:l305 INFO\tAssociating fragments to their original gene family...\r\n",
      "2022-11-21 12:10:36 cluster.py:l153 INFO\tAligning cluster representatives...\r\n",
      "2022-11-21 12:11:02 cluster.py:l159 INFO\tExtracting alignments...\r\n",
      "2022-11-21 12:11:03 cluster.py:l197 INFO\tStarting with 24102 families\r\n",
      "2022-11-21 12:11:03 cluster.py:l231 INFO\tEnding with 19918 gene families\r\n",
      "2022-11-21 12:11:03 cluster.py:l242 INFO\tAdding protein sequences to the gene families\r\n",
      "2022-11-21 12:11:03 cluster.py:l256 INFO\tAdding 581629 genes to the gene families\r\n",
      "100%|█████████████████████████████| 581629/581629 [00:02<00:00, 264722.85gene/s]\r\n",
      "2022-11-21 12:11:06 writeBinaries.py:l955 INFO\tWriting gene families and gene associations...\r\n",
      "100%|█████████████████████████| 19918/19918 [00:00<00:00, 37764.20gene family/s]\r\n",
      "2022-11-21 12:11:06 writeBinaries.py:l957 INFO\tWriting gene families information...\r\n",
      "100%|████████████████████████| 19918/19918 [00:00<00:00, 126573.64gene family/s]\r\n",
      "2022-11-21 12:11:07 writeBinaries.py:l823 INFO\tUpdating annotations with fragment information\r\n",
      "100%|█████████████████████████████| 591490/591490 [00:01<00:00, 310049.99gene/s]\r\n",
      "2022-11-21 12:11:08 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.cloacae/pangenome.h5\r\n",
      "2022-11-21 12:11:08 makeGraph.py:l87 INFO\tComputing the neighbors graph...\r\n",
      "Processing GCF_902387775.1_UHGG_MGYG-HGUT-02500_genomic: 100%|█| 129/129 [00:02<\r\n",
      "2022-11-21 12:11:11 makeGraph.py:l108 INFO\tDone making the neighbors graph.\r\n",
      "2022-11-21 12:11:11 partition.py:l482 INFO\tEstimating the optimal number of partitions...\r\n",
      "100%|█████████████████████████| 19/19 [01:09<00:00,  3.68s/Number of partitions]\r\n",
      "2022-11-21 12:12:22 partition.py:l485 INFO\tThe number of partitions has been evaluated at 15\r\n",
      "2022-11-21 12:12:22 partition.py:l503 INFO\tPartitioning...\r\n",
      "2022-11-21 12:13:19 partition.py:l578 INFO\tPartitioned 129 genomes in 57.01 seconds.\r\n",
      "2022-11-21 12:13:19 writeBinaries.py:l966 INFO\tWriting the edges...\r\n",
      "100%|████████████████████████████████| 28401/28401 [00:00<00:00, 59054.06edge/s]\r\n",
      "2022-11-21 12:13:20 writeBinaries.py:l808 INFO\tUpdating gene families with partition information\r\n",
      "100%|█████████████████████████| 19918/19918 [00:00<00:00, 79916.88gene family/s]\r\n",
      "2022-11-21 12:13:20 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.cloacae/pangenome.h5\r\n",
      "2022-11-21 12:13:20 genomicIsland.py:l250 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 12:13:20 genomicIsland.py:l252 INFO\tCompute Regions of Genomic Plasticity ...\r\n",
      "100%|████████████████████████████████████| 129/129 [00:01<00:00, 99.44genomes/s]\r\n",
      "2022-11-21 12:13:22 genomicIsland.py:l257 INFO\tPredicted 7161 RGP\r\n",
      "2022-11-21 12:13:22 spot.py:l191 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 12:13:22 spot.py:l194 INFO\tDetecting hotspots in the pan...\r\n",
      "2022-11-21 12:13:22 spot.py:l118 INFO\t2662 RGPs were not used as they are on a contig border (or have less than 3 persistent gene families until the contig border)\r\n",
      "2022-11-21 12:13:22 spot.py:l120 INFO\t4499 RGPs are being used to predict spots of insertion\r\n",
      "2022-11-21 12:13:22 spot.py:l122 INFO\t437 number of different pairs of flanking gene families\r\n",
      "2022-11-21 12:13:22 spot.py:l203 INFO\t237 spots were detected\r\n",
      "2022-11-21 12:13:22 module.py:l121 INFO\tBuilding the graph...\r\n",
      "100%|█████████████████████████████████████| 129/129 [00:08<00:00, 16.11genome/s]\r\n",
      "2022-11-21 12:13:31 module.py:l123 INFO\tTook 8.01 seconds to build the graph to find modules in\r\n",
      "2022-11-21 12:13:31 module.py:l124 INFO\tThere are 19918 nodes and 151193 edges\r\n",
      "2022-11-21 12:13:31 module.py:l137 INFO\tThere are 5017 families among 758 modules\r\n",
      "2022-11-21 12:13:31 module.py:l138 INFO\tComputing modules took 0.64 seconds\r\n",
      "2022-11-21 12:13:31 writeBinaries.py:l975 INFO\tWriting Regions of Genomic Plasticity...\r\n",
      "100%|███████████████████████████████| 7161/7161 [00:00<00:00, 114900.35region/s]\r\n",
      "2022-11-21 12:13:31 writeBinaries.py:l980 INFO\tWriting Spots of Insertion...\r\n",
      "100%|███████████████████████████████████| 237/237 [00:00<00:00, 105168.22spot/s]\r\n",
      "2022-11-21 12:13:31 writeBinaries.py:l985 INFO\tWriting Modules...\r\n",
      "100%|████████████████████████████████| 758/758 [00:00<00:00, 204429.17modules/s]\r\n",
      "2022-11-21 12:13:31 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.cloacae/pangenome.h5\r\n",
      "2022-11-21 12:13:32 draw_spot.py:l650 INFO\tDrawing 116 spots\r\n",
      "2022-11-21 12:13:32 draw_spot.py:l559 INFO\tOrdering genes among regions, and drawing spots...\r\n",
      "  3%|█▍                                       | 4/116 [00:00<00:11, 10.02spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_130.html' already exists, will be overwritten.\r\n",
      "  6%|██▍                                      | 7/116 [00:00<00:10,  9.97spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_127.html' already exists, will be overwritten.\r\n",
      "  7%|██▊                                      | 8/116 [00:00<00:11,  9.72spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_139.html' already exists, will be overwritten.\r\n",
      "  9%|███▍                                    | 10/116 [00:01<00:10,  9.78spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_133.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_137.html' already exists, will be overwritten.\r\n",
      " 10%|████▏                                   | 12/116 [00:01<00:10,  9.76spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_57.html' already exists, will be overwritten.\r\n",
      " 12%|████▊                                   | 14/116 [00:01<00:10,  9.74spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_68.html' already exists, will be overwritten.\r\n",
      " 13%|█████▏                                  | 15/116 [00:01<00:10,  9.74spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_65.html' already exists, will be overwritten.\r\n",
      " 14%|█████▌                                  | 16/116 [00:01<00:10,  9.20spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_69.html' already exists, will be overwritten.\r\n",
      " 15%|█████▊                                  | 17/116 [00:01<00:10,  9.03spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_71.html' already exists, will be overwritten.\r\n",
      " 16%|██████▏                                 | 18/116 [00:01<00:11,  8.31spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_79.html' already exists, will be overwritten.\r\n",
      " 20%|███████▉                                | 23/116 [00:02<00:11,  8.44spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_87.html' already exists, will be overwritten.\r\n",
      " 22%|████████▌                               | 25/116 [00:02<00:10,  8.59spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_88.html' already exists, will be overwritten.\r\n",
      " 28%|███████████                             | 32/116 [00:03<00:08,  9.55spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_94.html' already exists, will be overwritten.\r\n",
      " 29%|███████████▋                            | 34/116 [00:03<00:09,  8.95spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_96.html' already exists, will be overwritten.\r\n",
      " 30%|████████████                            | 35/116 [00:03<00:09,  8.99spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_98.html' already exists, will be overwritten.\r\n",
      " 33%|█████████████                           | 38/116 [00:04<00:08,  9.00spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_101.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_209.html' already exists, will be overwritten.\r\n",
      " 34%|█████████████▊                          | 40/116 [00:04<00:08,  9.40spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_106.html' already exists, will be overwritten.\r\n",
      " 36%|██████████████▍                         | 42/116 [00:04<00:07,  9.52spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_103.html' already exists, will be overwritten.\r\n",
      " 38%|███████████████▏                        | 44/116 [00:04<00:07,  9.53spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_215.html' already exists, will be overwritten.\r\n",
      " 41%|████████████████▏                       | 47/116 [00:05<00:07,  9.83spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_56.html' already exists, will be overwritten.\r\n",
      " 43%|█████████████████▏                      | 50/116 [00:05<00:06,  9.60spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_29.html' already exists, will be overwritten.\r\n",
      " 44%|█████████████████▌                      | 51/116 [00:05<00:07,  8.42spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_143.html' already exists, will be overwritten.\r\n",
      " 45%|█████████████████▉                      | 52/116 [00:05<00:07,  8.65spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_27.html' already exists, will be overwritten.\r\n",
      " 47%|██████████████████▌                     | 54/116 [00:05<00:06,  8.87spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_31.html' already exists, will be overwritten.\r\n",
      " 47%|██████████████████▉                     | 55/116 [00:06<00:07,  8.08spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_30.html' already exists, will be overwritten.\r\n",
      " 48%|███████████████████▎                    | 56/116 [00:06<00:07,  8.51spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_25.html' already exists, will be overwritten.\r\n",
      " 49%|███████████████████▋                    | 57/116 [00:06<00:08,  6.89spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_26.html' already exists, will be overwritten.\r\n",
      " 50%|████████████████████                    | 58/116 [00:06<00:07,  7.33spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_153.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_22.html' already exists, will be overwritten.\r\n",
      " 52%|████████████████████▋                   | 60/116 [00:06<00:06,  8.19spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_24.html' already exists, will be overwritten.\r\n",
      " 53%|█████████████████████                   | 61/116 [00:06<00:06,  8.54spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_20.html' already exists, will be overwritten.\r\n",
      " 53%|█████████████████████▍                  | 62/116 [00:06<00:06,  8.79spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_6.html' already exists, will be overwritten.\r\n",
      " 54%|█████████████████████▋                  | 63/116 [00:06<00:05,  9.04spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_21.html' already exists, will be overwritten.\r\n",
      " 56%|██████████████████████▍                 | 65/116 [00:07<00:05,  9.03spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_18.html' already exists, will be overwritten.\r\n",
      " 58%|███████████████████████                 | 67/116 [00:07<00:05,  9.42spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_159.html' already exists, will be overwritten.\r\n",
      " 59%|███████████████████████▍                | 68/116 [00:07<00:05,  9.51spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_19.html' already exists, will be overwritten.\r\n",
      " 60%|████████████████████████▏               | 70/116 [00:07<00:04,  9.39spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_16.html' already exists, will be overwritten.\r\n",
      " 61%|████████████████████████▍               | 71/116 [00:07<00:06,  7.35spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_154.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_14.html' already exists, will be overwritten.\r\n",
      " 63%|█████████████████████████▏              | 73/116 [00:08<00:05,  7.42spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_161.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_17.html' already exists, will be overwritten.\r\n",
      " 65%|█████████████████████████▊              | 75/116 [00:08<00:04,  8.29spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_12.html' already exists, will be overwritten.\r\n",
      " 66%|██████████████████████████▏             | 76/116 [00:08<00:04,  8.42spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_15.html' already exists, will be overwritten.\r\n",
      " 66%|██████████████████████████▌             | 77/116 [00:08<00:04,  8.15spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_13.html' already exists, will be overwritten.\r\n",
      " 67%|██████████████████████████▉             | 78/116 [00:08<00:04,  8.43spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_4.html' already exists, will be overwritten.\r\n",
      " 68%|███████████████████████████▏            | 79/116 [00:08<00:04,  8.77spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_10.html' already exists, will be overwritten.\r\n",
      " 69%|███████████████████████████▌            | 80/116 [00:08<00:04,  8.40spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_11.html' already exists, will be overwritten.\r\n",
      " 70%|███████████████████████████▉            | 81/116 [00:09<00:04,  8.75spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_8.html' already exists, will be overwritten.\r\n",
      " 71%|████████████████████████████▎           | 82/116 [00:09<00:03,  8.70spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_168.html' already exists, will be overwritten.\r\n",
      " 73%|█████████████████████████████▎          | 85/116 [00:09<00:03,  9.46spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_62.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_167.html' already exists, will be overwritten.\r\n",
      " 75%|██████████████████████████████          | 87/116 [00:09<00:03,  9.48spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_60.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_7.html' already exists, will be overwritten.\r\n",
      " 77%|██████████████████████████████▋         | 89/116 [00:09<00:02,  9.29spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_0.html' already exists, will be overwritten.\r\n",
      " 78%|███████████████████████████████         | 90/116 [00:10<00:02,  9.27spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_32.html' already exists, will be overwritten.\r\n",
      " 79%|███████████████████████████████▋        | 92/116 [00:10<00:02,  9.41spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_169.html' already exists, will be overwritten.\r\n",
      " 80%|████████████████████████████████        | 93/116 [00:10<00:02,  9.46spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_35.html' already exists, will be overwritten.\r\n",
      " 81%|████████████████████████████████▍       | 94/116 [00:10<00:02,  9.19spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_2.html' already exists, will be overwritten.\r\n",
      " 82%|████████████████████████████████▊       | 95/116 [00:10<00:02,  9.34spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_34.html' already exists, will be overwritten.\r\n",
      " 84%|█████████████████████████████████▍      | 97/116 [00:10<00:02,  7.79spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_174.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_39.html' already exists, will be overwritten.\r\n",
      " 85%|██████████████████████████████████▏     | 99/116 [00:11<00:01,  8.55spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_36.html' already exists, will be overwritten.\r\n",
      " 86%|█████████████████████████████████▌     | 100/116 [00:11<00:02,  7.18spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_41.html' already exists, will be overwritten.\r\n",
      " 87%|█████████████████████████████████▉     | 101/116 [00:11<00:02,  6.84spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_61.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_38.html' already exists, will be overwritten.\r\n",
      " 90%|██████████████████████████████████▉    | 104/116 [00:11<00:01,  7.53spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_42.html' already exists, will be overwritten.\r\n",
      " 91%|███████████████████████████████████▎   | 105/116 [00:11<00:01,  7.64spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_44.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_45.html' already exists, will be overwritten.\r\n",
      " 92%|███████████████████████████████████▉   | 107/116 [00:12<00:01,  7.67spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_5.html' already exists, will be overwritten.\r\n",
      " 93%|████████████████████████████████████▎  | 108/116 [00:12<00:01,  7.26spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_46.html' already exists, will be overwritten.\r\n",
      " 94%|████████████████████████████████████▋  | 109/116 [00:12<00:00,  7.14spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_47.html' already exists, will be overwritten.\r\n",
      " 97%|█████████████████████████████████████▋ | 112/116 [00:12<00:00,  8.23spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_1.html' already exists, will be overwritten.\r\n",
      " 97%|█████████████████████████████████████▉ | 113/116 [00:12<00:00,  8.02spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_49.html' already exists, will be overwritten.\r\n",
      " 98%|██████████████████████████████████████▎| 114/116 [00:13<00:00,  8.40spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_52.html' already exists, will be overwritten.\r\n",
      " 99%|██████████████████████████████████████▋| 115/116 [00:13<00:00,  8.62spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.cloacae/spot_figures/spot_53.html' already exists, will be overwritten.\r\n",
      "100%|███████████████████████████████████████| 116/116 [00:13<00:00,  8.73spot/s]\r\n",
      "2022-11-21 12:13:45 draw_spot.py:l617 INFO\tDone drawing spot(s), they can be found in the directory: 'data/pangenomes/Enterobacter.cloacae/spot_figures'\r\n",
      "2022-11-21 12:13:45 tile_plot.py:l38 INFO\tDrawing the tile plot...\r\n",
      "2022-11-21 12:13:45 tile_plot.py:l56 INFO\tstart with matrice\r\n",
      "2022-11-21 12:13:46 tile_plot.py:l72 INFO\tdone with making the dendrogram to order the organisms on the plot\r\n",
      "2022-11-21 12:13:46 tile_plot.py:l106 INFO\tGetting the gene name(s) and the number for each tile of the plot ...\r\n",
      "2022-11-21 12:13:47 tile_plot.py:l116 INFO\tDone extracting names and numbers. Making the heatmap ...\r\n",
      "2022-11-21 12:13:52 tile_plot.py:l173 INFO\tDrawing the figure itself...\r\n",
      "2022-11-21 12:14:17 tile_plot.py:l175 INFO\tDone with the tile plot : 'data/pangenomes/Enterobacter.cloacae/tile_plot.html' \r\n",
      "2022-11-21 12:14:17 ucurve.py:l23 INFO\tDrawing the U-shaped curve...\r\n",
      "2022-11-21 12:14:17 ucurve.py:l77 INFO\tDone drawing the U-shaped curve : 'data/pangenomes/Enterobacter.cloacae/Ushaped_plot.html'\r\n",
      "2022-11-21 12:14:17 writeFlat.py:l399 INFO\tWriting the gene presence absence file ...\r\n",
      "2022-11-21 12:14:17 writeFlat.py:l328 INFO\tWriting the .csv file ...\r\n",
      "2022-11-21 12:14:17 writeFlat.py:l305 INFO\tWriting the gexf file for the pangenome graph...\r\n",
      "2022-11-21 12:14:17 writeFlat.py:l305 INFO\tWriting the light gexf file for the pangenome graph...\r\n",
      "2022-11-21 12:14:17 writeFlat.py:l582 INFO\tWriting the projection files...\r\n",
      "2022-11-21 12:14:17 writeFlat.py:l431 INFO\tWriting pangenome statistics...\r\n",
      "2022-11-21 12:14:17 writeFlat.py:l432 INFO\tWriting statistics on persistent duplication...\r\n",
      "2022-11-21 12:14:17 writeFlat.py:l419 INFO\tDone writing the gene presence absence file : 'data/pangenomes/Enterobacter.cloacae/gene_presence_absence.Rtab'\r\n",
      "2022-11-21 12:14:17 writeFlat.py:l157 INFO\tWriting the json file for the pangenome graph...\r\n",
      "2022-11-21 12:14:17 writeFlat.py:l454 INFO\tDone writing stats on persistent duplication\r\n",
      "2022-11-21 12:14:17 writeFlat.py:l455 INFO\tWriting genome per genome statistics (completeness and counts)...\r\n",
      "2022-11-21 12:14:19 writeFlat.py:l520 INFO\tDone writing genome per genome statistics\r\n",
      "2022-11-21 12:14:19 writeFlat.py:l598 INFO\tWriting the list of gene families for each partition ...\r\n",
      "2022-11-21 12:14:19 writeFlat.py:l625 INFO\tDone writing the list of gene families for each partition\r\n",
      "2022-11-21 12:14:20 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Enterobacter.cloacae/pangenomeGraph_light.gexf'\r\n",
      "2022-11-21 12:14:20 writeFlat.py:l389 INFO\tDone writing the matrix : 'data/pangenomes/Enterobacter.cloacae/matrix.csv'\r\n",
      "2022-11-21 12:14:20 writeFlat.py:l776 INFO\tWriting functional modules...\r\n",
      "2022-11-21 12:14:20 writeFlat.py:l784 INFO\tDone writing functional modules to: 'data/pangenomes/Enterobacter.cloacae/functional_modules.tsv'\r\n",
      "2022-11-21 12:14:20 writeFlat.py:l752 INFO\tWriting functional modules summary...\r\n",
      "2022-11-21 12:14:20 writeFlat.py:l691 INFO\tDone writing spots in : 'data/pangenomes/Enterobacter.cloacae/summarize_spots.tsv'\r\n",
      "2022-11-21 12:14:20 writeFlat.py:l793 INFO\tWriting modules to organisms associations...\r\n",
      "2022-11-21 12:14:20 writeFlat.py:l767 INFO\tDone writing module summary: 'data/pangenomes/Enterobacter.cloacae/modules_summary.tsv'\r\n",
      "2022-11-21 12:14:20 writeFlat.py:l814 INFO\tWriting modules to spot associations...\r\n",
      "2022-11-21 12:14:20 writeFlat.py:l837 INFO\tDone writing module to spot associations to: data/pangenomes/Enterobacter.cloacae/modules_spots.tsv\r\n",
      "2022-11-21 12:14:20 writeFlat.py:l846 INFO\tClustering RGPs based on module content...\r\n",
      "2022-11-21 12:14:20 writeFlat.py:l881 INFO\tRGP and associated modules are listed in : data/pangenomes/Enterobacter.cloacae/modules_RGP_lists.tsv\r\n",
      "2022-11-21 12:14:22 writeFlat.py:l164 INFO\tDone writing the json file : 'data/pangenomes/Enterobacter.cloacae/pangenomeGraph.json'\r\n",
      "2022-11-21 12:14:22 writeFlat.py:l588 INFO\tDone writing the projection files\r\n",
      "2022-11-21 12:14:24 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Enterobacter.cloacae/pangenomeGraph.gexf'\r\n",
      "2022-11-21 12:14:32 writeFlat.py:l804 INFO\tDone writing modules to organisms associations to: 'data/pangenomes/Enterobacter.cloacae/modules_in_organisms.tsv'\r\n",
      "2022-11-21 12:14:32 all.py:l125 INFO\tAnnotation took : 18.7 seconds\r\n",
      "2022-11-21 12:14:32 all.py:l126 INFO\tClustering took : 46.58 seconds\r\n",
      "2022-11-21 12:14:32 all.py:l127 INFO\tBuilding the graph took : 2.87 seconds\r\n",
      "2022-11-21 12:14:32 all.py:l128 INFO\tPartitioning the pangenome took : 127.51 seconds\r\n",
      "2022-11-21 12:14:32 all.py:l129 INFO\tPredicting RGP took : 1.63 seconds\r\n",
      "2022-11-21 12:14:32 all.py:l130 INFO\tGathering RGP into spots took : 14.79 seconds\r\n",
      "2022-11-21 12:14:32 all.py:l131 INFO\tPredicting modules took : 8.66 seconds\r\n",
      "2022-11-21 12:14:32 all.py:l132 INFO\tWriting the pangenome data in HDF5 took : 9.05 seconds\r\n",
      "2022-11-21 12:14:32 all.py:l134 INFO\tWriting descriptive files for the pangenome took : 15.59 seconds\r\n",
      "Genes: 581629\r\n",
      "Organisms: 129\r\n",
      "Families: 19918\r\n",
      "Edges: 28401\r\n",
      "Persistent (min:0.83, max:0.96, sd:0.02, mean:0.95 ): 3630\r\n",
      "Shell ( min:0.03, max:0.95, sd:0.23, mean:0.21 ): 4002\r\n",
      "Cloud ( min:0.01, max:0.05, sd:0.01, mean:0.01 ): 12286\r\n",
      "Number of partitions: 15\r\n",
      "Shell S10 : 680\r\n",
      "Shell S12 : 378\r\n",
      "Shell S5 : 127\r\n",
      "Shell S1 : 158\r\n",
      "Shell S13 : 1691\r\n",
      "Shell S11 : 208\r\n",
      "Shell S3 : 87\r\n",
      "Shell S4 : 114\r\n",
      "Shell S7 : 26\r\n",
      "Shell S2 : 41\r\n",
      "Shell S8 : 109\r\n",
      "Shell S9 : 328\r\n",
      "Shell S6 : 50\r\n",
      "Shell S_ : 5\r\n",
      "RGPs: 7161\r\n",
      "Spots: 237\r\n",
      "Modules: 758\r\n",
      "Families in Modules: 5017\r\n",
      "2022-11-21 12:14:40 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin all --anno data/pangenomes/Enterobacter.hormaechei_A.list -o data/pangenomes/Enterobacter.hormaechei_A -c 6 -f\r\n",
      "2022-11-21 12:14:40 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 12:14:40 annotate.py:l448 INFO\tReading data/pangenomes/Enterobacter.hormaechei_A.list the list of organism files ...\r\n",
      "100%|███████████████████████████████████████| 159/159 [00:21<00:00,  7.29file/s]\r\n",
      "2022-11-21 12:15:03 annotate.py:l469 INFO\tgene identifiers used in the provided annotation files were unique, PPanGGOLiN will use them.\r\n",
      "2022-11-21 12:15:03 writeBinaries.py:l932 INFO\tWriting genome annotations...\r\n",
      "100%|█████████████████████████████████████| 159/159 [00:01<00:00, 81.08genome/s]\r\n",
      "2022-11-21 12:15:06 writeBinaries.py:l950 INFO\twriting the protein coding gene dna sequences\r\n",
      "100%|█████████████████████████████| 739586/739586 [00:01<00:00, 624530.73gene/s]\r\n",
      "2022-11-21 12:15:11 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.hormaechei_A/pangenome.h5\r\n",
      "2022-11-21 12:15:11 writeSequences.py:l39 INFO\tWriting all of the CDS sequences...\r\n",
      "100%|█████████████████████████████| 739586/739586 [00:01<00:00, 654460.08gene/s]\r\n",
      "2022-11-21 12:15:12 cluster.py:l297 INFO\tClustering all of the genes sequences...\r\n",
      "2022-11-21 12:15:12 cluster.py:l86 INFO\tCreating sequence database...\r\n",
      "2022-11-21 12:15:16 cluster.py:l93 INFO\tClustering sequences...\r\n",
      "2022-11-21 12:15:31 cluster.py:l99 INFO\tExtracting cluster representatives...\r\n",
      "2022-11-21 12:15:31 cluster.py:l108 INFO\tWriting gene to family informations\r\n",
      "2022-11-21 12:15:31 cluster.py:l305 INFO\tAssociating fragments to their original gene family...\r\n",
      "2022-11-21 12:15:32 cluster.py:l153 INFO\tAligning cluster representatives...\r\n",
      "2022-11-21 12:15:53 cluster.py:l159 INFO\tExtracting alignments...\r\n",
      "2022-11-21 12:15:54 cluster.py:l197 INFO\tStarting with 19672 families\r\n",
      "2022-11-21 12:15:54 cluster.py:l231 INFO\tEnding with 18180 gene families\r\n",
      "2022-11-21 12:15:54 cluster.py:l242 INFO\tAdding protein sequences to the gene families\r\n",
      "2022-11-21 12:15:54 cluster.py:l256 INFO\tAdding 739586 genes to the gene families\r\n",
      "100%|█████████████████████████████| 739586/739586 [00:02<00:00, 290873.30gene/s]\r\n",
      "2022-11-21 12:15:56 writeBinaries.py:l955 INFO\tWriting gene families and gene associations...\r\n",
      "100%|█████████████████████████| 18180/18180 [00:00<00:00, 29964.72gene family/s]\r\n",
      "2022-11-21 12:15:57 writeBinaries.py:l957 INFO\tWriting gene families information...\r\n",
      "100%|████████████████████████| 18180/18180 [00:00<00:00, 135213.28gene family/s]\r\n",
      "2022-11-21 12:15:58 writeBinaries.py:l823 INFO\tUpdating annotations with fragment information\r\n",
      "100%|█████████████████████████████| 757004/757004 [00:02<00:00, 352882.79gene/s]\r\n",
      "2022-11-21 12:16:00 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.hormaechei_A/pangenome.h5\r\n",
      "2022-11-21 12:16:00 makeGraph.py:l87 INFO\tComputing the neighbors graph...\r\n",
      "Processing GCF_900497145.1_ASM90049714v1_genomic: 100%|█| 159/159 [00:03<00:00, \r\n",
      "2022-11-21 12:16:03 makeGraph.py:l108 INFO\tDone making the neighbors graph.\r\n",
      "2022-11-21 12:16:03 partition.py:l482 INFO\tEstimating the optimal number of partitions...\r\n",
      "100%|█████████████████████████| 19/19 [01:10<00:00,  3.72s/Number of partitions]\r\n",
      "2022-11-21 12:17:14 partition.py:l485 INFO\tThe number of partitions has been evaluated at 7\r\n",
      "2022-11-21 12:17:14 partition.py:l503 INFO\tPartitioning...\r\n",
      "2022-11-21 12:17:35 partition.py:l578 INFO\tPartitioned 159 genomes in 20.44 seconds.\r\n",
      "2022-11-21 12:17:35 writeBinaries.py:l966 INFO\tWriting the edges...\r\n",
      "100%|████████████████████████████████| 29822/29822 [00:00<00:00, 50745.07edge/s]\r\n",
      "2022-11-21 12:17:36 writeBinaries.py:l808 INFO\tUpdating gene families with partition information\r\n",
      "100%|█████████████████████████| 18180/18180 [00:00<00:00, 82332.90gene family/s]\r\n",
      "2022-11-21 12:17:36 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.hormaechei_A/pangenome.h5\r\n",
      "2022-11-21 12:17:36 genomicIsland.py:l250 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 12:17:36 genomicIsland.py:l252 INFO\tCompute Regions of Genomic Plasticity ...\r\n",
      "100%|████████████████████████████████████| 159/159 [00:03<00:00, 43.58genomes/s]\r\n",
      "2022-11-21 12:17:40 genomicIsland.py:l257 INFO\tPredicted 5746 RGP\r\n",
      "2022-11-21 12:17:40 spot.py:l191 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 12:17:40 spot.py:l194 INFO\tDetecting hotspots in the pan...\r\n",
      "2022-11-21 12:17:40 spot.py:l118 INFO\t347 RGPs were not used as they are on a contig border (or have less than 3 persistent gene families until the contig border)\r\n",
      "2022-11-21 12:17:40 spot.py:l120 INFO\t5399 RGPs are being used to predict spots of insertion\r\n",
      "2022-11-21 12:17:40 spot.py:l122 INFO\t450 number of different pairs of flanking gene families\r\n",
      "2022-11-21 12:17:41 spot.py:l203 INFO\t279 spots were detected\r\n",
      "2022-11-21 12:17:41 module.py:l121 INFO\tBuilding the graph...\r\n",
      "100%|█████████████████████████████████████| 159/159 [00:08<00:00, 17.73genome/s]\r\n",
      "2022-11-21 12:17:50 module.py:l123 INFO\tTook 8.97 seconds to build the graph to find modules in\r\n",
      "2022-11-21 12:17:50 module.py:l124 INFO\tThere are 18180 nodes and 169529 edges\r\n",
      "2022-11-21 12:17:51 module.py:l137 INFO\tThere are 5321 families among 738 modules\r\n",
      "2022-11-21 12:17:51 module.py:l138 INFO\tComputing modules took 0.74 seconds\r\n",
      "2022-11-21 12:17:51 writeBinaries.py:l975 INFO\tWriting Regions of Genomic Plasticity...\r\n",
      "100%|████████████████████████████████| 5746/5746 [00:00<00:00, 69769.28region/s]\r\n",
      "2022-11-21 12:17:51 writeBinaries.py:l980 INFO\tWriting Spots of Insertion...\r\n",
      "100%|███████████████████████████████████| 279/279 [00:00<00:00, 153752.57spot/s]\r\n",
      "2022-11-21 12:17:51 writeBinaries.py:l985 INFO\tWriting Modules...\r\n",
      "100%|████████████████████████████████| 738/738 [00:00<00:00, 254471.91modules/s]\r\n",
      "2022-11-21 12:17:51 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.hormaechei_A/pangenome.h5\r\n",
      "2022-11-21 12:17:51 draw_spot.py:l650 INFO\tDrawing 130 spots\r\n",
      "2022-11-21 12:17:51 draw_spot.py:l559 INFO\tOrdering genes among regions, and drawing spots...\r\n",
      "  4%|█▌                                       | 5/130 [00:00<00:15,  8.29spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_151.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_152.html' already exists, will be overwritten.\r\n",
      "  5%|██▏                                      | 7/130 [00:00<00:13,  9.31spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_154.html' already exists, will be overwritten.\r\n",
      "  7%|██▊                                      | 9/130 [00:00<00:12,  9.89spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_157.html' already exists, will be overwritten.\r\n",
      "  8%|███▍                                    | 11/130 [00:01<00:11, 10.16spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_162.html' already exists, will be overwritten.\r\n",
      " 10%|████                                    | 13/130 [00:01<00:11, 10.07spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_163.html' already exists, will be overwritten.\r\n",
      " 11%|████▎                                   | 14/130 [00:01<00:11, 10.03spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_165.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_166.html' already exists, will be overwritten.\r\n",
      " 12%|████▉                                   | 16/130 [00:01<00:11, 10.32spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_167.html' already exists, will be overwritten.\r\n",
      " 14%|█████▌                                  | 18/130 [00:01<00:11,  9.69spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_170.html' already exists, will be overwritten.\r\n",
      " 15%|█████▊                                  | 19/130 [00:01<00:11,  9.46spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_171.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_172.html' already exists, will be overwritten.\r\n",
      " 17%|██████▊                                 | 22/130 [00:02<00:11,  9.27spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_53.html' already exists, will be overwritten.\r\n",
      " 18%|███████▍                                | 24/130 [00:02<00:17,  5.94spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_2.html' already exists, will be overwritten.\r\n",
      " 19%|███████▋                                | 25/130 [00:03<00:19,  5.48spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_0.html' already exists, will be overwritten.\r\n",
      " 21%|████████▎                               | 27/130 [00:03<00:14,  6.90spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_27.html' already exists, will be overwritten.\r\n",
      " 22%|████████▌                               | 28/130 [00:03<00:16,  6.19spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_54.html' already exists, will be overwritten.\r\n",
      " 22%|████████▉                               | 29/130 [00:03<00:20,  4.99spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_121.html' already exists, will be overwritten.\r\n",
      " 24%|█████████▌                              | 31/130 [00:03<00:15,  6.43spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_29.html' already exists, will be overwritten.\r\n",
      " 25%|█████████▊                              | 32/130 [00:04<00:14,  6.83spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_52.html' already exists, will be overwritten.\r\n",
      " 25%|██████████▏                             | 33/130 [00:04<00:13,  7.40spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_26.html' already exists, will be overwritten.\r\n",
      " 28%|███████████▍                            | 37/130 [00:04<00:12,  7.74spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_25.html' already exists, will be overwritten.\r\n",
      " 29%|███████████▋                            | 38/130 [00:04<00:11,  8.20spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_122.html' already exists, will be overwritten.\r\n",
      " 30%|████████████                            | 39/130 [00:04<00:11,  7.88spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_24.html' already exists, will be overwritten.\r\n",
      " 31%|████████████▎                           | 40/130 [00:04<00:11,  8.16spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_23.html' already exists, will be overwritten.\r\n",
      " 32%|████████████▌                           | 41/130 [00:05<00:10,  8.34spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_124.html' already exists, will be overwritten.\r\n",
      " 32%|████████████▉                           | 42/130 [00:05<00:10,  8.38spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_22.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_21.html' already exists, will be overwritten.\r\n",
      " 34%|█████████████▌                          | 44/130 [00:05<00:09,  8.83spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_50.html' already exists, will be overwritten.\r\n",
      " 35%|█████████████▊                          | 45/130 [00:05<00:16,  5.13spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_123.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_20.html' already exists, will be overwritten.\r\n",
      " 37%|██████████████▊                         | 48/130 [00:06<00:11,  7.01spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_19.html' already exists, will be overwritten.\r\n",
      " 38%|███████████████                         | 49/130 [00:06<00:11,  7.22spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_125.html' already exists, will be overwritten.\r\n",
      " 39%|███████████████▋                        | 51/130 [00:06<00:09,  8.30spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_18.html' already exists, will be overwritten.\r\n",
      " 40%|████████████████                        | 52/130 [00:06<00:09,  8.34spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_1.html' already exists, will be overwritten.\r\n",
      " 41%|████████████████▎                       | 53/130 [00:06<00:09,  8.47spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_17.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_16.html' already exists, will be overwritten.\r\n",
      " 42%|████████████████▉                       | 55/130 [00:06<00:09,  8.23spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_15.html' already exists, will be overwritten.\r\n",
      " 43%|█████████████████▏                      | 56/130 [00:07<00:10,  6.88spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_14.html' already exists, will be overwritten.\r\n",
      " 44%|█████████████████▌                      | 57/130 [00:07<00:11,  6.11spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_131.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_13.html' already exists, will be overwritten.\r\n",
      " 45%|██████████████████▏                     | 59/130 [00:07<00:09,  7.22spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_12.html' already exists, will be overwritten.\r\n",
      " 48%|███████████████████                     | 62/130 [00:07<00:07,  8.59spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_10.html' already exists, will be overwritten.\r\n",
      " 48%|███████████████████▍                    | 63/130 [00:07<00:07,  8.85spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_9.html' already exists, will be overwritten.\r\n",
      " 49%|███████████████████▋                    | 64/130 [00:08<00:07,  8.56spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_138.html' already exists, will be overwritten.\r\n",
      " 51%|████████████████████▎                   | 66/130 [00:08<00:06,  9.32spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_7.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_51.html' already exists, will be overwritten.\r\n",
      " 52%|████████████████████▉                   | 68/130 [00:08<00:06,  9.77spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_6.html' already exists, will be overwritten.\r\n",
      " 53%|█████████████████████▏                  | 69/130 [00:08<00:06,  9.57spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_5.html' already exists, will be overwritten.\r\n",
      " 54%|█████████████████████▌                  | 70/130 [00:08<00:06,  9.67spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_3.html' already exists, will be overwritten.\r\n",
      " 55%|█████████████████████▊                  | 71/130 [00:08<00:07,  7.41spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_30.html' already exists, will be overwritten.\r\n",
      " 55%|██████████████████████▏                 | 72/130 [00:09<00:07,  7.93spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_31.html' already exists, will be overwritten.\r\n",
      " 56%|██████████████████████▍                 | 73/130 [00:09<00:07,  7.70spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_32.html' already exists, will be overwritten.\r\n",
      " 58%|███████████████████████                 | 75/130 [00:09<00:06,  8.71spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_33.html' already exists, will be overwritten.\r\n",
      " 58%|███████████████████████▍                | 76/130 [00:09<00:06,  8.70spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_34.html' already exists, will be overwritten.\r\n",
      " 59%|███████████████████████▋                | 77/130 [00:09<00:06,  8.28spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_36.html' already exists, will be overwritten.\r\n",
      " 60%|████████████████████████                | 78/130 [00:09<00:06,  7.96spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_37.html' already exists, will be overwritten.\r\n",
      " 61%|████████████████████████▎               | 79/130 [00:09<00:06,  8.17spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_38.html' already exists, will be overwritten.\r\n",
      " 62%|████████████████████████▌               | 80/130 [00:10<00:07,  7.04spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_39.html' already exists, will be overwritten.\r\n",
      " 62%|████████████████████████▉               | 81/130 [00:10<00:07,  6.47spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_40.html' already exists, will be overwritten.\r\n",
      " 63%|█████████████████████████▏              | 82/130 [00:10<00:07,  6.01spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_41.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_42.html' already exists, will be overwritten.\r\n",
      " 65%|█████████████████████████▊              | 84/130 [00:10<00:06,  7.42spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_43.html' already exists, will be overwritten.\r\n",
      " 65%|██████████████████████████▏             | 85/130 [00:10<00:05,  7.80spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_44.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_45.html' already exists, will be overwritten.\r\n",
      " 67%|██████████████████████████▊             | 87/130 [00:10<00:04,  8.61spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_46.html' already exists, will be overwritten.\r\n",
      " 68%|███████████████████████████             | 88/130 [00:11<00:05,  8.19spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_47.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_48.html' already exists, will be overwritten.\r\n",
      " 69%|███████████████████████████▋            | 90/130 [00:11<00:04,  8.88spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_49.html' already exists, will be overwritten.\r\n",
      " 71%|████████████████████████████▎           | 92/130 [00:11<00:04,  8.34spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_58.html' already exists, will be overwritten.\r\n",
      " 72%|████████████████████████████▌           | 93/130 [00:11<00:04,  8.18spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_59.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_60.html' already exists, will be overwritten.\r\n",
      " 73%|█████████████████████████████▏          | 95/130 [00:11<00:04,  8.74spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_62.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_63.html' already exists, will be overwritten.\r\n",
      " 75%|█████████████████████████████▊          | 97/130 [00:12<00:03,  9.29spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_65.html' already exists, will be overwritten.\r\n",
      " 76%|██████████████████████████████▍         | 99/130 [00:12<00:03,  9.55spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_66.html' already exists, will be overwritten.\r\n",
      " 77%|██████████████████████████████         | 100/130 [00:12<00:03,  9.38spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_67.html' already exists, will be overwritten.\r\n",
      " 78%|██████████████████████████████▎        | 101/130 [00:12<00:03,  8.88spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_68.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_70.html' already exists, will be overwritten.\r\n",
      " 79%|██████████████████████████████▉        | 103/130 [00:12<00:02,  9.35spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_71.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_72.html' already exists, will be overwritten.\r\n",
      " 81%|███████████████████████████████▌       | 105/130 [00:14<00:10,  2.44spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_74.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_75.html' already exists, will be overwritten.\r\n",
      " 83%|████████████████████████████████▍      | 108/130 [00:14<00:05,  3.76spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_77.html' already exists, will be overwritten.\r\n",
      " 84%|████████████████████████████████▋      | 109/130 [00:15<00:04,  4.30spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_79.html' already exists, will be overwritten.\r\n",
      " 93%|████████████████████████████████████▎  | 121/130 [00:16<00:01,  8.60spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_225.html' already exists, will be overwritten.\r\n",
      " 96%|█████████████████████████████████████▌ | 125/130 [00:16<00:00,  9.65spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_118.html' already exists, will be overwritten.\r\n",
      " 98%|██████████████████████████████████████ | 127/130 [00:17<00:00,  9.91spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_119.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_120.html' already exists, will be overwritten.\r\n",
      " 99%|██████████████████████████████████████▋| 129/130 [00:17<00:00,  9.78spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures/spot_234.html' already exists, will be overwritten.\r\n",
      "100%|███████████████████████████████████████| 130/130 [00:17<00:00,  7.46spot/s]\r\n",
      "2022-11-21 12:18:09 draw_spot.py:l617 INFO\tDone drawing spot(s), they can be found in the directory: 'data/pangenomes/Enterobacter.hormaechei_A/spot_figures'\r\n",
      "2022-11-21 12:18:09 tile_plot.py:l38 INFO\tDrawing the tile plot...\r\n",
      "2022-11-21 12:18:09 tile_plot.py:l56 INFO\tstart with matrice\r\n",
      "2022-11-21 12:18:10 tile_plot.py:l72 INFO\tdone with making the dendrogram to order the organisms on the plot\r\n",
      "2022-11-21 12:18:10 tile_plot.py:l106 INFO\tGetting the gene name(s) and the number for each tile of the plot ...\r\n",
      "2022-11-21 12:18:11 tile_plot.py:l116 INFO\tDone extracting names and numbers. Making the heatmap ...\r\n",
      "2022-11-21 12:18:17 tile_plot.py:l173 INFO\tDrawing the figure itself...\r\n",
      "2022-11-21 12:18:43 tile_plot.py:l175 INFO\tDone with the tile plot : 'data/pangenomes/Enterobacter.hormaechei_A/tile_plot.html' \r\n",
      "2022-11-21 12:18:43 ucurve.py:l23 INFO\tDrawing the U-shaped curve...\r\n",
      "2022-11-21 12:18:43 ucurve.py:l77 INFO\tDone drawing the U-shaped curve : 'data/pangenomes/Enterobacter.hormaechei_A/Ushaped_plot.html'\r\n",
      "2022-11-21 12:18:44 writeFlat.py:l328 INFO\tWriting the .csv file ...\r\n",
      "2022-11-21 12:18:44 writeFlat.py:l399 INFO\tWriting the gene presence absence file ...\r\n",
      "2022-11-21 12:18:44 writeFlat.py:l305 INFO\tWriting the gexf file for the pangenome graph...\r\n",
      "2022-11-21 12:18:44 writeFlat.py:l305 INFO\tWriting the light gexf file for the pangenome graph...\r\n",
      "2022-11-21 12:18:44 writeFlat.py:l582 INFO\tWriting the projection files...\r\n",
      "2022-11-21 12:18:44 writeFlat.py:l431 INFO\tWriting pangenome statistics...\r\n",
      "2022-11-21 12:18:44 writeFlat.py:l432 INFO\tWriting statistics on persistent duplication...\r\n",
      "2022-11-21 12:18:44 writeFlat.py:l419 INFO\tDone writing the gene presence absence file : 'data/pangenomes/Enterobacter.hormaechei_A/gene_presence_absence.Rtab'\r\n",
      "2022-11-21 12:18:44 writeFlat.py:l157 INFO\tWriting the json file for the pangenome graph...\r\n",
      "2022-11-21 12:18:44 writeFlat.py:l454 INFO\tDone writing stats on persistent duplication\r\n",
      "2022-11-21 12:18:44 writeFlat.py:l455 INFO\tWriting genome per genome statistics (completeness and counts)...\r\n",
      "2022-11-21 12:18:46 writeFlat.py:l520 INFO\tDone writing genome per genome statistics\r\n",
      "2022-11-21 12:18:46 writeFlat.py:l598 INFO\tWriting the list of gene families for each partition ...\r\n",
      "2022-11-21 12:18:46 writeFlat.py:l389 INFO\tDone writing the matrix : 'data/pangenomes/Enterobacter.hormaechei_A/matrix.csv'\r\n",
      "2022-11-21 12:18:46 writeFlat.py:l625 INFO\tDone writing the list of gene families for each partition\r\n",
      "2022-11-21 12:18:47 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Enterobacter.hormaechei_A/pangenomeGraph_light.gexf'\r\n",
      "2022-11-21 12:18:47 writeFlat.py:l776 INFO\tWriting functional modules...\r\n",
      "2022-11-21 12:18:47 writeFlat.py:l784 INFO\tDone writing functional modules to: 'data/pangenomes/Enterobacter.hormaechei_A/functional_modules.tsv'\r\n",
      "2022-11-21 12:18:47 writeFlat.py:l752 INFO\tWriting functional modules summary...\r\n",
      "2022-11-21 12:18:47 writeFlat.py:l767 INFO\tDone writing module summary: 'data/pangenomes/Enterobacter.hormaechei_A/modules_summary.tsv'\r\n",
      "2022-11-21 12:18:47 writeFlat.py:l793 INFO\tWriting modules to organisms associations...\r\n",
      "2022-11-21 12:18:47 writeFlat.py:l691 INFO\tDone writing spots in : 'data/pangenomes/Enterobacter.hormaechei_A/summarize_spots.tsv'\r\n",
      "2022-11-21 12:18:47 writeFlat.py:l814 INFO\tWriting modules to spot associations...\r\n",
      "2022-11-21 12:18:47 writeFlat.py:l837 INFO\tDone writing module to spot associations to: data/pangenomes/Enterobacter.hormaechei_A/modules_spots.tsv\r\n",
      "2022-11-21 12:18:47 writeFlat.py:l846 INFO\tClustering RGPs based on module content...\r\n",
      "2022-11-21 12:18:47 writeFlat.py:l881 INFO\tRGP and associated modules are listed in : data/pangenomes/Enterobacter.hormaechei_A/modules_RGP_lists.tsv\r\n",
      "2022-11-21 12:18:50 writeFlat.py:l164 INFO\tDone writing the json file : 'data/pangenomes/Enterobacter.hormaechei_A/pangenomeGraph.json'\r\n",
      "2022-11-21 12:18:50 writeFlat.py:l588 INFO\tDone writing the projection files\r\n",
      "2022-11-21 12:18:51 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Enterobacter.hormaechei_A/pangenomeGraph.gexf'\r\n",
      "2022-11-21 12:19:00 writeFlat.py:l804 INFO\tDone writing modules to organisms associations to: 'data/pangenomes/Enterobacter.hormaechei_A/modules_in_organisms.tsv'\r\n",
      "2022-11-21 12:19:00 all.py:l125 INFO\tAnnotation took : 23.34 seconds\r\n",
      "2022-11-21 12:19:00 all.py:l126 INFO\tClustering took : 45.35 seconds\r\n",
      "2022-11-21 12:19:00 all.py:l127 INFO\tBuilding the graph took : 3.31 seconds\r\n",
      "2022-11-21 12:19:00 all.py:l128 INFO\tPartitioning the pangenome took : 91.66 seconds\r\n",
      "2022-11-21 12:19:00 all.py:l129 INFO\tPredicting RGP took : 4.06 seconds\r\n",
      "2022-11-21 12:19:00 all.py:l130 INFO\tGathering RGP into spots took : 19.33 seconds\r\n",
      "2022-11-21 12:19:00 all.py:l131 INFO\tPredicting modules took : 9.72 seconds\r\n",
      "2022-11-21 12:19:00 all.py:l132 INFO\tWriting the pangenome data in HDF5 took : 9.34 seconds\r\n",
      "2022-11-21 12:19:00 all.py:l134 INFO\tWriting descriptive files for the pangenome took : 16.84 seconds\r\n",
      "Genes: 739586\r\n",
      "Organisms: 159\r\n",
      "Families: 18180\r\n",
      "Edges: 29822\r\n",
      "Persistent (min:0.84, max:1.0, sd:0.02, mean:0.99 ): 3694\r\n",
      "Shell ( min:0.05, max:0.85, sd:0.17, mean:0.19 ): 4087\r\n",
      "Cloud ( min:0.01, max:0.05, sd:0.01, mean:0.01 ): 10399\r\n",
      "Number of partitions: 7\r\n",
      "Shell S5 : 3016\r\n",
      "Shell S3 : 331\r\n",
      "Shell S4 : 225\r\n",
      "Shell S1 : 371\r\n",
      "Shell S2 : 142\r\n",
      "Shell S_ : 2\r\n",
      "RGPs: 5746\r\n",
      "Spots: 279\r\n",
      "Modules: 738\r\n",
      "Families in Modules: 5321\r\n",
      "2022-11-21 12:19:08 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin all --anno data/pangenomes/Enterobacter.kobei.list -o data/pangenomes/Enterobacter.kobei -c 6 -f\r\n",
      "2022-11-21 12:19:08 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 12:19:08 annotate.py:l448 INFO\tReading data/pangenomes/Enterobacter.kobei.list the list of organism files ...\r\n",
      "100%|███████████████████████████████████████| 141/141 [00:18<00:00,  7.56file/s]\r\n",
      "2022-11-21 12:19:28 annotate.py:l469 INFO\tgene identifiers used in the provided annotation files were unique, PPanGGOLiN will use them.\r\n",
      "2022-11-21 12:19:28 writeBinaries.py:l932 INFO\tWriting genome annotations...\r\n",
      "100%|█████████████████████████████████████| 141/141 [00:01<00:00, 84.51genome/s]\r\n",
      "2022-11-21 12:19:30 writeBinaries.py:l950 INFO\twriting the protein coding gene dna sequences\r\n",
      "100%|█████████████████████████████| 623869/623869 [00:00<00:00, 675327.69gene/s]\r\n",
      "2022-11-21 12:19:33 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.kobei/pangenome.h5\r\n",
      "2022-11-21 12:19:33 writeSequences.py:l39 INFO\tWriting all of the CDS sequences...\r\n",
      "100%|█████████████████████████████| 623869/623869 [00:00<00:00, 676431.01gene/s]\r\n",
      "2022-11-21 12:19:34 cluster.py:l297 INFO\tClustering all of the genes sequences...\r\n",
      "2022-11-21 12:19:34 cluster.py:l86 INFO\tCreating sequence database...\r\n",
      "2022-11-21 12:19:38 cluster.py:l93 INFO\tClustering sequences...\r\n",
      "2022-11-21 12:19:50 cluster.py:l99 INFO\tExtracting cluster representatives...\r\n",
      "2022-11-21 12:19:50 cluster.py:l108 INFO\tWriting gene to family informations\r\n",
      "2022-11-21 12:19:50 cluster.py:l305 INFO\tAssociating fragments to their original gene family...\r\n",
      "2022-11-21 12:19:50 cluster.py:l153 INFO\tAligning cluster representatives...\r\n",
      "2022-11-21 12:20:11 cluster.py:l159 INFO\tExtracting alignments...\r\n",
      "2022-11-21 12:20:11 cluster.py:l197 INFO\tStarting with 19301 families\r\n",
      "2022-11-21 12:20:11 cluster.py:l231 INFO\tEnding with 18170 gene families\r\n",
      "2022-11-21 12:20:12 cluster.py:l242 INFO\tAdding protein sequences to the gene families\r\n",
      "2022-11-21 12:20:12 cluster.py:l256 INFO\tAdding 623869 genes to the gene families\r\n",
      "100%|█████████████████████████████| 623869/623869 [00:02<00:00, 212466.23gene/s]\r\n",
      "2022-11-21 12:20:15 writeBinaries.py:l955 INFO\tWriting gene families and gene associations...\r\n",
      "100%|█████████████████████████| 18170/18170 [00:00<00:00, 36752.64gene family/s]\r\n",
      "2022-11-21 12:20:15 writeBinaries.py:l957 INFO\tWriting gene families information...\r\n",
      "100%|████████████████████████| 18170/18170 [00:00<00:00, 148998.22gene family/s]\r\n",
      "2022-11-21 12:20:15 writeBinaries.py:l823 INFO\tUpdating annotations with fragment information\r\n",
      "100%|█████████████████████████████| 634334/634334 [00:01<00:00, 329854.83gene/s]\r\n",
      "2022-11-21 12:20:17 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.kobei/pangenome.h5\r\n",
      "2022-11-21 12:20:17 makeGraph.py:l87 INFO\tComputing the neighbors graph...\r\n",
      "Processing GCF_902807225.1_PBIO694_genomic: 100%|█| 141/141 [00:01<00:00, 70.86o\r\n",
      "2022-11-21 12:20:19 makeGraph.py:l108 INFO\tDone making the neighbors graph.\r\n",
      "2022-11-21 12:20:19 partition.py:l482 INFO\tEstimating the optimal number of partitions...\r\n",
      "100%|█████████████████████████| 19/19 [01:01<00:00,  3.25s/Number of partitions]\r\n",
      "2022-11-21 12:21:22 partition.py:l485 INFO\tThe number of partitions has been evaluated at 5\r\n",
      "2022-11-21 12:21:22 partition.py:l503 INFO\tPartitioning...\r\n",
      "2022-11-21 12:21:35 partition.py:l578 INFO\tPartitioned 141 genomes in 13.56 seconds.\r\n",
      "2022-11-21 12:21:35 writeBinaries.py:l966 INFO\tWriting the edges...\r\n",
      "100%|████████████████████████████████| 26021/26021 [00:00<00:00, 55304.60edge/s]\r\n",
      "2022-11-21 12:21:36 writeBinaries.py:l808 INFO\tUpdating gene families with partition information\r\n",
      "100%|█████████████████████████| 18170/18170 [00:00<00:00, 86297.76gene family/s]\r\n",
      "2022-11-21 12:21:36 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.kobei/pangenome.h5\r\n",
      "2022-11-21 12:21:36 genomicIsland.py:l250 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 12:21:37 genomicIsland.py:l252 INFO\tCompute Regions of Genomic Plasticity ...\r\n",
      "100%|████████████████████████████████████| 141/141 [00:02<00:00, 64.02genomes/s]\r\n",
      "2022-11-21 12:21:39 genomicIsland.py:l257 INFO\tPredicted 4903 RGP\r\n",
      "2022-11-21 12:21:39 spot.py:l191 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 12:21:39 spot.py:l194 INFO\tDetecting hotspots in the pan...\r\n",
      "2022-11-21 12:21:39 spot.py:l118 INFO\t2476 RGPs were not used as they are on a contig border (or have less than 3 persistent gene families until the contig border)\r\n",
      "2022-11-21 12:21:39 spot.py:l120 INFO\t2427 RGPs are being used to predict spots of insertion\r\n",
      "2022-11-21 12:21:39 spot.py:l122 INFO\t248 number of different pairs of flanking gene families\r\n",
      "2022-11-21 12:21:39 spot.py:l203 INFO\t164 spots were detected\r\n",
      "2022-11-21 12:21:39 module.py:l121 INFO\tBuilding the graph...\r\n",
      "100%|█████████████████████████████████████| 141/141 [00:07<00:00, 20.05genome/s]\r\n",
      "2022-11-21 12:21:46 module.py:l123 INFO\tTook 7.03 seconds to build the graph to find modules in\r\n",
      "2022-11-21 12:21:46 module.py:l124 INFO\tThere are 18170 nodes and 143261 edges\r\n",
      "2022-11-21 12:21:47 module.py:l137 INFO\tThere are 4260 families among 639 modules\r\n",
      "2022-11-21 12:21:47 module.py:l138 INFO\tComputing modules took 0.63 seconds\r\n",
      "2022-11-21 12:21:47 writeBinaries.py:l975 INFO\tWriting Regions of Genomic Plasticity...\r\n",
      "100%|███████████████████████████████| 4903/4903 [00:00<00:00, 121884.23region/s]\r\n",
      "2022-11-21 12:21:47 writeBinaries.py:l980 INFO\tWriting Spots of Insertion...\r\n",
      "100%|███████████████████████████████████| 164/164 [00:00<00:00, 167649.49spot/s]\r\n",
      "2022-11-21 12:21:47 writeBinaries.py:l985 INFO\tWriting Modules...\r\n",
      "100%|████████████████████████████████| 639/639 [00:00<00:00, 210870.20modules/s]\r\n",
      "2022-11-21 12:21:47 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.kobei/pangenome.h5\r\n",
      "2022-11-21 12:21:47 draw_spot.py:l650 INFO\tDrawing 86 spots\r\n",
      "2022-11-21 12:21:47 draw_spot.py:l559 INFO\tOrdering genes among regions, and drawing spots...\r\n",
      "  0%|                                                  | 0/86 [00:00<?, ?spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_26.html' already exists, will be overwritten.\r\n",
      "  1%|▍                                         | 1/86 [00:00<00:09,  9.42spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_25.html' already exists, will be overwritten.\r\n",
      "  2%|▉                                         | 2/86 [00:00<00:09,  9.05spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_18.html' already exists, will be overwritten.\r\n",
      "  5%|█▉                                        | 4/86 [00:00<00:09,  8.70spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_20.html' already exists, will be overwritten.\r\n",
      "  6%|██▍                                       | 5/86 [00:00<00:09,  8.47spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_4.html' already exists, will be overwritten.\r\n",
      "  7%|██▉                                       | 6/86 [00:00<00:09,  8.50spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_17.html' already exists, will be overwritten.\r\n",
      "  8%|███▍                                      | 7/86 [00:00<00:09,  7.94spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_28.html' already exists, will be overwritten.\r\n",
      " 10%|████▍                                     | 9/86 [00:01<00:08,  8.67spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_86.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_63.html' already exists, will be overwritten.\r\n",
      " 13%|█████▏                                   | 11/86 [00:01<00:07,  9.45spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_62.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_10.html' already exists, will be overwritten.\r\n",
      " 15%|██████▏                                  | 13/86 [00:01<00:07,  9.44spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_90.html' already exists, will be overwritten.\r\n",
      " 17%|███████▏                                 | 15/86 [00:01<00:07,  9.78spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_45.html' already exists, will be overwritten.\r\n",
      " 20%|████████                                 | 17/86 [00:01<00:06,  9.92spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_120.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_24.html' already exists, will be overwritten.\r\n",
      " 22%|█████████                                | 19/86 [00:02<00:06,  9.70spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_73.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_16.html' already exists, will be overwritten.\r\n",
      " 26%|██████████▍                              | 22/86 [00:02<00:07,  9.14spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_119.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_79.html' already exists, will be overwritten.\r\n",
      " 28%|███████████▍                             | 24/86 [00:02<00:06,  9.62spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_8.html' already exists, will be overwritten.\r\n",
      " 29%|███████████▉                             | 25/86 [00:02<00:06,  9.28spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_78.html' already exists, will be overwritten.\r\n",
      " 30%|████████████▍                            | 26/86 [00:02<00:06,  9.29spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_118.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_12.html' already exists, will be overwritten.\r\n",
      " 33%|█████████████▎                           | 28/86 [00:03<00:05,  9.72spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_65.html' already exists, will be overwritten.\r\n",
      " 35%|██████████████▎                          | 30/86 [00:03<00:05, 10.11spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_27.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_13.html' already exists, will be overwritten.\r\n",
      " 37%|███████████████▎                         | 32/86 [00:03<00:05,  9.90spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_2.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_15.html' already exists, will be overwritten.\r\n",
      " 40%|████████████████▏                        | 34/86 [00:03<00:05,  9.85spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_14.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_29.html' already exists, will be overwritten.\r\n",
      " 42%|█████████████████▏                       | 36/86 [00:03<00:05,  9.93spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_72.html' already exists, will be overwritten.\r\n",
      " 44%|██████████████████                       | 38/86 [00:04<00:05,  9.34spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_6.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_21.html' already exists, will be overwritten.\r\n",
      " 47%|███████████████████                      | 40/86 [00:04<00:05,  8.66spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_3.html' already exists, will be overwritten.\r\n",
      " 48%|███████████████████▌                     | 41/86 [00:04<00:05,  8.85spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_106.html' already exists, will be overwritten.\r\n",
      " 49%|████████████████████                     | 42/86 [00:04<00:04,  9.08spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_5.html' already exists, will be overwritten.\r\n",
      " 50%|████████████████████▌                    | 43/86 [00:04<00:05,  8.22spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_11.html' already exists, will be overwritten.\r\n",
      " 51%|████████████████████▉                    | 44/86 [00:04<00:05,  8.23spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_75.html' already exists, will be overwritten.\r\n",
      " 52%|█████████████████████▍                   | 45/86 [00:04<00:04,  8.58spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_61.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_7.html' already exists, will be overwritten.\r\n",
      " 55%|██████████████████████▍                  | 47/86 [00:05<00:04,  9.05spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_0.html' already exists, will be overwritten.\r\n",
      " 56%|██████████████████████▉                  | 48/86 [00:05<00:04,  9.17spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_34.html' already exists, will be overwritten.\r\n",
      " 57%|███████████████████████▎                 | 49/86 [00:05<00:04,  8.08spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_31.html' already exists, will be overwritten.\r\n",
      " 59%|████████████████████████▎                | 51/86 [00:05<00:04,  8.13spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_1.html' already exists, will be overwritten.\r\n",
      " 60%|████████████████████████▊                | 52/86 [00:05<00:04,  8.50spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_36.html' already exists, will be overwritten.\r\n",
      " 63%|█████████████████████████▋               | 54/86 [00:05<00:03,  9.29spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_68.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_33.html' already exists, will be overwritten.\r\n",
      " 65%|██████████████████████████▋              | 56/86 [00:06<00:03,  9.64spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_66.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_40.html' already exists, will be overwritten.\r\n",
      " 67%|███████████████████████████▋             | 58/86 [00:06<00:02,  9.70spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_35.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_41.html' already exists, will be overwritten.\r\n",
      " 71%|█████████████████████████████            | 61/86 [00:06<00:02,  9.83spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_39.html' already exists, will be overwritten.\r\n",
      " 72%|█████████████████████████████▌           | 62/86 [00:06<00:02,  9.29spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_37.html' already exists, will be overwritten.\r\n",
      " 73%|██████████████████████████████           | 63/86 [00:06<00:02,  9.05spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_47.html' already exists, will be overwritten.\r\n",
      " 74%|██████████████████████████████▌          | 64/86 [00:06<00:02,  8.74spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_42.html' already exists, will be overwritten.\r\n",
      " 76%|██████████████████████████████▉          | 65/86 [00:07<00:02,  8.90spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_43.html' already exists, will be overwritten.\r\n",
      " 77%|███████████████████████████████▍         | 66/86 [00:07<00:02,  7.50spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_9.html' already exists, will be overwritten.\r\n",
      " 79%|████████████████████████████████▍        | 68/86 [00:07<00:02,  8.34spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_97.html' already exists, will be overwritten.\r\n",
      " 81%|█████████████████████████████████▎       | 70/86 [00:07<00:01,  9.01spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_48.html' already exists, will be overwritten.\r\n",
      " 83%|█████████████████████████████████▊       | 71/86 [00:07<00:01,  8.99spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_50.html' already exists, will be overwritten.\r\n",
      " 84%|██████████████████████████████████▎      | 72/86 [00:07<00:01,  9.14spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_49.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_96.html' already exists, will be overwritten.\r\n",
      " 86%|███████████████████████████████████▎     | 74/86 [00:08<00:01,  9.49spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_46.html' already exists, will be overwritten.\r\n",
      " 87%|███████████████████████████████████▊     | 75/86 [00:08<00:01,  9.22spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_51.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_59.html' already exists, will be overwritten.\r\n",
      " 90%|████████████████████████████████████▋    | 77/86 [00:08<00:00,  9.56spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_98.html' already exists, will be overwritten.\r\n",
      " 91%|█████████████████████████████████████▏   | 78/86 [00:08<00:00,  9.64spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_52.html' already exists, will be overwritten.\r\n",
      " 93%|██████████████████████████████████████▏  | 80/86 [00:08<00:00,  9.98spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_55.html' already exists, will be overwritten.\r\n",
      " 94%|██████████████████████████████████████▌  | 81/86 [00:08<00:00,  9.68spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_87.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_23.html' already exists, will be overwritten.\r\n",
      " 97%|███████████████████████████████████████▌ | 83/86 [00:09<00:00,  8.92spot/s]INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_94.html' already exists, will be overwritten.\r\n",
      "INFO:bokeh.io.state:Session output file 'data/pangenomes/Enterobacter.kobei/spot_figures/spot_54.html' already exists, will be overwritten.\r\n",
      "100%|█████████████████████████████████████████| 86/86 [00:09<00:00,  9.18spot/s]\r\n",
      "2022-11-21 12:21:57 draw_spot.py:l617 INFO\tDone drawing spot(s), they can be found in the directory: 'data/pangenomes/Enterobacter.kobei/spot_figures'\r\n",
      "2022-11-21 12:21:57 tile_plot.py:l38 INFO\tDrawing the tile plot...\r\n",
      "2022-11-21 12:21:57 tile_plot.py:l56 INFO\tstart with matrice\r\n",
      "2022-11-21 12:21:57 tile_plot.py:l72 INFO\tdone with making the dendrogram to order the organisms on the plot\r\n",
      "2022-11-21 12:21:58 tile_plot.py:l106 INFO\tGetting the gene name(s) and the number for each tile of the plot ...\r\n",
      "2022-11-21 12:21:59 tile_plot.py:l116 INFO\tDone extracting names and numbers. Making the heatmap ...\r\n",
      "2022-11-21 12:22:04 tile_plot.py:l173 INFO\tDrawing the figure itself...\r\n",
      "2022-11-21 12:22:29 tile_plot.py:l175 INFO\tDone with the tile plot : 'data/pangenomes/Enterobacter.kobei/tile_plot.html' \r\n",
      "2022-11-21 12:22:29 ucurve.py:l23 INFO\tDrawing the U-shaped curve...\r\n",
      "2022-11-21 12:22:29 ucurve.py:l77 INFO\tDone drawing the U-shaped curve : 'data/pangenomes/Enterobacter.kobei/Ushaped_plot.html'\r\n",
      "2022-11-21 12:22:29 writeFlat.py:l328 INFO\tWriting the .csv file ...\r\n",
      "2022-11-21 12:22:29 writeFlat.py:l399 INFO\tWriting the gene presence absence file ...\r\n",
      "2022-11-21 12:22:29 writeFlat.py:l305 INFO\tWriting the gexf file for the pangenome graph...\r\n",
      "2022-11-21 12:22:29 writeFlat.py:l305 INFO\tWriting the light gexf file for the pangenome graph...\r\n",
      "2022-11-21 12:22:29 writeFlat.py:l431 INFO\tWriting pangenome statistics...\r\n",
      "2022-11-21 12:22:29 writeFlat.py:l582 INFO\tWriting the projection files...\r\n",
      "2022-11-21 12:22:29 writeFlat.py:l432 INFO\tWriting statistics on persistent duplication...\r\n",
      "2022-11-21 12:22:29 writeFlat.py:l454 INFO\tDone writing stats on persistent duplication\r\n",
      "2022-11-21 12:22:29 writeFlat.py:l455 INFO\tWriting genome per genome statistics (completeness and counts)...\r\n",
      "2022-11-21 12:22:29 writeFlat.py:l419 INFO\tDone writing the gene presence absence file : 'data/pangenomes/Enterobacter.kobei/gene_presence_absence.Rtab'\r\n",
      "2022-11-21 12:22:29 writeFlat.py:l157 INFO\tWriting the json file for the pangenome graph...\r\n",
      "2022-11-21 12:22:31 writeFlat.py:l520 INFO\tDone writing genome per genome statistics\r\n",
      "2022-11-21 12:22:31 writeFlat.py:l598 INFO\tWriting the list of gene families for each partition ...\r\n",
      "2022-11-21 12:22:31 writeFlat.py:l625 INFO\tDone writing the list of gene families for each partition\r\n",
      "2022-11-21 12:22:31 writeFlat.py:l691 INFO\tDone writing spots in : 'data/pangenomes/Enterobacter.kobei/summarize_spots.tsv'\r\n",
      "2022-11-21 12:22:31 writeFlat.py:l776 INFO\tWriting functional modules...\r\n",
      "2022-11-21 12:22:31 writeFlat.py:l784 INFO\tDone writing functional modules to: 'data/pangenomes/Enterobacter.kobei/functional_modules.tsv'\r\n",
      "2022-11-21 12:22:31 writeFlat.py:l752 INFO\tWriting functional modules summary...\r\n",
      "2022-11-21 12:22:31 writeFlat.py:l767 INFO\tDone writing module summary: 'data/pangenomes/Enterobacter.kobei/modules_summary.tsv'\r\n",
      "2022-11-21 12:22:31 writeFlat.py:l793 INFO\tWriting modules to organisms associations...\r\n",
      "2022-11-21 12:22:32 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Enterobacter.kobei/pangenomeGraph_light.gexf'\r\n",
      "2022-11-21 12:22:32 writeFlat.py:l814 INFO\tWriting modules to spot associations...\r\n",
      "2022-11-21 12:22:32 writeFlat.py:l837 INFO\tDone writing module to spot associations to: data/pangenomes/Enterobacter.kobei/modules_spots.tsv\r\n",
      "2022-11-21 12:22:32 writeFlat.py:l846 INFO\tClustering RGPs based on module content...\r\n",
      "2022-11-21 12:22:32 writeFlat.py:l881 INFO\tRGP and associated modules are listed in : data/pangenomes/Enterobacter.kobei/modules_RGP_lists.tsv\r\n",
      "2022-11-21 12:22:32 writeFlat.py:l389 INFO\tDone writing the matrix : 'data/pangenomes/Enterobacter.kobei/matrix.csv'\r\n",
      "2022-11-21 12:22:34 writeFlat.py:l588 INFO\tDone writing the projection files\r\n",
      "2022-11-21 12:22:35 writeFlat.py:l164 INFO\tDone writing the json file : 'data/pangenomes/Enterobacter.kobei/pangenomeGraph.json'\r\n",
      "2022-11-21 12:22:35 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Enterobacter.kobei/pangenomeGraph.gexf'\r\n",
      "2022-11-21 12:22:41 writeFlat.py:l804 INFO\tDone writing modules to organisms associations to: 'data/pangenomes/Enterobacter.kobei/modules_in_organisms.tsv'\r\n",
      "2022-11-21 12:22:41 all.py:l125 INFO\tAnnotation took : 19.85 seconds\r\n",
      "2022-11-21 12:22:41 all.py:l126 INFO\tClustering took : 41.1 seconds\r\n",
      "2022-11-21 12:22:41 all.py:l127 INFO\tBuilding the graph took : 1.99 seconds\r\n",
      "2022-11-21 12:22:41 all.py:l128 INFO\tPartitioning the pangenome took : 75.87 seconds\r\n",
      "2022-11-21 12:22:41 all.py:l129 INFO\tPredicting RGP took : 2.58 seconds\r\n",
      "2022-11-21 12:22:41 all.py:l130 INFO\tGathering RGP into spots took : 10.37 seconds\r\n",
      "2022-11-21 12:22:41 all.py:l131 INFO\tPredicting modules took : 7.68 seconds\r\n",
      "2022-11-21 12:22:41 all.py:l132 INFO\tWriting the pangenome data in HDF5 took : 7.03 seconds\r\n",
      "2022-11-21 12:22:41 all.py:l134 INFO\tWriting descriptive files for the pangenome took : 12.15 seconds\r\n",
      "Genes: 623869\r\n",
      "Organisms: 141\r\n",
      "Families: 18170\r\n",
      "Edges: 26021\r\n",
      "Persistent (min:0.82, max:0.96, sd:0.02, mean:0.95 ): 3821\r\n",
      "Shell ( min:0.06, max:0.85, sd:0.18, mean:0.21 ): 2569\r\n",
      "Cloud ( min:0.01, max:0.06, sd:0.01, mean:0.02 ): 11780\r\n",
      "Number of partitions: 5\r\n",
      "Shell S2 : 152\r\n",
      "Shell S3 : 2128\r\n",
      "Shell S1 : 289\r\n",
      "RGPs: 4903\r\n",
      "Spots: 164\r\n",
      "Modules: 639\r\n",
      "Families in Modules: 4260\r\n",
      "2022-11-21 12:22:48 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin all --anno data/pangenomes/Enterobacter.roggenkampii.list -o data/pangenomes/Enterobacter.roggenkampii -c 6 -f\r\n",
      "2022-11-21 12:22:48 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 12:22:48 annotate.py:l448 INFO\tReading data/pangenomes/Enterobacter.roggenkampii.list the list of organism files ...\r\n",
      "100%|███████████████████████████████████████| 197/197 [00:26<00:00,  7.36file/s]\r\n",
      "2022-11-21 12:23:16 annotate.py:l469 INFO\tgene identifiers used in the provided annotation files were unique, PPanGGOLiN will use them.\r\n",
      "2022-11-21 12:23:16 writeBinaries.py:l932 INFO\tWriting genome annotations...\r\n",
      "100%|█████████████████████████████████████| 197/197 [00:02<00:00, 79.77genome/s]\r\n",
      "2022-11-21 12:23:19 writeBinaries.py:l950 INFO\twriting the protein coding gene dna sequences\r\n",
      "100%|█████████████████████████████| 896840/896840 [00:01<00:00, 636039.17gene/s]\r\n",
      "2022-11-21 12:23:27 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.roggenkampii/pangenome.h5\r\n",
      "2022-11-21 12:23:27 writeSequences.py:l39 INFO\tWriting all of the CDS sequences...\r\n",
      "100%|█████████████████████████████| 896840/896840 [00:01<00:00, 684607.85gene/s]\r\n",
      "2022-11-21 12:23:28 cluster.py:l297 INFO\tClustering all of the genes sequences...\r\n",
      "2022-11-21 12:23:28 cluster.py:l86 INFO\tCreating sequence database...\r\n",
      "2022-11-21 12:23:33 cluster.py:l93 INFO\tClustering sequences...\r\n",
      "2022-11-21 12:23:51 cluster.py:l99 INFO\tExtracting cluster representatives...\r\n",
      "2022-11-21 12:23:52 cluster.py:l108 INFO\tWriting gene to family informations\r\n",
      "2022-11-21 12:23:52 cluster.py:l305 INFO\tAssociating fragments to their original gene family...\r\n",
      "2022-11-21 12:23:52 cluster.py:l153 INFO\tAligning cluster representatives...\r\n",
      "2022-11-21 12:24:22 cluster.py:l159 INFO\tExtracting alignments...\r\n",
      "2022-11-21 12:24:23 cluster.py:l197 INFO\tStarting with 25196 families\r\n",
      "2022-11-21 12:24:23 cluster.py:l231 INFO\tEnding with 23062 gene families\r\n",
      "2022-11-21 12:24:23 cluster.py:l242 INFO\tAdding protein sequences to the gene families\r\n",
      "2022-11-21 12:24:23 cluster.py:l256 INFO\tAdding 896840 genes to the gene families\r\n",
      "100%|█████████████████████████████| 896840/896840 [00:03<00:00, 288748.25gene/s]\r\n",
      "2022-11-21 12:24:27 writeBinaries.py:l955 INFO\tWriting gene families and gene associations...\r\n",
      "100%|█████████████████████████| 23062/23062 [00:00<00:00, 30960.48gene family/s]\r\n",
      "2022-11-21 12:24:28 writeBinaries.py:l957 INFO\tWriting gene families information...\r\n",
      "100%|████████████████████████| 23062/23062 [00:00<00:00, 142729.88gene family/s]\r\n",
      "2022-11-21 12:24:28 writeBinaries.py:l823 INFO\tUpdating annotations with fragment information\r\n",
      "100%|█████████████████████████████| 912560/912560 [00:02<00:00, 312769.40gene/s]\r\n",
      "2022-11-21 12:24:31 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.roggenkampii/pangenome.h5\r\n",
      "2022-11-21 12:24:31 makeGraph.py:l87 INFO\tComputing the neighbors graph...\r\n",
      "Processing GCF_902387855.1_UHGG_MGYG-HGUT-02499_genomic: 100%|█| 197/197 [00:04<\r\n",
      "2022-11-21 12:24:35 makeGraph.py:l108 INFO\tDone making the neighbors graph.\r\n",
      "2022-11-21 12:24:35 partition.py:l482 INFO\tEstimating the optimal number of partitions...\r\n",
      "100%|█████████████████████████| 19/19 [02:04<00:00,  6.56s/Number of partitions]\r\n",
      "2022-11-21 12:26:40 partition.py:l485 INFO\tThe number of partitions has been evaluated at 5\r\n",
      "2022-11-21 12:26:40 partition.py:l503 INFO\tPartitioning...\r\n",
      "2022-11-21 12:27:09 partition.py:l578 INFO\tPartitioned 197 genomes in 28.45 seconds.\r\n",
      "2022-11-21 12:27:09 writeBinaries.py:l966 INFO\tWriting the edges...\r\n",
      "100%|████████████████████████████████| 36129/36129 [00:00<00:00, 44074.00edge/s]\r\n",
      "2022-11-21 12:27:10 writeBinaries.py:l808 INFO\tUpdating gene families with partition information\r\n",
      "100%|█████████████████████████| 23062/23062 [00:00<00:00, 84304.50gene family/s]\r\n",
      "2022-11-21 12:27:10 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.roggenkampii/pangenome.h5\r\n",
      "2022-11-21 12:27:10 genomicIsland.py:l250 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 12:27:11 genomicIsland.py:l252 INFO\tCompute Regions of Genomic Plasticity ...\r\n",
      "100%|████████████████████████████████████| 197/197 [00:02<00:00, 98.37genomes/s]\r\n",
      "2022-11-21 12:27:13 genomicIsland.py:l257 INFO\tPredicted 10537 RGP\r\n",
      "2022-11-21 12:27:13 spot.py:l191 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 12:27:13 spot.py:l194 INFO\tDetecting hotspots in the pan...\r\n",
      "2022-11-21 12:27:13 spot.py:l118 INFO\t4708 RGPs were not used as they are on a contig border (or have less than 3 persistent gene families until the contig border)\r\n",
      "2022-11-21 12:27:13 spot.py:l120 INFO\t5829 RGPs are being used to predict spots of insertion\r\n",
      "2022-11-21 12:27:13 spot.py:l122 INFO\t469 number of different pairs of flanking gene families\r\n",
      "2022-11-21 12:27:14 spot.py:l203 INFO\t267 spots were detected\r\n",
      "2022-11-21 12:27:14 module.py:l121 INFO\tBuilding the graph...\r\n",
      "100%|█████████████████████████████████████| 197/197 [00:11<00:00, 17.25genome/s]\r\n",
      "2022-11-21 12:27:25 module.py:l123 INFO\tTook 11.42 seconds to build the graph to find modules in\r\n",
      "2022-11-21 12:27:25 module.py:l124 INFO\tThere are 23062 nodes and 199707 edges\r\n",
      "2022-11-21 12:27:26 module.py:l137 INFO\tThere are 6395 families among 964 modules\r\n",
      "2022-11-21 12:27:26 module.py:l138 INFO\tComputing modules took 0.88 seconds\r\n",
      "2022-11-21 12:27:26 writeBinaries.py:l975 INFO\tWriting Regions of Genomic Plasticity...\r\n",
      "100%|█████████████████████████████| 10537/10537 [00:00<00:00, 117930.14region/s]\r\n",
      "2022-11-21 12:27:26 writeBinaries.py:l980 INFO\tWriting Spots of Insertion...\r\n",
      "100%|███████████████████████████████████| 267/267 [00:00<00:00, 134941.46spot/s]\r\n",
      "2022-11-21 12:27:26 writeBinaries.py:l985 INFO\tWriting Modules...\r\n",
      "100%|████████████████████████████████| 964/964 [00:00<00:00, 265378.65modules/s]\r\n",
      "2022-11-21 12:27:27 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterobacter.roggenkampii/pangenome.h5\r\n",
      "2022-11-21 12:27:27 draw_spot.py:l650 INFO\tDrawing 129 spots\r\n",
      "2022-11-21 12:27:27 draw_spot.py:l559 INFO\tOrdering genes among regions, and drawing spots...\r\n",
      "100%|███████████████████████████████████████| 129/129 [00:14<00:00,  9.08spot/s]\r\n",
      "2022-11-21 12:27:42 draw_spot.py:l617 INFO\tDone drawing spot(s), they can be found in the directory: 'data/pangenomes/Enterobacter.roggenkampii/spot_figures'\r\n",
      "2022-11-21 12:27:42 tile_plot.py:l38 INFO\tDrawing the tile plot...\r\n",
      "2022-11-21 12:27:42 tile_plot.py:l56 INFO\tstart with matrice\r\n",
      "2022-11-21 12:27:42 tile_plot.py:l72 INFO\tdone with making the dendrogram to order the organisms on the plot\r\n",
      "2022-11-21 12:27:42 tile_plot.py:l106 INFO\tGetting the gene name(s) and the number for each tile of the plot ...\r\n",
      "2022-11-21 12:27:44 tile_plot.py:l116 INFO\tDone extracting names and numbers. Making the heatmap ...\r\n",
      "2022-11-21 12:27:53 tile_plot.py:l173 INFO\tDrawing the figure itself...\r\n",
      "2022-11-21 12:28:34 tile_plot.py:l175 INFO\tDone with the tile plot : 'data/pangenomes/Enterobacter.roggenkampii/tile_plot.html' \r\n",
      "2022-11-21 12:28:34 ucurve.py:l23 INFO\tDrawing the U-shaped curve...\r\n",
      "2022-11-21 12:28:34 ucurve.py:l77 INFO\tDone drawing the U-shaped curve : 'data/pangenomes/Enterobacter.roggenkampii/Ushaped_plot.html'\r\n",
      "2022-11-21 12:28:35 writeFlat.py:l328 INFO\tWriting the .csv file ...\r\n",
      "2022-11-21 12:28:35 writeFlat.py:l305 INFO\tWriting the gexf file for the pangenome graph...\r\n",
      "2022-11-21 12:28:35 writeFlat.py:l399 INFO\tWriting the gene presence absence file ...\r\n",
      "2022-11-21 12:28:35 writeFlat.py:l305 INFO\tWriting the light gexf file for the pangenome graph...\r\n",
      "2022-11-21 12:28:35 writeFlat.py:l431 INFO\tWriting pangenome statistics...\r\n",
      "2022-11-21 12:28:35 writeFlat.py:l582 INFO\tWriting the projection files...\r\n",
      "2022-11-21 12:28:35 writeFlat.py:l432 INFO\tWriting statistics on persistent duplication...\r\n",
      "2022-11-21 12:28:35 writeFlat.py:l454 INFO\tDone writing stats on persistent duplication\r\n",
      "2022-11-21 12:28:35 writeFlat.py:l455 INFO\tWriting genome per genome statistics (completeness and counts)...\r\n",
      "2022-11-21 12:28:35 writeFlat.py:l419 INFO\tDone writing the gene presence absence file : 'data/pangenomes/Enterobacter.roggenkampii/gene_presence_absence.Rtab'\r\n",
      "2022-11-21 12:28:35 writeFlat.py:l157 INFO\tWriting the json file for the pangenome graph...\r\n",
      "2022-11-21 12:28:37 writeFlat.py:l520 INFO\tDone writing genome per genome statistics\r\n",
      "2022-11-21 12:28:37 writeFlat.py:l598 INFO\tWriting the list of gene families for each partition ...\r\n",
      "2022-11-21 12:28:37 writeFlat.py:l625 INFO\tDone writing the list of gene families for each partition\r\n",
      "2022-11-21 12:28:38 writeFlat.py:l691 INFO\tDone writing spots in : 'data/pangenomes/Enterobacter.roggenkampii/summarize_spots.tsv'\r\n",
      "2022-11-21 12:28:38 writeFlat.py:l776 INFO\tWriting functional modules...\r\n",
      "2022-11-21 12:28:38 writeFlat.py:l784 INFO\tDone writing functional modules to: 'data/pangenomes/Enterobacter.roggenkampii/functional_modules.tsv'\r\n",
      "2022-11-21 12:28:38 writeFlat.py:l752 INFO\tWriting functional modules summary...\r\n",
      "2022-11-21 12:28:38 writeFlat.py:l767 INFO\tDone writing module summary: 'data/pangenomes/Enterobacter.roggenkampii/modules_summary.tsv'\r\n",
      "2022-11-21 12:28:38 writeFlat.py:l793 INFO\tWriting modules to organisms associations...\r\n",
      "2022-11-21 12:28:39 writeFlat.py:l389 INFO\tDone writing the matrix : 'data/pangenomes/Enterobacter.roggenkampii/matrix.csv'\r\n",
      "2022-11-21 12:28:39 writeFlat.py:l814 INFO\tWriting modules to spot associations...\r\n",
      "2022-11-21 12:28:39 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Enterobacter.roggenkampii/pangenomeGraph_light.gexf'\r\n",
      "2022-11-21 12:28:39 writeFlat.py:l846 INFO\tClustering RGPs based on module content...\r\n",
      "2022-11-21 12:28:40 writeFlat.py:l881 INFO\tRGP and associated modules are listed in : data/pangenomes/Enterobacter.roggenkampii/modules_RGP_lists.tsv\r\n",
      "2022-11-21 12:28:40 writeFlat.py:l837 INFO\tDone writing module to spot associations to: data/pangenomes/Enterobacter.roggenkampii/modules_spots.tsv\r\n",
      "2022-11-21 12:28:43 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Enterobacter.roggenkampii/pangenomeGraph.gexf'\r\n",
      "2022-11-21 12:28:43 writeFlat.py:l164 INFO\tDone writing the json file : 'data/pangenomes/Enterobacter.roggenkampii/pangenomeGraph.json'\r\n",
      "2022-11-21 12:28:43 writeFlat.py:l588 INFO\tDone writing the projection files\r\n",
      "2022-11-21 12:28:54 writeFlat.py:l804 INFO\tDone writing modules to organisms associations to: 'data/pangenomes/Enterobacter.roggenkampii/modules_in_organisms.tsv'\r\n",
      "2022-11-21 12:28:54 all.py:l125 INFO\tAnnotation took : 28.5 seconds\r\n",
      "2022-11-21 12:28:54 all.py:l126 INFO\tClustering took : 59.81 seconds\r\n",
      "2022-11-21 12:28:54 all.py:l127 INFO\tBuilding the graph took : 4.1 seconds\r\n",
      "2022-11-21 12:28:54 all.py:l128 INFO\tPartitioning the pangenome took : 153.81 seconds\r\n",
      "2022-11-21 12:28:54 all.py:l129 INFO\tPredicting RGP took : 2.51 seconds\r\n",
      "2022-11-21 12:28:54 all.py:l130 INFO\tGathering RGP into spots took : 16.21 seconds\r\n",
      "2022-11-21 12:28:54 all.py:l131 INFO\tPredicting modules took : 12.32 seconds\r\n",
      "2022-11-21 12:28:54 all.py:l132 INFO\tWriting the pangenome data in HDF5 took : 12.3 seconds\r\n",
      "2022-11-21 12:28:54 all.py:l134 INFO\tWriting descriptive files for the pangenome took : 19.9 seconds\r\n",
      "Genes: 896840\r\n",
      "Organisms: 197\r\n",
      "Families: 23062\r\n",
      "Edges: 36129\r\n",
      "Persistent (min:0.87, max:0.99, sd:0.02, mean:0.98 ): 3603\r\n",
      "Shell ( min:0.04, max:0.94, sd:0.21, mean:0.18 ): 4424\r\n",
      "Cloud ( min:0.01, max:0.04, sd:0.01, mean:0.01 ): 15035\r\n",
      "Number of partitions: 5\r\n",
      "Shell S2 : 905\r\n",
      "Shell S3 : 3164\r\n",
      "Shell S1 : 355\r\n",
      "RGPs: 10537\r\n",
      "Spots: 267\r\n",
      "Modules: 964\r\n",
      "Families in Modules: 6395\r\n",
      "2022-11-21 12:29:04 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin all --anno data/pangenomes/Enterococcus_B.faecium.list -o data/pangenomes/Enterococcus_B.faecium -c 6 -f\r\n",
      "2022-11-21 12:29:04 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 12:29:04 annotate.py:l448 INFO\tReading data/pangenomes/Enterococcus_B.faecium.list the list of organism files ...\r\n",
      "100%|███████████████████████████████████████| 207/207 [00:15<00:00, 12.99file/s]\r\n",
      "2022-11-21 12:29:21 annotate.py:l469 INFO\tgene identifiers used in the provided annotation files were unique, PPanGGOLiN will use them.\r\n",
      "2022-11-21 12:29:21 writeBinaries.py:l932 INFO\tWriting genome annotations...\r\n",
      "100%|████████████████████████████████████| 207/207 [00:01<00:00, 134.73genome/s]\r\n",
      "2022-11-21 12:29:23 writeBinaries.py:l950 INFO\twriting the protein coding gene dna sequences\r\n",
      "100%|█████████████████████████████| 570393/570393 [00:00<00:00, 691971.14gene/s]\r\n",
      "2022-11-21 12:29:25 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterococcus_B.faecium/pangenome.h5\r\n",
      "2022-11-21 12:29:25 writeSequences.py:l39 INFO\tWriting all of the CDS sequences...\r\n",
      "100%|█████████████████████████████| 570393/570393 [00:00<00:00, 706775.65gene/s]\r\n",
      "2022-11-21 12:29:25 cluster.py:l297 INFO\tClustering all of the genes sequences...\r\n",
      "2022-11-21 12:29:25 cluster.py:l86 INFO\tCreating sequence database...\r\n",
      "2022-11-21 12:29:28 cluster.py:l93 INFO\tClustering sequences...\r\n",
      "2022-11-21 12:29:38 cluster.py:l99 INFO\tExtracting cluster representatives...\r\n",
      "2022-11-21 12:29:38 cluster.py:l108 INFO\tWriting gene to family informations\r\n",
      "2022-11-21 12:29:38 cluster.py:l305 INFO\tAssociating fragments to their original gene family...\r\n",
      "2022-11-21 12:29:38 cluster.py:l153 INFO\tAligning cluster representatives...\r\n",
      "2022-11-21 12:29:47 cluster.py:l159 INFO\tExtracting alignments...\r\n",
      "2022-11-21 12:29:47 cluster.py:l197 INFO\tStarting with 9113 families\r\n",
      "2022-11-21 12:29:47 cluster.py:l231 INFO\tEnding with 7889 gene families\r\n",
      "2022-11-21 12:29:48 cluster.py:l242 INFO\tAdding protein sequences to the gene families\r\n",
      "2022-11-21 12:29:48 cluster.py:l256 INFO\tAdding 570393 genes to the gene families\r\n",
      "100%|█████████████████████████████| 570393/570393 [00:02<00:00, 216370.28gene/s]\r\n",
      "2022-11-21 12:29:50 writeBinaries.py:l955 INFO\tWriting gene families and gene associations...\r\n",
      "100%|███████████████████████████| 7889/7889 [00:00<00:00, 17460.54gene family/s]\r\n",
      "2022-11-21 12:29:51 writeBinaries.py:l957 INFO\tWriting gene families information...\r\n",
      "100%|██████████████████████████| 7889/7889 [00:00<00:00, 244872.34gene family/s]\r\n",
      "2022-11-21 12:29:51 writeBinaries.py:l823 INFO\tUpdating annotations with fragment information\r\n",
      "100%|█████████████████████████████| 588226/588226 [00:01<00:00, 327139.96gene/s]\r\n",
      "2022-11-21 12:29:53 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterococcus_B.faecium/pangenome.h5\r\n",
      "2022-11-21 12:29:53 makeGraph.py:l87 INFO\tComputing the neighbors graph...\r\n",
      "Processing GCF_907177285.1_USZ_VRE53_P46_hybrid_assembly_genomic: 100%|█| 207/20\r\n",
      "2022-11-21 12:29:55 makeGraph.py:l108 INFO\tDone making the neighbors graph.\r\n",
      "2022-11-21 12:29:55 partition.py:l482 INFO\tEstimating the optimal number of partitions...\r\n",
      "100%|█████████████████████████| 19/19 [00:27<00:00,  1.44s/Number of partitions]\r\n",
      "2022-11-21 12:30:22 partition.py:l485 INFO\tThe number of partitions has been evaluated at 14\r\n",
      "2022-11-21 12:30:22 partition.py:l503 INFO\tPartitioning...\r\n",
      "2022-11-21 12:30:42 partition.py:l578 INFO\tPartitioned 207 genomes in 19.75 seconds.\r\n",
      "2022-11-21 12:30:42 writeBinaries.py:l966 INFO\tWriting the edges...\r\n",
      "100%|████████████████████████████████| 18642/18642 [00:00<00:00, 37124.27edge/s]\r\n",
      "2022-11-21 12:30:43 writeBinaries.py:l808 INFO\tUpdating gene families with partition information\r\n",
      "100%|██████████████████████████| 7889/7889 [00:00<00:00, 155711.57gene family/s]\r\n",
      "2022-11-21 12:30:43 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterococcus_B.faecium/pangenome.h5\r\n",
      "2022-11-21 12:30:43 genomicIsland.py:l250 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 12:30:43 genomicIsland.py:l252 INFO\tCompute Regions of Genomic Plasticity ...\r\n",
      "100%|████████████████████████████████████| 207/207 [00:02<00:00, 82.85genomes/s]\r\n",
      "2022-11-21 12:30:46 genomicIsland.py:l257 INFO\tPredicted 5939 RGP\r\n",
      "2022-11-21 12:30:46 spot.py:l191 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 12:30:46 spot.py:l194 INFO\tDetecting hotspots in the pan...\r\n",
      "2022-11-21 12:30:46 spot.py:l118 INFO\t803 RGPs were not used as they are on a contig border (or have less than 3 persistent gene families until the contig border)\r\n",
      "2022-11-21 12:30:46 spot.py:l120 INFO\t5136 RGPs are being used to predict spots of insertion\r\n",
      "2022-11-21 12:30:46 spot.py:l122 INFO\t361 number of different pairs of flanking gene families\r\n",
      "2022-11-21 12:30:46 spot.py:l203 INFO\t185 spots were detected\r\n",
      "2022-11-21 12:30:46 module.py:l121 INFO\tBuilding the graph...\r\n",
      "100%|█████████████████████████████████████| 207/207 [00:06<00:00, 30.50genome/s]\r\n",
      "2022-11-21 12:30:53 module.py:l123 INFO\tTook 6.79 seconds to build the graph to find modules in\r\n",
      "2022-11-21 12:30:53 module.py:l124 INFO\tThere are 7889 nodes and 96663 edges\r\n",
      "2022-11-21 12:30:54 module.py:l137 INFO\tThere are 2391 families among 317 modules\r\n",
      "2022-11-21 12:30:54 module.py:l138 INFO\tComputing modules took 0.51 seconds\r\n",
      "2022-11-21 12:30:54 writeBinaries.py:l975 INFO\tWriting Regions of Genomic Plasticity...\r\n",
      "100%|████████████████████████████████| 5939/5939 [00:00<00:00, 75280.59region/s]\r\n",
      "2022-11-21 12:30:54 writeBinaries.py:l980 INFO\tWriting Spots of Insertion...\r\n",
      "100%|████████████████████████████████████| 185/185 [00:00<00:00, 86408.27spot/s]\r\n",
      "2022-11-21 12:30:54 writeBinaries.py:l985 INFO\tWriting Modules...\r\n",
      "100%|████████████████████████████████| 317/317 [00:00<00:00, 252630.51modules/s]\r\n",
      "2022-11-21 12:30:54 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Enterococcus_B.faecium/pangenome.h5\r\n",
      "2022-11-21 12:30:55 draw_spot.py:l650 INFO\tDrawing 84 spots\r\n",
      "2022-11-21 12:30:55 draw_spot.py:l559 INFO\tOrdering genes among regions, and drawing spots...\r\n",
      "100%|█████████████████████████████████████████| 84/84 [00:10<00:00,  8.04spot/s]\r\n",
      "2022-11-21 12:31:06 draw_spot.py:l617 INFO\tDone drawing spot(s), they can be found in the directory: 'data/pangenomes/Enterococcus_B.faecium/spot_figures'\r\n",
      "2022-11-21 12:31:06 tile_plot.py:l38 INFO\tDrawing the tile plot...\r\n",
      "2022-11-21 12:31:06 tile_plot.py:l56 INFO\tstart with matrice\r\n",
      "2022-11-21 12:31:06 tile_plot.py:l72 INFO\tdone with making the dendrogram to order the organisms on the plot\r\n",
      "2022-11-21 12:31:06 tile_plot.py:l106 INFO\tGetting the gene name(s) and the number for each tile of the plot ...\r\n",
      "2022-11-21 12:31:07 tile_plot.py:l116 INFO\tDone extracting names and numbers. Making the heatmap ...\r\n",
      "2022-11-21 12:31:10 tile_plot.py:l173 INFO\tDrawing the figure itself...\r\n",
      "2022-11-21 12:31:26 tile_plot.py:l175 INFO\tDone with the tile plot : 'data/pangenomes/Enterococcus_B.faecium/tile_plot.html' \r\n",
      "2022-11-21 12:31:26 ucurve.py:l23 INFO\tDrawing the U-shaped curve...\r\n",
      "2022-11-21 12:31:26 ucurve.py:l77 INFO\tDone drawing the U-shaped curve : 'data/pangenomes/Enterococcus_B.faecium/Ushaped_plot.html'\r\n",
      "2022-11-21 12:31:26 writeFlat.py:l328 INFO\tWriting the .csv file ...\r\n",
      "2022-11-21 12:31:26 writeFlat.py:l399 INFO\tWriting the gene presence absence file ...\r\n",
      "2022-11-21 12:31:26 writeFlat.py:l305 INFO\tWriting the gexf file for the pangenome graph...\r\n",
      "2022-11-21 12:31:26 writeFlat.py:l305 INFO\tWriting the light gexf file for the pangenome graph...\r\n",
      "2022-11-21 12:31:26 writeFlat.py:l582 INFO\tWriting the projection files...\r\n",
      "2022-11-21 12:31:26 writeFlat.py:l431 INFO\tWriting pangenome statistics...\r\n",
      "2022-11-21 12:31:26 writeFlat.py:l432 INFO\tWriting statistics on persistent duplication...\r\n",
      "2022-11-21 12:31:26 writeFlat.py:l454 INFO\tDone writing stats on persistent duplication\r\n",
      "2022-11-21 12:31:26 writeFlat.py:l455 INFO\tWriting genome per genome statistics (completeness and counts)...\r\n",
      "2022-11-21 12:31:26 writeFlat.py:l419 INFO\tDone writing the gene presence absence file : 'data/pangenomes/Enterococcus_B.faecium/gene_presence_absence.Rtab'\r\n",
      "2022-11-21 12:31:26 writeFlat.py:l157 INFO\tWriting the json file for the pangenome graph...\r\n",
      "2022-11-21 12:31:28 writeFlat.py:l389 INFO\tDone writing the matrix : 'data/pangenomes/Enterococcus_B.faecium/matrix.csv'\r\n",
      "2022-11-21 12:31:28 writeFlat.py:l598 INFO\tWriting the list of gene families for each partition ...\r\n",
      "2022-11-21 12:31:28 writeFlat.py:l520 INFO\tDone writing genome per genome statistics\r\n",
      "2022-11-21 12:31:28 writeFlat.py:l625 INFO\tDone writing the list of gene families for each partition\r\n",
      "2022-11-21 12:31:28 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Enterococcus_B.faecium/pangenomeGraph_light.gexf'\r\n",
      "2022-11-21 12:31:28 writeFlat.py:l776 INFO\tWriting functional modules...\r\n",
      "2022-11-21 12:31:28 writeFlat.py:l784 INFO\tDone writing functional modules to: 'data/pangenomes/Enterococcus_B.faecium/functional_modules.tsv'\r\n",
      "2022-11-21 12:31:28 writeFlat.py:l752 INFO\tWriting functional modules summary...\r\n",
      "2022-11-21 12:31:29 writeFlat.py:l767 INFO\tDone writing module summary: 'data/pangenomes/Enterococcus_B.faecium/modules_summary.tsv'\r\n",
      "2022-11-21 12:31:29 writeFlat.py:l793 INFO\tWriting modules to organisms associations...\r\n",
      "2022-11-21 12:31:29 writeFlat.py:l814 INFO\tWriting modules to spot associations...\r\n",
      "2022-11-21 12:31:29 writeFlat.py:l691 INFO\tDone writing spots in : 'data/pangenomes/Enterococcus_B.faecium/summarize_spots.tsv'\r\n",
      "2022-11-21 12:31:29 writeFlat.py:l846 INFO\tClustering RGPs based on module content...\r\n",
      "2022-11-21 12:31:29 writeFlat.py:l881 INFO\tRGP and associated modules are listed in : data/pangenomes/Enterococcus_B.faecium/modules_RGP_lists.tsv\r\n",
      "2022-11-21 12:31:30 writeFlat.py:l837 INFO\tDone writing module to spot associations to: data/pangenomes/Enterococcus_B.faecium/modules_spots.tsv\r\n",
      "2022-11-21 12:31:31 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Enterococcus_B.faecium/pangenomeGraph.gexf'\r\n",
      "2022-11-21 12:31:31 writeFlat.py:l164 INFO\tDone writing the json file : 'data/pangenomes/Enterococcus_B.faecium/pangenomeGraph.json'\r\n",
      "2022-11-21 12:31:37 writeFlat.py:l804 INFO\tDone writing modules to organisms associations to: 'data/pangenomes/Enterococcus_B.faecium/modules_in_organisms.tsv'\r\n",
      "2022-11-21 12:31:37 writeFlat.py:l588 INFO\tDone writing the projection files\r\n",
      "2022-11-21 12:31:37 all.py:l125 INFO\tAnnotation took : 16.99 seconds\r\n",
      "2022-11-21 12:31:37 all.py:l126 INFO\tClustering took : 25.55 seconds\r\n",
      "2022-11-21 12:31:37 all.py:l127 INFO\tBuilding the graph took : 1.81 seconds\r\n",
      "2022-11-21 12:31:37 all.py:l128 INFO\tPartitioning the pangenome took : 47.5 seconds\r\n",
      "2022-11-21 12:31:37 all.py:l129 INFO\tPredicting RGP took : 2.76 seconds\r\n",
      "2022-11-21 12:31:37 all.py:l130 INFO\tGathering RGP into spots took : 12.38 seconds\r\n",
      "2022-11-21 12:31:37 all.py:l131 INFO\tPredicting modules took : 7.3 seconds\r\n",
      "2022-11-21 12:31:37 all.py:l132 INFO\tWriting the pangenome data in HDF5 took : 4.43 seconds\r\n",
      "2022-11-21 12:31:37 all.py:l134 INFO\tWriting descriptive files for the pangenome took : 11.38 seconds\r\n",
      "Genes: 570393\r\n",
      "Organisms: 207\r\n",
      "Families: 7889\r\n",
      "Edges: 18642\r\n",
      "Persistent (min:0.91, max:0.99, sd:0.01, mean:0.98 ): 1891\r\n",
      "Shell ( min:0.04, max:0.95, sd:0.27, mean:0.31 ): 2324\r\n",
      "Cloud ( min:0.0, max:0.04, sd:0.01, mean:0.01 ): 3674\r\n",
      "Number of partitions: 12\r\n",
      "Shell S2 : 142\r\n",
      "Shell S4 : 29\r\n",
      "Shell S8 : 104\r\n",
      "Shell S12 : 1071\r\n",
      "Shell S10 : 306\r\n",
      "Shell S_ : 139\r\n",
      "Shell S9 : 88\r\n",
      "Shell S1 : 160\r\n",
      "Shell S3 : 148\r\n",
      "Shell S5 : 47\r\n",
      "Shell S11 : 90\r\n",
      "RGPs: 5939\r\n",
      "Spots: 185\r\n",
      "Modules: 317\r\n",
      "Families in Modules: 2391\r\n",
      "2022-11-21 12:31:44 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin all --anno data/pangenomes/Klebsiella.pneumoniae.list -o data/pangenomes/Klebsiella.pneumoniae -c 6 -f\r\n",
      "2022-11-21 12:31:44 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 12:31:44 annotate.py:l448 INFO\tReading data/pangenomes/Klebsiella.pneumoniae.list the list of organism files ...\r\n",
      "100%|███████████████████████████████████████| 600/600 [01:32<00:00,  6.47file/s]\r\n",
      "2022-11-21 12:33:23 annotate.py:l469 INFO\tgene identifiers used in the provided annotation files were unique, PPanGGOLiN will use them.\r\n",
      "2022-11-21 12:33:23 writeBinaries.py:l932 INFO\tWriting genome annotations...\r\n",
      "100%|█████████████████████████████████████| 600/600 [00:08<00:00, 70.99genome/s]\r\n",
      "2022-11-21 12:33:33 writeBinaries.py:l950 INFO\twriting the protein coding gene dna sequences\r\n",
      "100%|███████████████████████████| 3100191/3100191 [00:04<00:00, 654879.74gene/s]\r\n",
      "2022-11-21 12:33:47 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Klebsiella.pneumoniae/pangenome.h5\r\n",
      "2022-11-21 12:33:47 writeSequences.py:l39 INFO\tWriting all of the CDS sequences...\r\n",
      "100%|███████████████████████████| 3100191/3100191 [00:04<00:00, 672164.44gene/s]\r\n",
      "2022-11-21 12:33:52 cluster.py:l297 INFO\tClustering all of the genes sequences...\r\n",
      "2022-11-21 12:33:52 cluster.py:l86 INFO\tCreating sequence database...\r\n",
      "2022-11-21 12:34:07 cluster.py:l93 INFO\tClustering sequences...\r\n",
      "2022-11-21 12:35:03 cluster.py:l99 INFO\tExtracting cluster representatives...\r\n",
      "2022-11-21 12:35:04 cluster.py:l108 INFO\tWriting gene to family informations\r\n",
      "2022-11-21 12:35:05 cluster.py:l305 INFO\tAssociating fragments to their original gene family...\r\n",
      "2022-11-21 12:35:05 cluster.py:l153 INFO\tAligning cluster representatives...\r\n",
      "2022-11-21 12:35:52 cluster.py:l159 INFO\tExtracting alignments...\r\n",
      "2022-11-21 12:35:55 cluster.py:l197 INFO\tStarting with 40178 families\r\n",
      "2022-11-21 12:35:55 cluster.py:l231 INFO\tEnding with 29144 gene families\r\n",
      "2022-11-21 12:35:56 cluster.py:l242 INFO\tAdding protein sequences to the gene families\r\n",
      "2022-11-21 12:35:56 cluster.py:l256 INFO\tAdding 3100191 genes to the gene families\r\n",
      "100%|███████████████████████████| 3100191/3100191 [00:11<00:00, 275347.73gene/s]\r\n",
      "2022-11-21 12:36:08 writeBinaries.py:l955 INFO\tWriting gene families and gene associations...\r\n",
      "100%|█████████████████████████| 29144/29144 [00:02<00:00, 11136.06gene family/s]\r\n",
      "2022-11-21 12:36:12 writeBinaries.py:l957 INFO\tWriting gene families information...\r\n",
      "100%|████████████████████████| 29144/29144 [00:00<00:00, 148564.23gene family/s]\r\n",
      "2022-11-21 12:36:12 writeBinaries.py:l823 INFO\tUpdating annotations with fragment information\r\n",
      "100%|███████████████████████████| 3166261/3166261 [00:09<00:00, 318284.73gene/s]\r\n",
      "2022-11-21 12:36:22 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Klebsiella.pneumoniae/pangenome.h5\r\n",
      "2022-11-21 12:36:22 makeGraph.py:l87 INFO\tComputing the neighbors graph...\r\n",
      "Processing GCF_910589455.1_Kp1-VIM_genomic: 100%|█| 600/600 [00:14<00:00, 42.22o\r\n",
      "2022-11-21 12:36:37 makeGraph.py:l108 INFO\tDone making the neighbors graph.\r\n",
      "2022-11-21 12:36:37 partition.py:l482 INFO\tEstimating the optimal number of partitions...\r\n",
      "100%|█████████████████████████| 19/19 [07:03<00:00, 22.26s/Number of partitions]\r\n",
      "2022-11-21 12:43:42 partition.py:l485 INFO\tThe number of partitions has been evaluated at 13\r\n",
      "2022-11-21 12:43:42 partition.py:l503 INFO\tPartitioning...\r\n",
      "2022-11-21 12:43:42 partition.py:l547 INFO\tLaunching NEM\r\n",
      "100%|██████████████████████████| 6/6 [18:17<00:00, 182.93s/ samples partitioned]\r\n",
      "2022-11-21 13:02:00 partition.py:l566 INFO\tDid 6 partitioning with chunks of size 500 among 600 genomes in 1098.49 seconds.\r\n",
      "2022-11-21 13:02:00 writeBinaries.py:l966 INFO\tWriting the edges...\r\n",
      "100%|████████████████████████████████| 61922/61922 [00:02<00:00, 20786.95edge/s]\r\n",
      "2022-11-21 13:02:04 writeBinaries.py:l808 INFO\tUpdating gene families with partition information\r\n",
      "100%|█████████████████████████| 29144/29144 [00:00<00:00, 96739.29gene family/s]\r\n",
      "2022-11-21 13:02:05 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Klebsiella.pneumoniae/pangenome.h5\r\n",
      "2022-11-21 13:02:05 genomicIsland.py:l250 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 13:02:07 genomicIsland.py:l252 INFO\tCompute Regions of Genomic Plasticity ...\r\n",
      "100%|████████████████████████████████████| 600/600 [00:16<00:00, 36.02genomes/s]\r\n",
      "2022-11-21 13:02:23 genomicIsland.py:l257 INFO\tPredicted 24874 RGP\r\n",
      "2022-11-21 13:02:23 spot.py:l191 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 13:02:25 spot.py:l194 INFO\tDetecting hotspots in the pan...\r\n",
      "2022-11-21 13:02:26 spot.py:l118 INFO\t1599 RGPs were not used as they are on a contig border (or have less than 3 persistent gene families until the contig border)\r\n",
      "2022-11-21 13:02:26 spot.py:l120 INFO\t23275 RGPs are being used to predict spots of insertion\r\n",
      "2022-11-21 13:02:26 spot.py:l122 INFO\t1216 number of different pairs of flanking gene families\r\n",
      "2022-11-21 13:02:30 spot.py:l203 INFO\t523 spots were detected\r\n",
      "2022-11-21 13:02:30 module.py:l121 INFO\tBuilding the graph...\r\n",
      "100%|█████████████████████████████████████| 600/600 [00:43<00:00, 13.84genome/s]\r\n",
      "2022-11-21 13:03:13 module.py:l123 INFO\tTook 43.35 seconds to build the graph to find modules in\r\n",
      "2022-11-21 13:03:13 module.py:l124 INFO\tThere are 29144 nodes and 336518 edges\r\n",
      "2022-11-21 13:03:17 module.py:l137 INFO\tThere are 8197 families among 1167 modules\r\n",
      "2022-11-21 13:03:17 module.py:l138 INFO\tComputing modules took 3.68 seconds\r\n",
      "2022-11-21 13:03:17 writeBinaries.py:l975 INFO\tWriting Regions of Genomic Plasticity...\r\n",
      "100%|██████████████████████████████| 24874/24874 [00:00<00:00, 81582.20region/s]\r\n",
      "2022-11-21 13:03:17 writeBinaries.py:l980 INFO\tWriting Spots of Insertion...\r\n",
      "100%|████████████████████████████████████| 523/523 [00:00<00:00, 33770.88spot/s]\r\n",
      "2022-11-21 13:03:17 writeBinaries.py:l985 INFO\tWriting Modules...\r\n",
      "100%|██████████████████████████████| 1167/1167 [00:00<00:00, 187869.53modules/s]\r\n",
      "2022-11-21 13:03:18 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Klebsiella.pneumoniae/pangenome.h5\r\n",
      "2022-11-21 13:03:28 draw_spot.py:l650 INFO\tDrawing 191 spots\r\n",
      "2022-11-21 13:03:28 draw_spot.py:l559 INFO\tOrdering genes among regions, and drawing spots...\r\n",
      "100%|███████████████████████████████████████| 191/191 [00:38<00:00,  4.96spot/s]\r\n",
      "2022-11-21 13:04:08 draw_spot.py:l617 INFO\tDone drawing spot(s), they can be found in the directory: 'data/pangenomes/Klebsiella.pneumoniae/spot_figures'\r\n",
      "2022-11-21 13:04:08 tile_plot.py:l38 INFO\tDrawing the tile plot...\r\n",
      "2022-11-21 13:04:08 tile_plot.py:l56 INFO\tstart with matrice\r\n",
      "2022-11-21 13:04:13 tile_plot.py:l72 INFO\tdone with making the dendrogram to order the organisms on the plot\r\n",
      "2022-11-21 13:04:14 tile_plot.py:l106 INFO\tGetting the gene name(s) and the number for each tile of the plot ...\r\n",
      "2022-11-21 13:04:19 tile_plot.py:l116 INFO\tDone extracting names and numbers. Making the heatmap ...\r\n",
      "2022-11-21 13:04:33 tile_plot.py:l173 INFO\tDrawing the figure itself...\r\n",
      "2022-11-21 13:05:40 tile_plot.py:l175 INFO\tDone with the tile plot : 'data/pangenomes/Klebsiella.pneumoniae/tile_plot.html' \r\n",
      "2022-11-21 13:05:41 ucurve.py:l23 INFO\tDrawing the U-shaped curve...\r\n",
      "2022-11-21 13:05:41 ucurve.py:l77 INFO\tDone drawing the U-shaped curve : 'data/pangenomes/Klebsiella.pneumoniae/Ushaped_plot.html'\r\n",
      "2022-11-21 13:05:42 writeFlat.py:l328 INFO\tWriting the .csv file ...\r\n",
      "2022-11-21 13:05:42 writeFlat.py:l399 INFO\tWriting the gene presence absence file ...\r\n",
      "2022-11-21 13:05:42 writeFlat.py:l305 INFO\tWriting the gexf file for the pangenome graph...\r\n",
      "2022-11-21 13:05:42 writeFlat.py:l305 INFO\tWriting the light gexf file for the pangenome graph...\r\n",
      "2022-11-21 13:05:42 writeFlat.py:l582 INFO\tWriting the projection files...\r\n",
      "2022-11-21 13:05:42 writeFlat.py:l431 INFO\tWriting pangenome statistics...\r\n",
      "2022-11-21 13:05:42 writeFlat.py:l432 INFO\tWriting statistics on persistent duplication...\r\n",
      "2022-11-21 13:05:42 writeFlat.py:l454 INFO\tDone writing stats on persistent duplication\r\n",
      "2022-11-21 13:05:42 writeFlat.py:l455 INFO\tWriting genome per genome statistics (completeness and counts)...\r\n",
      "2022-11-21 13:05:43 writeFlat.py:l419 INFO\tDone writing the gene presence absence file : 'data/pangenomes/Klebsiella.pneumoniae/gene_presence_absence.Rtab'\r\n",
      "2022-11-21 13:05:43 writeFlat.py:l157 INFO\tWriting the json file for the pangenome graph...\r\n",
      "2022-11-21 13:05:51 writeFlat.py:l520 INFO\tDone writing genome per genome statistics\r\n",
      "2022-11-21 13:05:51 writeFlat.py:l598 INFO\tWriting the list of gene families for each partition ...\r\n",
      "2022-11-21 13:05:52 writeFlat.py:l625 INFO\tDone writing the list of gene families for each partition\r\n",
      "2022-11-21 13:05:55 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Klebsiella.pneumoniae/pangenomeGraph_light.gexf'\r\n",
      "2022-11-21 13:05:57 writeFlat.py:l389 INFO\tDone writing the matrix : 'data/pangenomes/Klebsiella.pneumoniae/matrix.csv'\r\n",
      "2022-11-21 13:05:57 writeFlat.py:l776 INFO\tWriting functional modules...\r\n",
      "2022-11-21 13:05:57 writeFlat.py:l784 INFO\tDone writing functional modules to: 'data/pangenomes/Klebsiella.pneumoniae/functional_modules.tsv'\r\n",
      "2022-11-21 13:05:57 writeFlat.py:l752 INFO\tWriting functional modules summary...\r\n",
      "2022-11-21 13:05:57 writeFlat.py:l767 INFO\tDone writing module summary: 'data/pangenomes/Klebsiella.pneumoniae/modules_summary.tsv'\r\n",
      "2022-11-21 13:05:57 writeFlat.py:l793 INFO\tWriting modules to organisms associations...\r\n",
      "2022-11-21 13:05:58 writeFlat.py:l814 INFO\tWriting modules to spot associations...\r\n",
      "2022-11-21 13:06:03 writeFlat.py:l691 INFO\tDone writing spots in : 'data/pangenomes/Klebsiella.pneumoniae/summarize_spots.tsv'\r\n",
      "2022-11-21 13:06:03 writeFlat.py:l846 INFO\tClustering RGPs based on module content...\r\n",
      "2022-11-21 13:06:03 writeFlat.py:l881 INFO\tRGP and associated modules are listed in : data/pangenomes/Klebsiella.pneumoniae/modules_RGP_lists.tsv\r\n",
      "2022-11-21 13:06:11 writeFlat.py:l837 INFO\tDone writing module to spot associations to: data/pangenomes/Klebsiella.pneumoniae/modules_spots.tsv\r\n",
      "2022-11-21 13:06:19 writeFlat.py:l164 INFO\tDone writing the json file : 'data/pangenomes/Klebsiella.pneumoniae/pangenomeGraph.json'\r\n",
      "2022-11-21 13:06:27 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Klebsiella.pneumoniae/pangenomeGraph.gexf'\r\n",
      "2022-11-21 13:06:34 writeFlat.py:l588 INFO\tDone writing the projection files\r\n",
      "2022-11-21 13:07:19 writeFlat.py:l804 INFO\tDone writing modules to organisms associations to: 'data/pangenomes/Klebsiella.pneumoniae/modules_in_organisms.tsv'\r\n",
      "2022-11-21 13:07:20 all.py:l125 INFO\tAnnotation took : 99.49 seconds\r\n",
      "2022-11-21 13:07:20 all.py:l126 INFO\tClustering took : 140.82 seconds\r\n",
      "2022-11-21 13:07:20 all.py:l127 INFO\tBuilding the graph took : 14.21 seconds\r\n",
      "2022-11-21 13:07:20 all.py:l128 INFO\tPartitioning the pangenome took : 1523.59 seconds\r\n",
      "2022-11-21 13:07:20 all.py:l129 INFO\tPredicting RGP took : 18.54 seconds\r\n",
      "2022-11-21 13:07:20 all.py:l130 INFO\tGathering RGP into spots took : 56.8 seconds\r\n",
      "2022-11-21 13:07:20 all.py:l131 INFO\tPredicting modules took : 47.05 seconds\r\n",
      "2022-11-21 13:07:20 all.py:l132 INFO\tWriting the pangenome data in HDF5 took : 29.37 seconds\r\n",
      "2022-11-21 13:07:20 all.py:l134 INFO\tWriting descriptive files for the pangenome took : 98.79 seconds\r\n",
      "Genes: 3100191\r\n",
      "Organisms: 600\r\n",
      "Families: 29144\r\n",
      "Edges: 61922\r\n",
      "Persistent (min:0.9, max:0.99, sd:0.01, mean:0.97 ): 4199\r\n",
      "Shell ( min:0.02, max:0.94, sd:0.2, mean:0.16 ): 5711\r\n",
      "Cloud ( min:0.0, max:0.02, sd:0.0, mean:0.0 ): 19234\r\n",
      "Number of partitions: 3\r\n",
      "RGPs: 24874\r\n",
      "Spots: 523\r\n",
      "Modules: 1167\r\n",
      "Families in Modules: 8197\r\n",
      "2022-11-21 13:07:54 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin all --anno data/pangenomes/Pseudomonas.aeruginosa.list -o data/pangenomes/Pseudomonas.aeruginosa -c 6 -f\r\n",
      "2022-11-21 13:07:54 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 13:07:54 annotate.py:l448 INFO\tReading data/pangenomes/Pseudomonas.aeruginosa.list the list of organism files ...\r\n",
      "100%|███████████████████████████████████████| 313/313 [00:59<00:00,  5.25file/s]\r\n",
      "2022-11-21 13:08:58 annotate.py:l469 INFO\tgene identifiers used in the provided annotation files were unique, PPanGGOLiN will use them.\r\n",
      "2022-11-21 13:08:58 writeBinaries.py:l932 INFO\tWriting genome annotations...\r\n",
      "100%|█████████████████████████████████████| 313/313 [00:05<00:00, 54.50genome/s]\r\n",
      "2022-11-21 13:09:05 writeBinaries.py:l950 INFO\twriting the protein coding gene dna sequences\r\n",
      "100%|███████████████████████████| 1892863/1892863 [00:03<00:00, 564281.35gene/s]\r\n",
      "2022-11-21 13:09:19 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Pseudomonas.aeruginosa/pangenome.h5\r\n",
      "2022-11-21 13:09:19 writeSequences.py:l39 INFO\tWriting all of the CDS sequences...\r\n",
      "100%|███████████████████████████| 1892863/1892863 [00:03<00:00, 587332.53gene/s]\r\n",
      "2022-11-21 13:09:23 cluster.py:l297 INFO\tClustering all of the genes sequences...\r\n",
      "2022-11-21 13:09:23 cluster.py:l86 INFO\tCreating sequence database...\r\n",
      "2022-11-21 13:09:33 cluster.py:l93 INFO\tClustering sequences...\r\n",
      "2022-11-21 13:10:11 cluster.py:l99 INFO\tExtracting cluster representatives...\r\n",
      "2022-11-21 13:10:12 cluster.py:l108 INFO\tWriting gene to family informations\r\n",
      "2022-11-21 13:10:13 cluster.py:l305 INFO\tAssociating fragments to their original gene family...\r\n",
      "2022-11-21 13:10:13 cluster.py:l153 INFO\tAligning cluster representatives...\r\n",
      "2022-11-21 13:10:59 cluster.py:l159 INFO\tExtracting alignments...\r\n",
      "2022-11-21 13:11:01 cluster.py:l197 INFO\tStarting with 27256 families\r\n",
      "2022-11-21 13:11:01 cluster.py:l231 INFO\tEnding with 23687 gene families\r\n",
      "2022-11-21 13:11:02 cluster.py:l242 INFO\tAdding protein sequences to the gene families\r\n",
      "2022-11-21 13:11:02 cluster.py:l256 INFO\tAdding 1892863 genes to the gene families\r\n",
      "100%|███████████████████████████| 1892863/1892863 [00:07<00:00, 241585.99gene/s]\r\n",
      "2022-11-21 13:11:10 writeBinaries.py:l955 INFO\tWriting gene families and gene associations...\r\n",
      "100%|█████████████████████████| 23687/23687 [00:01<00:00, 12562.99gene family/s]\r\n",
      "2022-11-21 13:11:13 writeBinaries.py:l957 INFO\tWriting gene families information...\r\n",
      "100%|████████████████████████| 23687/23687 [00:00<00:00, 101086.74gene family/s]\r\n",
      "2022-11-21 13:11:13 writeBinaries.py:l823 INFO\tUpdating annotations with fragment information\r\n",
      "100%|███████████████████████████| 1916555/1916555 [00:07<00:00, 269384.15gene/s]\r\n",
      "2022-11-21 13:11:20 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Pseudomonas.aeruginosa/pangenome.h5\r\n",
      "2022-11-21 13:11:20 makeGraph.py:l87 INFO\tComputing the neighbors graph...\r\n",
      "Processing GCF_902703215.1_C7-25_genomic: 100%|█| 313/313 [00:09<00:00, 31.39org\r\n",
      "2022-11-21 13:11:30 makeGraph.py:l108 INFO\tDone making the neighbors graph.\r\n",
      "2022-11-21 13:11:30 partition.py:l482 INFO\tEstimating the optimal number of partitions...\r\n",
      "100%|█████████████████████████| 19/19 [04:12<00:00, 13.29s/Number of partitions]\r\n",
      "2022-11-21 13:15:44 partition.py:l485 INFO\tThe number of partitions has been evaluated at 13\r\n",
      "2022-11-21 13:15:44 partition.py:l503 INFO\tPartitioning...\r\n",
      "2022-11-21 13:18:37 partition.py:l578 INFO\tPartitioned 313 genomes in 172.48 seconds.\r\n",
      "2022-11-21 13:18:37 writeBinaries.py:l966 INFO\tWriting the edges...\r\n",
      "100%|████████████████████████████████| 42119/42119 [00:02<00:00, 15391.68edge/s]\r\n",
      "2022-11-21 13:18:40 writeBinaries.py:l808 INFO\tUpdating gene families with partition information\r\n",
      "100%|█████████████████████████| 23687/23687 [00:00<00:00, 44983.48gene family/s]\r\n",
      "2022-11-21 13:18:41 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Pseudomonas.aeruginosa/pangenome.h5\r\n",
      "2022-11-21 13:18:41 genomicIsland.py:l250 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 13:18:43 genomicIsland.py:l252 INFO\tCompute Regions of Genomic Plasticity ...\r\n",
      "100%|████████████████████████████████████| 313/313 [00:13<00:00, 22.67genomes/s]\r\n",
      "2022-11-21 13:18:57 genomicIsland.py:l257 INFO\tPredicted 10771 RGP\r\n",
      "2022-11-21 13:18:57 spot.py:l191 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 13:18:58 spot.py:l194 INFO\tDetecting hotspots in the pan...\r\n",
      "2022-11-21 13:18:59 spot.py:l118 INFO\t25 RGPs were not used as they are on a contig border (or have less than 3 persistent gene families until the contig border)\r\n",
      "2022-11-21 13:18:59 spot.py:l120 INFO\t10746 RGPs are being used to predict spots of insertion\r\n",
      "2022-11-21 13:18:59 spot.py:l122 INFO\t901 number of different pairs of flanking gene families\r\n",
      "2022-11-21 13:19:02 spot.py:l203 INFO\t539 spots were detected\r\n",
      "2022-11-21 13:19:02 module.py:l121 INFO\tBuilding the graph...\r\n",
      "100%|█████████████████████████████████████| 313/313 [00:34<00:00,  9.08genome/s]\r\n",
      "2022-11-21 13:19:36 module.py:l123 INFO\tTook 34.49 seconds to build the graph to find modules in\r\n",
      "2022-11-21 13:19:36 module.py:l124 INFO\tThere are 23687 nodes and 234908 edges\r\n",
      "2022-11-21 13:19:39 module.py:l137 INFO\tThere are 6358 families among 915 modules\r\n",
      "2022-11-21 13:19:39 module.py:l138 INFO\tComputing modules took 2.31 seconds\r\n",
      "2022-11-21 13:19:39 writeBinaries.py:l975 INFO\tWriting Regions of Genomic Plasticity...\r\n",
      "100%|██████████████████████████████| 10771/10771 [00:00<00:00, 62150.10region/s]\r\n",
      "2022-11-21 13:19:39 writeBinaries.py:l980 INFO\tWriting Spots of Insertion...\r\n",
      "100%|████████████████████████████████████| 539/539 [00:00<00:00, 63902.14spot/s]\r\n",
      "2022-11-21 13:19:39 writeBinaries.py:l985 INFO\tWriting Modules...\r\n",
      "100%|████████████████████████████████| 915/915 [00:00<00:00, 129953.55modules/s]\r\n",
      "2022-11-21 13:19:39 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Pseudomonas.aeruginosa/pangenome.h5\r\n",
      "2022-11-21 13:19:43 draw_spot.py:l650 INFO\tDrawing 182 spots\r\n",
      "2022-11-21 13:19:43 draw_spot.py:l559 INFO\tOrdering genes among regions, and drawing spots...\r\n",
      "100%|███████████████████████████████████████| 182/182 [00:41<00:00,  4.44spot/s]\r\n",
      "2022-11-21 13:20:26 draw_spot.py:l617 INFO\tDone drawing spot(s), they can be found in the directory: 'data/pangenomes/Pseudomonas.aeruginosa/spot_figures'\r\n",
      "2022-11-21 13:20:26 tile_plot.py:l38 INFO\tDrawing the tile plot...\r\n",
      "2022-11-21 13:20:26 tile_plot.py:l56 INFO\tstart with matrice\r\n",
      "2022-11-21 13:20:28 tile_plot.py:l72 INFO\tdone with making the dendrogram to order the organisms on the plot\r\n",
      "2022-11-21 13:20:28 tile_plot.py:l106 INFO\tGetting the gene name(s) and the number for each tile of the plot ...\r\n",
      "2022-11-21 13:20:33 tile_plot.py:l116 INFO\tDone extracting names and numbers. Making the heatmap ...\r\n",
      "2022-11-21 13:20:55 tile_plot.py:l173 INFO\tDrawing the figure itself...\r\n",
      "2022-11-21 13:22:29 tile_plot.py:l175 INFO\tDone with the tile plot : 'data/pangenomes/Pseudomonas.aeruginosa/tile_plot.html' \r\n",
      "2022-11-21 13:22:29 ucurve.py:l23 INFO\tDrawing the U-shaped curve...\r\n",
      "2022-11-21 13:22:29 ucurve.py:l77 INFO\tDone drawing the U-shaped curve : 'data/pangenomes/Pseudomonas.aeruginosa/Ushaped_plot.html'\r\n",
      "2022-11-21 13:22:30 writeFlat.py:l328 INFO\tWriting the .csv file ...\r\n",
      "2022-11-21 13:22:30 writeFlat.py:l399 INFO\tWriting the gene presence absence file ...\r\n",
      "2022-11-21 13:22:30 writeFlat.py:l305 INFO\tWriting the gexf file for the pangenome graph...\r\n",
      "2022-11-21 13:22:30 writeFlat.py:l305 INFO\tWriting the light gexf file for the pangenome graph...\r\n",
      "2022-11-21 13:22:30 writeFlat.py:l582 INFO\tWriting the projection files...\r\n",
      "2022-11-21 13:22:30 writeFlat.py:l431 INFO\tWriting pangenome statistics...\r\n",
      "2022-11-21 13:22:30 writeFlat.py:l432 INFO\tWriting statistics on persistent duplication...\r\n",
      "2022-11-21 13:22:31 writeFlat.py:l419 INFO\tDone writing the gene presence absence file : 'data/pangenomes/Pseudomonas.aeruginosa/gene_presence_absence.Rtab'\r\n",
      "2022-11-21 13:22:31 writeFlat.py:l157 INFO\tWriting the json file for the pangenome graph...\r\n",
      "2022-11-21 13:22:31 writeFlat.py:l454 INFO\tDone writing stats on persistent duplication\r\n",
      "2022-11-21 13:22:31 writeFlat.py:l455 INFO\tWriting genome per genome statistics (completeness and counts)...\r\n",
      "2022-11-21 13:22:37 writeFlat.py:l520 INFO\tDone writing genome per genome statistics\r\n",
      "2022-11-21 13:22:37 writeFlat.py:l598 INFO\tWriting the list of gene families for each partition ...\r\n",
      "2022-11-21 13:22:38 writeFlat.py:l625 INFO\tDone writing the list of gene families for each partition\r\n",
      "2022-11-21 13:22:39 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Pseudomonas.aeruginosa/pangenomeGraph_light.gexf'\r\n",
      "2022-11-21 13:22:41 writeFlat.py:l389 INFO\tDone writing the matrix : 'data/pangenomes/Pseudomonas.aeruginosa/matrix.csv'\r\n",
      "2022-11-21 13:22:41 writeFlat.py:l776 INFO\tWriting functional modules...\r\n",
      "2022-11-21 13:22:41 writeFlat.py:l784 INFO\tDone writing functional modules to: 'data/pangenomes/Pseudomonas.aeruginosa/functional_modules.tsv'\r\n",
      "2022-11-21 13:22:41 writeFlat.py:l752 INFO\tWriting functional modules summary...\r\n",
      "2022-11-21 13:22:41 writeFlat.py:l767 INFO\tDone writing module summary: 'data/pangenomes/Pseudomonas.aeruginosa/modules_summary.tsv'\r\n",
      "2022-11-21 13:22:41 writeFlat.py:l793 INFO\tWriting modules to organisms associations...\r\n",
      "2022-11-21 13:22:42 writeFlat.py:l814 INFO\tWriting modules to spot associations...\r\n",
      "2022-11-21 13:22:43 writeFlat.py:l691 INFO\tDone writing spots in : 'data/pangenomes/Pseudomonas.aeruginosa/summarize_spots.tsv'\r\n",
      "2022-11-21 13:22:43 writeFlat.py:l846 INFO\tClustering RGPs based on module content...\r\n",
      "2022-11-21 13:22:43 writeFlat.py:l881 INFO\tRGP and associated modules are listed in : data/pangenomes/Pseudomonas.aeruginosa/modules_RGP_lists.tsv\r\n",
      "2022-11-21 13:22:46 writeFlat.py:l837 INFO\tDone writing module to spot associations to: data/pangenomes/Pseudomonas.aeruginosa/modules_spots.tsv\r\n",
      "2022-11-21 13:22:52 writeFlat.py:l588 INFO\tDone writing the projection files\r\n",
      "2022-11-21 13:22:54 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Pseudomonas.aeruginosa/pangenomeGraph.gexf'\r\n",
      "2022-11-21 13:22:55 writeFlat.py:l164 INFO\tDone writing the json file : 'data/pangenomes/Pseudomonas.aeruginosa/pangenomeGraph.json'\r\n",
      "2022-11-21 13:23:29 writeFlat.py:l804 INFO\tDone writing modules to organisms associations to: 'data/pangenomes/Pseudomonas.aeruginosa/modules_in_organisms.tsv'\r\n",
      "2022-11-21 13:23:30 all.py:l125 INFO\tAnnotation took : 64.03 seconds\r\n",
      "2022-11-21 13:23:30 all.py:l126 INFO\tClustering took : 110.49 seconds\r\n",
      "2022-11-21 13:23:30 all.py:l127 INFO\tBuilding the graph took : 9.97 seconds\r\n",
      "2022-11-21 13:23:30 all.py:l128 INFO\tPartitioning the pangenome took : 426.48 seconds\r\n",
      "2022-11-21 13:23:30 all.py:l129 INFO\tPredicting RGP took : 15.54 seconds\r\n",
      "2022-11-21 13:23:30 all.py:l130 INFO\tGathering RGP into spots took : 51.56 seconds\r\n",
      "2022-11-21 13:23:30 all.py:l131 INFO\tPredicting modules took : 36.8 seconds\r\n",
      "2022-11-21 13:23:30 all.py:l132 INFO\tWriting the pangenome data in HDF5 took : 25.83 seconds\r\n",
      "2022-11-21 13:23:30 all.py:l134 INFO\tWriting descriptive files for the pangenome took : 60.68 seconds\r\n",
      "Genes: 1892863\r\n",
      "Organisms: 313\r\n",
      "Families: 23687\r\n",
      "Edges: 42119\r\n",
      "Persistent (min:0.89, max:0.99, sd:0.01, mean:0.98 ): 5284\r\n",
      "Shell ( min:0.02, max:0.9, sd:0.16, mean:0.13 ): 5380\r\n",
      "Cloud ( min:0.0, max:0.02, sd:0.0, mean:0.01 ): 13023\r\n",
      "Number of partitions: 13\r\n",
      "Shell S9 : 77\r\n",
      "Shell S10 : 1210\r\n",
      "Shell S8 : 435\r\n",
      "Shell S11 : 3280\r\n",
      "Shell S5 : 76\r\n",
      "Shell S3 : 55\r\n",
      "Shell S1 : 106\r\n",
      "Shell S2 : 27\r\n",
      "Shell S7 : 51\r\n",
      "Shell S4 : 36\r\n",
      "Shell S6 : 27\r\n",
      "RGPs: 10771\r\n",
      "Spots: 539\r\n",
      "Modules: 915\r\n",
      "Families in Modules: 6358\r\n",
      "2022-11-21 13:23:57 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin all --anno data/pangenomes/Staphylococcus.aureus.list -o data/pangenomes/Staphylococcus.aureus -c 6 -f\r\n",
      "2022-11-21 13:23:57 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 13:23:57 annotate.py:l448 INFO\tReading data/pangenomes/Staphylococcus.aureus.list the list of organism files ...\r\n",
      "100%|███████████████████████████████████████| 600/600 [01:05<00:00,  9.17file/s]\r\n",
      "2022-11-21 13:25:04 annotate.py:l472 INFO\tgene identifiers used in the provided annotation files were not unique, PPanGGOLiN will use self-generated identifiers.\r\n",
      "2022-11-21 13:25:04 writeBinaries.py:l932 INFO\tWriting genome annotations...\r\n",
      "100%|█████████████████████████████████████| 600/600 [00:06<00:00, 95.62genome/s]\r\n",
      "2022-11-21 13:25:12 writeBinaries.py:l950 INFO\twriting the protein coding gene dna sequences\r\n",
      "100%|███████████████████████████| 1586295/1586295 [00:03<00:00, 450148.82gene/s]\r\n",
      "2022-11-21 13:25:24 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Staphylococcus.aureus/pangenome.h5\r\n",
      "2022-11-21 13:25:24 writeSequences.py:l39 INFO\tWriting all of the CDS sequences...\r\n",
      "100%|███████████████████████████| 1586295/1586295 [00:03<00:00, 478791.67gene/s]\r\n",
      "2022-11-21 13:25:27 cluster.py:l297 INFO\tClustering all of the genes sequences...\r\n",
      "2022-11-21 13:25:27 cluster.py:l86 INFO\tCreating sequence database...\r\n",
      "2022-11-21 13:25:39 cluster.py:l93 INFO\tClustering sequences...\r\n",
      "2022-11-21 13:26:48 cluster.py:l99 INFO\tExtracting cluster representatives...\r\n",
      "2022-11-21 13:26:49 cluster.py:l108 INFO\tWriting gene to family informations\r\n",
      "2022-11-21 13:26:49 cluster.py:l305 INFO\tAssociating fragments to their original gene family...\r\n",
      "2022-11-21 13:26:50 cluster.py:l153 INFO\tAligning cluster representatives...\r\n",
      "2022-11-21 13:27:38 cluster.py:l159 INFO\tExtracting alignments...\r\n",
      "2022-11-21 13:27:41 cluster.py:l197 INFO\tStarting with 10521 families\r\n",
      "2022-11-21 13:27:41 cluster.py:l231 INFO\tEnding with 7005 gene families\r\n",
      "2022-11-21 13:27:42 cluster.py:l242 INFO\tAdding protein sequences to the gene families\r\n",
      "2022-11-21 13:27:42 cluster.py:l256 INFO\tAdding 1586295 genes to the gene families\r\n",
      "100%|███████████████████████████| 1586295/1586295 [00:12<00:00, 130026.57gene/s]\r\n",
      "2022-11-21 13:27:54 writeBinaries.py:l955 INFO\tWriting gene families and gene associations...\r\n",
      "100%|████████████████████████████| 7005/7005 [00:03<00:00, 1796.70gene family/s]\r\n",
      "2022-11-21 13:28:00 writeBinaries.py:l957 INFO\tWriting gene families information...\r\n",
      "100%|███████████████████████████| 7005/7005 [00:00<00:00, 53417.17gene family/s]\r\n",
      "2022-11-21 13:28:00 writeBinaries.py:l823 INFO\tUpdating annotations with fragment information\r\n",
      "100%|███████████████████████████| 1631806/1631806 [00:07<00:00, 212962.22gene/s]\r\n",
      "2022-11-21 13:28:07 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Staphylococcus.aureus/pangenome.h5\r\n",
      "2022-11-21 13:28:07 makeGraph.py:l87 INFO\tComputing the neighbors graph...\r\n",
      "Processing GCF_903932595.1_P3: 100%|████| 600/600 [00:07<00:00, 79.20organism/s]\r\n",
      "2022-11-21 13:28:15 makeGraph.py:l108 INFO\tDone making the neighbors graph.\r\n",
      "2022-11-21 13:28:15 partition.py:l482 INFO\tEstimating the optimal number of partitions...\r\n",
      "100%|█████████████████████████| 19/19 [01:17<00:00,  4.08s/Number of partitions]\r\n",
      "2022-11-21 13:29:34 partition.py:l485 INFO\tThe number of partitions has been evaluated at 15\r\n",
      "2022-11-21 13:29:34 partition.py:l503 INFO\tPartitioning...\r\n",
      "2022-11-21 13:29:34 partition.py:l547 INFO\tLaunching NEM\r\n",
      "100%|███████████████████████████| 6/6 [01:22<00:00, 13.74s/ samples partitioned]\r\n",
      "2022-11-21 13:30:57 partition.py:l566 INFO\tDid 6 partitioning with chunks of size 500 among 600 genomes in 83.1 seconds.\r\n",
      "2022-11-21 13:30:57 writeBinaries.py:l966 INFO\tWriting the edges...\r\n",
      "100%|█████████████████████████████████| 17874/17874 [00:02<00:00, 6090.69edge/s]\r\n",
      "2022-11-21 13:31:00 writeBinaries.py:l808 INFO\tUpdating gene families with partition information\r\n",
      "100%|███████████████████████████| 7005/7005 [00:00<00:00, 36987.74gene family/s]\r\n",
      "2022-11-21 13:31:01 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Staphylococcus.aureus/pangenome.h5\r\n",
      "2022-11-21 13:31:01 genomicIsland.py:l250 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 13:31:02 genomicIsland.py:l252 INFO\tCompute Regions of Genomic Plasticity ...\r\n",
      "100%|████████████████████████████████████| 600/600 [00:08<00:00, 69.83genomes/s]\r\n",
      "2022-11-21 13:31:11 genomicIsland.py:l257 INFO\tPredicted 11307 RGP\r\n",
      "2022-11-21 13:31:11 spot.py:l191 INFO\tDetecting multigenic families...\r\n",
      "2022-11-21 13:31:12 spot.py:l194 INFO\tDetecting hotspots in the pan...\r\n",
      "2022-11-21 13:31:12 spot.py:l118 INFO\t295 RGPs were not used as they are on a contig border (or have less than 3 persistent gene families until the contig border)\r\n",
      "2022-11-21 13:31:12 spot.py:l120 INFO\t11012 RGPs are being used to predict spots of insertion\r\n",
      "2022-11-21 13:31:12 spot.py:l122 INFO\t571 number of different pairs of flanking gene families\r\n",
      "2022-11-21 13:31:13 spot.py:l203 INFO\t257 spots were detected\r\n",
      "2022-11-21 13:31:13 module.py:l121 INFO\tBuilding the graph...\r\n",
      "100%|█████████████████████████████████████| 600/600 [00:27<00:00, 22.01genome/s]\r\n",
      "2022-11-21 13:31:40 module.py:l123 INFO\tTook 27.26 seconds to build the graph to find modules in\r\n",
      "2022-11-21 13:31:40 module.py:l124 INFO\tThere are 7005 nodes and 94259 edges\r\n",
      "2022-11-21 13:31:42 module.py:l137 INFO\tThere are 1276 families among 199 modules\r\n",
      "2022-11-21 13:31:42 module.py:l138 INFO\tComputing modules took 1.87 seconds\r\n",
      "2022-11-21 13:31:42 writeBinaries.py:l975 INFO\tWriting Regions of Genomic Plasticity...\r\n",
      "100%|██████████████████████████████| 11307/11307 [00:00<00:00, 63700.55region/s]\r\n",
      "2022-11-21 13:31:42 writeBinaries.py:l980 INFO\tWriting Spots of Insertion...\r\n",
      "100%|████████████████████████████████████| 257/257 [00:00<00:00, 36802.19spot/s]\r\n",
      "2022-11-21 13:31:42 writeBinaries.py:l985 INFO\tWriting Modules...\r\n",
      "100%|████████████████████████████████| 199/199 [00:00<00:00, 188030.30modules/s]\r\n",
      "2022-11-21 13:31:43 writeBinaries.py:l993 INFO\tDone writing the pangenome. It is in file : data/pangenomes/Staphylococcus.aureus/pangenome.h5\r\n",
      "2022-11-21 13:31:52 draw_spot.py:l650 INFO\tDrawing 116 spots\r\n",
      "2022-11-21 13:31:52 draw_spot.py:l559 INFO\tOrdering genes among regions, and drawing spots...\r\n",
      "100%|███████████████████████████████████████| 116/116 [00:20<00:00,  5.67spot/s]\r\n",
      "2022-11-21 13:32:13 draw_spot.py:l617 INFO\tDone drawing spot(s), they can be found in the directory: 'data/pangenomes/Staphylococcus.aureus/spot_figures'\r\n",
      "2022-11-21 13:32:13 tile_plot.py:l38 INFO\tDrawing the tile plot...\r\n",
      "2022-11-21 13:32:13 tile_plot.py:l56 INFO\tstart with matrice\r\n",
      "2022-11-21 13:32:16 tile_plot.py:l72 INFO\tdone with making the dendrogram to order the organisms on the plot\r\n",
      "2022-11-21 13:32:16 tile_plot.py:l106 INFO\tGetting the gene name(s) and the number for each tile of the plot ...\r\n",
      "2022-11-21 13:32:19 tile_plot.py:l116 INFO\tDone extracting names and numbers. Making the heatmap ...\r\n",
      "2022-11-21 13:32:25 tile_plot.py:l173 INFO\tDrawing the figure itself...\r\n",
      "2022-11-21 13:32:53 tile_plot.py:l175 INFO\tDone with the tile plot : 'data/pangenomes/Staphylococcus.aureus/tile_plot.html' \r\n",
      "2022-11-21 13:32:53 ucurve.py:l23 INFO\tDrawing the U-shaped curve...\r\n",
      "2022-11-21 13:32:53 ucurve.py:l77 INFO\tDone drawing the U-shaped curve : 'data/pangenomes/Staphylococcus.aureus/Ushaped_plot.html'\r\n",
      "2022-11-21 13:32:54 writeFlat.py:l328 INFO\tWriting the .csv file ...\r\n",
      "2022-11-21 13:32:54 writeFlat.py:l399 INFO\tWriting the gene presence absence file ...\r\n",
      "2022-11-21 13:32:54 writeFlat.py:l305 INFO\tWriting the gexf file for the pangenome graph...\r\n",
      "2022-11-21 13:32:54 writeFlat.py:l305 INFO\tWriting the light gexf file for the pangenome graph...\r\n",
      "2022-11-21 13:32:54 writeFlat.py:l582 INFO\tWriting the projection files...\r\n",
      "2022-11-21 13:32:54 writeFlat.py:l431 INFO\tWriting pangenome statistics...\r\n",
      "2022-11-21 13:32:54 writeFlat.py:l432 INFO\tWriting statistics on persistent duplication...\r\n",
      "2022-11-21 13:32:55 writeFlat.py:l454 INFO\tDone writing stats on persistent duplication\r\n",
      "2022-11-21 13:32:55 writeFlat.py:l455 INFO\tWriting genome per genome statistics (completeness and counts)...\r\n",
      "2022-11-21 13:32:55 writeFlat.py:l419 INFO\tDone writing the gene presence absence file : 'data/pangenomes/Staphylococcus.aureus/gene_presence_absence.Rtab'\r\n",
      "2022-11-21 13:32:55 writeFlat.py:l157 INFO\tWriting the json file for the pangenome graph...\r\n",
      "2022-11-21 13:33:00 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Staphylococcus.aureus/pangenomeGraph_light.gexf'\r\n",
      "2022-11-21 13:33:00 writeFlat.py:l598 INFO\tWriting the list of gene families for each partition ...\r\n",
      "2022-11-21 13:33:01 writeFlat.py:l389 INFO\tDone writing the matrix : 'data/pangenomes/Staphylococcus.aureus/matrix.csv'\r\n",
      "2022-11-21 13:33:02 writeFlat.py:l625 INFO\tDone writing the list of gene families for each partition\r\n",
      "2022-11-21 13:33:02 writeFlat.py:l520 INFO\tDone writing genome per genome statistics\r\n",
      "2022-11-21 13:33:02 writeFlat.py:l776 INFO\tWriting functional modules...\r\n",
      "2022-11-21 13:33:02 writeFlat.py:l784 INFO\tDone writing functional modules to: 'data/pangenomes/Staphylococcus.aureus/functional_modules.tsv'\r\n",
      "2022-11-21 13:33:02 writeFlat.py:l752 INFO\tWriting functional modules summary...\r\n",
      "2022-11-21 13:33:02 writeFlat.py:l767 INFO\tDone writing module summary: 'data/pangenomes/Staphylococcus.aureus/modules_summary.tsv'\r\n",
      "2022-11-21 13:33:02 writeFlat.py:l793 INFO\tWriting modules to organisms associations...\r\n",
      "2022-11-21 13:33:03 writeFlat.py:l814 INFO\tWriting modules to spot associations...\r\n",
      "2022-11-21 13:33:08 writeFlat.py:l691 INFO\tDone writing spots in : 'data/pangenomes/Staphylococcus.aureus/summarize_spots.tsv'\r\n",
      "2022-11-21 13:33:08 writeFlat.py:l846 INFO\tClustering RGPs based on module content...\r\n",
      "2022-11-21 13:33:08 writeFlat.py:l881 INFO\tRGP and associated modules are listed in : data/pangenomes/Staphylococcus.aureus/modules_RGP_lists.tsv\r\n",
      "2022-11-21 13:33:10 writeFlat.py:l837 INFO\tDone writing module to spot associations to: data/pangenomes/Staphylococcus.aureus/modules_spots.tsv\r\n",
      "2022-11-21 13:33:13 writeFlat.py:l804 INFO\tDone writing modules to organisms associations to: 'data/pangenomes/Staphylococcus.aureus/modules_in_organisms.tsv'\r\n",
      "2022-11-21 13:33:13 writeFlat.py:l164 INFO\tDone writing the json file : 'data/pangenomes/Staphylococcus.aureus/pangenomeGraph.json'\r\n",
      "2022-11-21 13:33:15 writeFlat.py:l314 INFO\tDone writing the gexf file : 'data/pangenomes/Staphylococcus.aureus/pangenomeGraph.gexf'\r\n",
      "2022-11-21 13:33:16 writeFlat.py:l588 INFO\tDone writing the projection files\r\n",
      "2022-11-21 13:33:16 all.py:l125 INFO\tAnnotation took : 67.87 seconds\r\n",
      "2022-11-21 13:33:16 all.py:l126 INFO\tClustering took : 150.6 seconds\r\n",
      "2022-11-21 13:33:16 all.py:l127 INFO\tBuilding the graph took : 7.58 seconds\r\n",
      "2022-11-21 13:33:16 all.py:l128 INFO\tPartitioning the pangenome took : 161.83 seconds\r\n",
      "2022-11-21 13:33:16 all.py:l129 INFO\tPredicting RGP took : 9.86 seconds\r\n",
      "2022-11-21 13:33:16 all.py:l130 INFO\tGathering RGP into spots took : 33.11 seconds\r\n",
      "2022-11-21 13:33:16 all.py:l131 INFO\tPredicting modules took : 29.13 seconds\r\n",
      "2022-11-21 13:33:16 all.py:l132 INFO\tWriting the pangenome data in HDF5 took : 23.65 seconds\r\n",
      "2022-11-21 13:33:16 all.py:l134 INFO\tWriting descriptive files for the pangenome took : 22.57 seconds\r\n",
      "Genes: 1586295\r\n",
      "Organisms: 600\r\n",
      "Families: 7005\r\n",
      "Edges: 17874\r\n",
      "Persistent (min:0.94, max:0.99, sd:0.01, mean:0.98 ): 2140\r\n",
      "Shell ( min:0.03, max:0.95, sd:0.27, mean:0.29 ): 1601\r\n",
      "Cloud ( min:0.0, max:0.03, sd:0.01, mean:0.01 ): 3264\r\n",
      "Number of partitions: 3\r\n",
      "RGPs: 11307\r\n",
      "Spots: 257\r\n",
      "Modules: 199\r\n",
      "Families in Modules: 1276\r\n"
     ]
    }
   ],
   "source": [
    "!ppanggolin all --anno $PANGENOMES/Acinetobacter.baumannii.list -o $PANGENOMES/Acinetobacter.baumannii -c $CPU -f\n",
    "!ppanggolin all --anno $PANGENOMES/Enterobacter.bugandensis.list -o $PANGENOMES/Enterobacter.bugandensis -c $CPU -f\n",
    "!ppanggolin all --anno $PANGENOMES/Enterobacter.cloacae.list -o $PANGENOMES/Enterobacter.cloacae -c $CPU -f\n",
    "!ppanggolin all --anno $PANGENOMES/Enterobacter.hormaechei_A.list -o $PANGENOMES/Enterobacter.hormaechei_A -c $CPU -f\n",
    "!ppanggolin all --anno $PANGENOMES/Enterobacter.kobei.list -o $PANGENOMES/Enterobacter.kobei -c $CPU -f\n",
    "!ppanggolin all --anno $PANGENOMES/Enterobacter.roggenkampii.list -o $PANGENOMES/Enterobacter.roggenkampii -c $CPU -f\n",
    "!ppanggolin all --anno $PANGENOMES/Enterococcus_B.faecium.list -o $PANGENOMES/Enterococcus_B.faecium -c $CPU -f\n",
    "!ppanggolin all --anno $PANGENOMES/Klebsiella.pneumoniae.list -o $PANGENOMES/Klebsiella.pneumoniae -c $CPU -f\n",
    "!ppanggolin all --anno $PANGENOMES/Pseudomonas.aeruginosa.list -o $PANGENOMES/Pseudomonas.aeruginosa -c $CPU -f\n",
    "!ppanggolin all --anno $PANGENOMES/Staphylococcus.aureus.list -o $PANGENOMES/Staphylococcus.aureus -c $CPU -f"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Pangenome AMR annotation with CARD database and RGI\n",
    "### Get gene faimilies reference sequences"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mkdir: impossible de créer le répertoire « data/GF_fasta »: Le fichier existe\r\n",
      "2022-11-21 13:34:28 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin fasta -p data/pangenomes/Acinetobacter.baumannii/pangenome.h5 -o data/GF_fasta/Acinetobacter.baumannii --gene_families all -f\r\n",
      "2022-11-21 13:34:28 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 13:34:28 readBinaries.py:l51 INFO\tGetting the current pangenome status\r\n",
      "2022-11-21 13:34:28 readBinaries.py:l505 INFO\tReading pangenome annotations...\r\n",
      "100%|███████████████████████████| 1070798/1070798 [00:02<00:00, 363321.08gene/s]\r\n",
      "100%|███████████████████████████████████| 284/284 [00:15<00:00, 17.84organism/s]\r\n",
      "2022-11-21 13:34:47 readBinaries.py:l519 INFO\tReading pangenome gene families...\r\n",
      "100%|████████████████████| 1045048/1045048 [00:04<00:00, 231330.27gene family/s]\r\n",
      "100%|█████████████████████████| 14427/14427 [00:00<00:00, 67431.59gene family/s]\r\n",
      "2022-11-21 13:34:51 writeSequences.py:l95 INFO\tWriting all of the representative nucleotide sequences of the gene families...\r\n",
      "2022-11-21 13:34:51 readBinaries.py:l130 INFO\tExtracting and writing CDS sequences from a data/pangenomes/Acinetobacter.baumannii/pangenome.h5 file to a fasta file...\r\n",
      "100%|███████████████████████████| 1045048/1045048 [00:01<00:00, 880225.28gene/s]\r\n",
      "2022-11-21 13:34:59 writeSequences.py:l149 INFO\tDone writing the representative nucleotide sequences of the gene families : 'data/GF_fasta/Acinetobacter.baumannii/all_nucleotide_families.fasta'\r\n",
      "2022-11-21 13:35:02 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin fasta -p data/pangenomes/Enterobacter.bugandensis/pangenome.h5 -o data/GF_fasta/Enterobacter.bugandensis --gene_families all -f\r\n",
      "2022-11-21 13:35:02 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 13:35:02 readBinaries.py:l51 INFO\tGetting the current pangenome status\r\n",
      "2022-11-21 13:35:02 readBinaries.py:l505 INFO\tReading pangenome annotations...\r\n",
      "100%|█████████████████████████████| 500674/500674 [00:01<00:00, 375309.09gene/s]\r\n",
      "100%|███████████████████████████████████| 114/114 [00:07<00:00, 16.09organism/s]\r\n",
      "2022-11-21 13:35:11 readBinaries.py:l519 INFO\tReading pangenome gene families...\r\n",
      "100%|██████████████████████| 492746/492746 [00:02<00:00, 228228.54gene family/s]\r\n",
      "100%|█████████████████████████| 11542/11542 [00:00<00:00, 89119.60gene family/s]\r\n",
      "2022-11-21 13:35:13 writeSequences.py:l95 INFO\tWriting all of the representative nucleotide sequences of the gene families...\r\n",
      "2022-11-21 13:35:13 readBinaries.py:l130 INFO\tExtracting and writing CDS sequences from a data/pangenomes/Enterobacter.bugandensis/pangenome.h5 file to a fasta file...\r\n",
      "100%|█████████████████████████████| 492746/492746 [00:00<00:00, 856167.42gene/s]\r\n",
      "2022-11-21 13:35:18 writeSequences.py:l149 INFO\tDone writing the representative nucleotide sequences of the gene families : 'data/GF_fasta/Enterobacter.bugandensis/all_nucleotide_families.fasta'\r\n",
      "2022-11-21 13:35:21 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin fasta -p data/pangenomes/Enterobacter.cloacae/pangenome.h5 -o data/GF_fasta/Enterobacter.cloacae --gene_families all -f\r\n",
      "2022-11-21 13:35:21 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 13:35:21 readBinaries.py:l51 INFO\tGetting the current pangenome status\r\n",
      "2022-11-21 13:35:21 readBinaries.py:l505 INFO\tReading pangenome annotations...\r\n",
      "100%|█████████████████████████████| 591490/591490 [00:01<00:00, 388123.49gene/s]\r\n",
      "100%|███████████████████████████████████| 124/124 [00:08<00:00, 14.86organism/s]\r\n",
      "2022-11-21 13:35:31 readBinaries.py:l519 INFO\tReading pangenome gene families...\r\n",
      "100%|██████████████████████| 581629/581629 [00:02<00:00, 197104.94gene family/s]\r\n",
      "100%|█████████████████████████| 19918/19918 [00:00<00:00, 86131.43gene family/s]\r\n",
      "2022-11-21 13:35:34 writeSequences.py:l95 INFO\tWriting all of the representative nucleotide sequences of the gene families...\r\n",
      "2022-11-21 13:35:34 readBinaries.py:l130 INFO\tExtracting and writing CDS sequences from a data/pangenomes/Enterobacter.cloacae/pangenome.h5 file to a fasta file...\r\n",
      "100%|█████████████████████████████| 581629/581629 [00:00<00:00, 806349.99gene/s]\r\n",
      "2022-11-21 13:35:41 writeSequences.py:l149 INFO\tDone writing the representative nucleotide sequences of the gene families : 'data/GF_fasta/Enterobacter.cloacae/all_nucleotide_families.fasta'\r\n",
      "2022-11-21 13:35:44 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin fasta -p data/pangenomes/Enterobacter.hormaechei_A/pangenome.h5 -o data/GF_fasta/Enterobacter.hormaechei_A --gene_families all -f\r\n",
      "2022-11-21 13:35:44 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 13:35:44 readBinaries.py:l51 INFO\tGetting the current pangenome status\r\n",
      "2022-11-21 13:35:44 readBinaries.py:l505 INFO\tReading pangenome annotations...\r\n",
      "100%|█████████████████████████████| 757004/757004 [00:01<00:00, 402712.75gene/s]\r\n",
      "100%|███████████████████████████████████| 159/159 [00:10<00:00, 15.20organism/s]\r\n",
      "2022-11-21 13:35:56 readBinaries.py:l519 INFO\tReading pangenome gene families...\r\n",
      "100%|██████████████████████| 739586/739586 [00:03<00:00, 196317.91gene family/s]\r\n",
      "100%|█████████████████████████| 18180/18180 [00:00<00:00, 88777.74gene family/s]\r\n",
      "2022-11-21 13:36:00 writeSequences.py:l95 INFO\tWriting all of the representative nucleotide sequences of the gene families...\r\n",
      "2022-11-21 13:36:00 readBinaries.py:l130 INFO\tExtracting and writing CDS sequences from a data/pangenomes/Enterobacter.hormaechei_A/pangenome.h5 file to a fasta file...\r\n",
      "100%|█████████████████████████████| 739586/739586 [00:00<00:00, 858211.06gene/s]\r\n",
      "2022-11-21 13:36:07 writeSequences.py:l149 INFO\tDone writing the representative nucleotide sequences of the gene families : 'data/GF_fasta/Enterobacter.hormaechei_A/all_nucleotide_families.fasta'\r\n",
      "2022-11-21 13:36:10 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin fasta -p data/pangenomes/Enterobacter.kobei/pangenome.h5 -o data/GF_fasta/Enterobacter.kobei --gene_families all -f\r\n",
      "2022-11-21 13:36:10 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 13:36:10 readBinaries.py:l51 INFO\tGetting the current pangenome status\r\n",
      "2022-11-21 13:36:10 readBinaries.py:l505 INFO\tReading pangenome annotations...\r\n",
      "100%|█████████████████████████████| 634334/634334 [00:01<00:00, 388316.32gene/s]\r\n",
      "100%|███████████████████████████████████| 136/136 [00:09<00:00, 15.11organism/s]\r\n",
      "2022-11-21 13:36:21 readBinaries.py:l519 INFO\tReading pangenome gene families...\r\n",
      "100%|██████████████████████| 623869/623869 [00:02<00:00, 224640.76gene family/s]\r\n",
      "100%|█████████████████████████| 18170/18170 [00:00<00:00, 84448.73gene family/s]\r\n",
      "2022-11-21 13:36:24 writeSequences.py:l95 INFO\tWriting all of the representative nucleotide sequences of the gene families...\r\n",
      "2022-11-21 13:36:24 readBinaries.py:l130 INFO\tExtracting and writing CDS sequences from a data/pangenomes/Enterobacter.kobei/pangenome.h5 file to a fasta file...\r\n",
      "100%|█████████████████████████████| 623869/623869 [00:00<00:00, 663872.91gene/s]\r\n",
      "2022-11-21 13:36:30 writeSequences.py:l149 INFO\tDone writing the representative nucleotide sequences of the gene families : 'data/GF_fasta/Enterobacter.kobei/all_nucleotide_families.fasta'\r\n",
      "2022-11-21 13:36:33 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin fasta -p data/pangenomes/Enterobacter.roggenkampii/pangenome.h5 -o data/GF_fasta/Enterobacter.roggenkampii --gene_families all -f\r\n",
      "2022-11-21 13:36:33 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 13:36:33 readBinaries.py:l51 INFO\tGetting the current pangenome status\r\n",
      "2022-11-21 13:36:33 readBinaries.py:l505 INFO\tReading pangenome annotations...\r\n",
      "100%|█████████████████████████████| 912560/912560 [00:03<00:00, 297369.41gene/s]\r\n",
      "100%|███████████████████████████████████| 195/195 [00:16<00:00, 12.01organism/s]\r\n",
      "2022-11-21 13:36:52 readBinaries.py:l519 INFO\tReading pangenome gene families...\r\n",
      "100%|██████████████████████| 896840/896840 [00:05<00:00, 151697.97gene family/s]\r\n",
      "100%|█████████████████████████| 23062/23062 [00:00<00:00, 80568.59gene family/s]\r\n",
      "2022-11-21 13:36:59 writeSequences.py:l95 INFO\tWriting all of the representative nucleotide sequences of the gene families...\r\n",
      "2022-11-21 13:36:59 readBinaries.py:l130 INFO\tExtracting and writing CDS sequences from a data/pangenomes/Enterobacter.roggenkampii/pangenome.h5 file to a fasta file...\r\n",
      "100%|█████████████████████████████| 896840/896840 [00:01<00:00, 656131.96gene/s]\r\n",
      "2022-11-21 13:37:11 writeSequences.py:l149 INFO\tDone writing the representative nucleotide sequences of the gene families : 'data/GF_fasta/Enterobacter.roggenkampii/all_nucleotide_families.fasta'\r\n",
      "2022-11-21 13:37:15 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin fasta -p data/pangenomes/Enterococcus_B.faecium/pangenome.h5 -o data/GF_fasta/Enterococcus_B.faecium --gene_families all -f\r\n",
      "2022-11-21 13:37:15 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 13:37:15 readBinaries.py:l51 INFO\tGetting the current pangenome status\r\n",
      "2022-11-21 13:37:15 readBinaries.py:l505 INFO\tReading pangenome annotations...\r\n",
      "100%|█████████████████████████████| 588226/588226 [00:01<00:00, 324977.29gene/s]\r\n",
      "100%|███████████████████████████████████| 204/204 [00:10<00:00, 19.51organism/s]\r\n",
      "2022-11-21 13:37:27 readBinaries.py:l519 INFO\tReading pangenome gene families...\r\n",
      "100%|██████████████████████| 570393/570393 [00:03<00:00, 188429.42gene family/s]\r\n",
      "100%|██████████████████████████| 7889/7889 [00:00<00:00, 142772.11gene family/s]\r\n",
      "2022-11-21 13:37:30 writeSequences.py:l95 INFO\tWriting all of the representative nucleotide sequences of the gene families...\r\n",
      "2022-11-21 13:37:30 readBinaries.py:l130 INFO\tExtracting and writing CDS sequences from a data/pangenomes/Enterococcus_B.faecium/pangenome.h5 file to a fasta file...\r\n",
      "100%|█████████████████████████████| 570393/570393 [00:00<00:00, 738033.16gene/s]\r\n",
      "2022-11-21 13:37:32 writeSequences.py:l149 INFO\tDone writing the representative nucleotide sequences of the gene families : 'data/GF_fasta/Enterococcus_B.faecium/all_nucleotide_families.fasta'\r\n",
      "2022-11-21 13:37:35 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin fasta -p data/pangenomes/Klebsiella.pneumoniae/pangenome.h5 -o data/GF_fasta/Klebsiella.pneumoniae --gene_families all -f\r\n",
      "2022-11-21 13:37:35 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 13:37:35 readBinaries.py:l51 INFO\tGetting the current pangenome status\r\n",
      "2022-11-21 13:37:35 readBinaries.py:l505 INFO\tReading pangenome annotations...\r\n",
      "100%|███████████████████████████| 3166261/3166261 [00:09<00:00, 327290.44gene/s]\r\n",
      "100%|███████████████████████████████████| 593/593 [00:45<00:00, 13.02organism/s]\r\n",
      "2022-11-21 13:38:30 readBinaries.py:l519 INFO\tReading pangenome gene families...\r\n",
      "100%|████████████████████| 3100191/3100191 [00:15<00:00, 200644.47gene family/s]\r\n",
      "100%|█████████████████████████| 29144/29144 [00:00<00:00, 96691.54gene family/s]\r\n",
      "2022-11-21 13:38:46 writeSequences.py:l95 INFO\tWriting all of the representative nucleotide sequences of the gene families...\r\n",
      "2022-11-21 13:38:46 readBinaries.py:l130 INFO\tExtracting and writing CDS sequences from a data/pangenomes/Klebsiella.pneumoniae/pangenome.h5 file to a fasta file...\r\n",
      "100%|███████████████████████████| 3100191/3100191 [00:03<00:00, 865900.31gene/s]\r\n",
      "2022-11-21 13:39:01 writeSequences.py:l149 INFO\tDone writing the representative nucleotide sequences of the gene families : 'data/GF_fasta/Klebsiella.pneumoniae/all_nucleotide_families.fasta'\r\n",
      "2022-11-21 13:39:08 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin fasta -p data/pangenomes/Pseudomonas.aeruginosa/pangenome.h5 -o data/GF_fasta/Pseudomonas.aeruginosa --gene_families all -f\r\n",
      "2022-11-21 13:39:08 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 13:39:08 readBinaries.py:l51 INFO\tGetting the current pangenome status\r\n",
      "2022-11-21 13:39:08 readBinaries.py:l505 INFO\tReading pangenome annotations...\r\n",
      "100%|███████████████████████████| 1916555/1916555 [00:04<00:00, 390620.89gene/s]\r\n",
      "100%|███████████████████████████████████| 311/311 [00:27<00:00, 11.33organism/s]\r\n",
      "2022-11-21 13:39:40 readBinaries.py:l519 INFO\tReading pangenome gene families...\r\n",
      "100%|████████████████████| 1892863/1892863 [00:09<00:00, 193349.69gene family/s]\r\n",
      "100%|█████████████████████████| 23687/23687 [00:00<00:00, 79422.69gene family/s]\r\n",
      "2022-11-21 13:39:50 writeSequences.py:l95 INFO\tWriting all of the representative nucleotide sequences of the gene families...\r\n",
      "2022-11-21 13:39:50 readBinaries.py:l130 INFO\tExtracting and writing CDS sequences from a data/pangenomes/Pseudomonas.aeruginosa/pangenome.h5 file to a fasta file...\r\n",
      "100%|███████████████████████████| 1892863/1892863 [00:02<00:00, 869169.89gene/s]\r\n",
      "2022-11-21 13:40:06 writeSequences.py:l149 INFO\tDone writing the representative nucleotide sequences of the gene families : 'data/GF_fasta/Pseudomonas.aeruginosa/all_nucleotide_families.fasta'\r\n",
      "2022-11-21 13:40:11 utils.py:l116 INFO\tCommand: /home/jarnoux/anaconda3/envs/pangraph/bin/ppanggolin fasta -p data/pangenomes/Staphylococcus.aureus/pangenome.h5 -o data/GF_fasta/Staphylococcus.aureus --gene_families all -f\r\n",
      "2022-11-21 13:40:11 utils.py:l117 INFO\tPPanGGOLiN version: 1.2.105\r\n",
      "2022-11-21 13:40:11 readBinaries.py:l51 INFO\tGetting the current pangenome status\r\n",
      "2022-11-21 13:40:11 readBinaries.py:l505 INFO\tReading pangenome annotations...\r\n",
      "100%|███████████████████████████| 1631806/1631806 [00:04<00:00, 385585.33gene/s]\r\n",
      "100%|███████████████████████████████████| 595/595 [00:24<00:00, 24.46organism/s]\r\n",
      "2022-11-21 13:40:40 readBinaries.py:l519 INFO\tReading pangenome gene families...\r\n",
      "100%|████████████████████| 1586295/1586295 [00:07<00:00, 213309.26gene family/s]\r\n",
      "100%|███████████████████████████| 7005/7005 [00:00<00:00, 53132.49gene family/s]\r\n",
      "2022-11-21 13:40:47 writeSequences.py:l95 INFO\tWriting all of the representative nucleotide sequences of the gene families...\r\n",
      "2022-11-21 13:40:47 readBinaries.py:l130 INFO\tExtracting and writing CDS sequences from a data/pangenomes/Staphylococcus.aureus/pangenome.h5 file to a fasta file...\r\n",
      "100%|███████████████████████████| 1586295/1586295 [00:01<00:00, 805079.92gene/s]\r\n",
      "2022-11-21 13:40:57 writeSequences.py:l149 INFO\tDone writing the representative nucleotide sequences of the gene families : 'data/GF_fasta/Staphylococcus.aureus/all_nucleotide_families.fasta'\r\n"
     ]
    }
   ],
   "source": [
    "!mkdir $GF\n",
    "!ppanggolin fasta -p $PANGENOMES/Acinetobacter.baumannii/pangenome.h5 -o $GF/Acinetobacter.baumannii --gene_families all -f\n",
    "!ppanggolin fasta -p $PANGENOMES/Enterobacter.bugandensis/pangenome.h5 -o $GF/Enterobacter.bugandensis --gene_families all -f\n",
    "!ppanggolin fasta -p $PANGENOMES/Enterobacter.cloacae/pangenome.h5 -o $GF/Enterobacter.cloacae --gene_families all -f\n",
    "!ppanggolin fasta -p $PANGENOMES/Enterobacter.hormaechei_A/pangenome.h5 -o $GF/Enterobacter.hormaechei_A --gene_families all -f\n",
    "!ppanggolin fasta -p $PANGENOMES/Enterobacter.kobei/pangenome.h5 -o $GF/Enterobacter.kobei --gene_families all -f\n",
    "!ppanggolin fasta -p $PANGENOMES/Enterobacter.roggenkampii/pangenome.h5 -o $GF/Enterobacter.roggenkampii --gene_families all -f\n",
    "!ppanggolin fasta -p $PANGENOMES/Enterococcus_B.faecium/pangenome.h5 -o $GF/Enterococcus_B.faecium --gene_families all -f\n",
    "!ppanggolin fasta -p $PANGENOMES/Klebsiella.pneumoniae/pangenome.h5 -o $GF/Klebsiella.pneumoniae --gene_families all -f\n",
    "!ppanggolin fasta -p $PANGENOMES/Pseudomonas.aeruginosa/pangenome.h5 -o $GF/Pseudomonas.aeruginosa --gene_families all -f\n",
    "!ppanggolin fasta -p $PANGENOMES/Staphylococcus.aureus/pangenome.h5 -o $GF/Staphylococcus.aureus --gene_families all -f"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Annotate with RGI\n",
    "RGI is only compatible with python version 3.6. It's why it's not possible to run it in the current notebook.\n",
    "To annotate your pangenome you have to follow the installation instruction [here](https://github.com/arpcard/rgi#id52).\n",
    "\n",
    "To annotate a pangenome you can use the following commands :\n",
    "```\n",
    "rgi load --card_json /path/to/card.json --local\n",
    "rgi main -i $GF/speciesall_nucleotide_families.fasta -o $CARD/species/card_res -n $CPU --include_nudge\n",
    "```\n",
    "When it's done for all pangenome, you can run the next block of code."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████| 634334/634334 [00:01<00:00, 396499.19gene/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 136/136 [00:08<00:00, 15.77organism/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████| 623869/623869 [00:02<00:00, 214094.19gene family/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████| 18170/18170 [00:00<00:00, 87089.36gene family/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████| 615049/615049 [00:02<00:00, 253097.96contig adjacency/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████| 88304/88304 [00:00<00:00, 110802.42region/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 2427/2427 [00:00<00:00, 16834.23spot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 4260/4260 [00:00<00:00, 375301.62module/s]\n",
      "0gene family [00:00, ?gene family/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Erasing the formerly computed gene family annotations...\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "99annot [00:00, 17419.81annot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 3166261/3166261 [00:07<00:00, 405114.75gene/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 593/593 [00:43<00:00, 13.52organism/s]\n",
      "100%|██████████████████████████████████████████████████████████████████████████████████████| 3100191/3100191 [00:12<00:00, 245271.57gene family/s]\n",
      "100%|██████████████████████████████████████████████████████████████████████████████████████████| 29144/29144 [00:00<00:00, 111575.42gene family/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████| 3094977/3094977 [00:15<00:00, 205050.76contig adjacency/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 564463/564463 [00:01<00:00, 507331.72region/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████| 23275/23275 [00:07<00:00, 3066.33spot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 8197/8197 [00:00<00:00, 430889.96module/s]\n",
      "There is no annotation in Klebsiella.pneumoniae\n",
      "140annot [00:00, 19160.20annot/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████| 500674/500674 [00:01<00:00, 431183.80gene/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 114/114 [00:05<00:00, 20.72organism/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████| 492746/492746 [00:01<00:00, 313075.58gene family/s]\n",
      "100%|██████████████████████████████████████████████████████████████████████████████████████████| 11542/11542 [00:00<00:00, 103727.05gene family/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████| 488568/488568 [00:09<00:00, 50748.75contig adjacency/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████| 53520/53520 [00:00<00:00, 346870.23region/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 3048/3048 [00:00<00:00, 22895.72spot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 2635/2635 [00:00<00:00, 349083.73module/s]\n",
      "There is no annotation in Enterobacter.bugandensis\n",
      "49annot [00:00, 21665.71annot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 1070798/1070798 [00:02<00:00, 421815.19gene/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 284/284 [00:16<00:00, 17.42organism/s]\n",
      "100%|██████████████████████████████████████████████████████████████████████████████████████| 1045048/1045048 [00:04<00:00, 209598.02gene family/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████| 14427/14427 [00:00<00:00, 68608.39gene family/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████| 1043916/1043916 [00:05<00:00, 196227.69contig adjacency/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 257053/257053 [00:00<00:00, 468876.45region/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 8933/8933 [00:02<00:00, 3646.58spot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 3789/3789 [00:00<00:00, 213032.41module/s]\n",
      "There is no annotation in Acinetobacter.baumannii\n",
      "97annot [00:00, 31782.48annot/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████| 591490/591490 [00:01<00:00, 375142.48gene/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 124/124 [00:10<00:00, 12.36organism/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████| 581629/581629 [00:02<00:00, 232095.52gene family/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████| 19918/19918 [00:00<00:00, 85152.99gene family/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████| 572409/572409 [00:02<00:00, 193229.51contig adjacency/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 112136/112136 [00:00<00:00, 428936.65region/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 4499/4499 [00:00<00:00, 16791.91spot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 5017/5017 [00:00<00:00, 358798.65module/s]\n",
      "There is no annotation in Enterobacter.cloacae\n",
      "117annot [00:00, 32443.05annot/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████| 588226/588226 [00:01<00:00, 385994.41gene/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 204/204 [00:09<00:00, 21.43organism/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████| 570393/570393 [00:02<00:00, 204418.88gene family/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████████| 7889/7889 [00:00<00:00, 175098.37gene family/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████| 568947/568947 [00:02<00:00, 229923.65contig adjacency/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 165391/165391 [00:00<00:00, 456535.75region/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 5136/5136 [00:00<00:00, 6724.82spot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 2391/2391 [00:00<00:00, 374886.20module/s]\n",
      "There is no annotation in Enterococcus_B.faecium\n",
      "62annot [00:00, 35755.10annot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 1916555/1916555 [00:04<00:00, 394550.86gene/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 311/311 [00:29<00:00, 10.49organism/s]\n",
      "100%|██████████████████████████████████████████████████████████████████████████████████████| 1892863/1892863 [00:08<00:00, 231763.74gene family/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████| 23687/23687 [00:00<00:00, 78935.73gene family/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████| 1891546/1891546 [00:09<00:00, 192099.07contig adjacency/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 234118/234118 [00:00<00:00, 453065.98region/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████| 10746/10746 [00:02<00:00, 4883.97spot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 6358/6358 [00:00<00:00, 268756.71module/s]\n",
      "There is no annotation in Pseudomonas.aeruginosa\n",
      "153annot [00:00, 26549.52annot/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████| 912560/912560 [00:02<00:00, 370910.78gene/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 195/195 [00:15<00:00, 12.32organism/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████| 896840/896840 [00:03<00:00, 242688.08gene family/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████| 23062/23062 [00:00<00:00, 97315.36gene family/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████| 882941/882941 [00:04<00:00, 197778.54contig adjacency/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 170998/170998 [00:00<00:00, 445991.46region/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 5829/5829 [00:00<00:00, 11964.01spot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 6395/6395 [00:00<00:00, 265081.87module/s]\n",
      "There is no annotation in Enterobacter.roggenkampii\n",
      "108annot [00:00, 20745.81annot/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████| 757004/757004 [00:01<00:00, 405297.89gene/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 159/159 [00:12<00:00, 12.46organism/s]\n",
      "100%|████████████████████████████████████████████████████████████████████████████████████████| 739586/739586 [00:03<00:00, 238026.96gene family/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████| 18180/18180 [00:00<00:00, 91222.63gene family/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████| 739063/739063 [00:03<00:00, 192506.17contig adjacency/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 140090/140090 [00:00<00:00, 464741.81region/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 5399/5399 [00:00<00:00, 9971.70spot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 5321/5321 [00:00<00:00, 368927.36module/s]\n",
      "There is no annotation in Enterobacter.hormaechei_A\n",
      "118annot [00:00, 26370.84annot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 1631806/1631806 [00:04<00:00, 381852.58gene/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 595/595 [00:25<00:00, 23.03organism/s]\n",
      "100%|██████████████████████████████████████████████████████████████████████████████████████| 1586295/1586295 [00:07<00:00, 218451.71gene family/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 7005/7005 [00:00<00:00, 53137.39gene family/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████| 1583889/1583889 [00:08<00:00, 183570.82contig adjacency/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████| 217265/217265 [00:00<00:00, 429508.37region/s]\n",
      "100%|███████████████████████████████████████████████████████████████████████████████████████████████████| 11012/11012 [00:04<00:00, 2385.05spot/s]\n",
      "100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 1276/1276 [00:00<00:00, 171503.30module/s]\n",
      "There is no annotation in Staphylococcus.aureus\n",
      "71annot [00:00, 47900.21annot/s]\n"
     ]
    }
   ],
   "source": [
    "import tables\n",
    "from script.python import Pangenome\n",
    "from script.python.utils import check_pangenome_info\n",
    "from script.python.annot import card_parse, annotation_to_families, write_gene_fam_annot\n",
    "\n",
    "pangenome_dir = [pan for pan in PANGENOMES.iterdir() if pan.is_dir()]\n",
    "for card, pan in zip(CARD.iterdir(), pangenome_dir):\n",
    "    current_pangenome = Pangenome(name=pan.name)\n",
    "    current_pangenome.add_file(f'{pan}/pangenome.h5')\n",
    "    check_pangenome_info(current_pangenome, need_annotations=True, need_families=True, need_graph=True,\n",
    "                         need_rgp=True, need_spots=True, need_modules=True, need_anntation_fam=True,\n",
    "                         disable_bar=False)\n",
    "    card_data = card_parse(f'{card}/card_res.txt')\n",
    "    annotation_to_families(card_data, pangenome=current_pangenome, source='CARD')\n",
    "    h5f = tables.open_file(current_pangenome.file, \"a\")\n",
    "    write_gene_fam_annot(current_pangenome, h5f, force=True, disable_bar=False)\n",
    "    h5f.close()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Compute similarity\n",
    "## Align sequence"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "convertalis data/similarity/sequenceDB data/similarity/sequenceDB data/similarity/alignDB data/similarity/align.tsv --format-output query,target,pident,qcov,tcov --threads 6 \r\n",
      "\r\n",
      "MMseqs Version:        \t14.7e284\r\n",
      "Substitution matrix    \taa:blosum62.out,nucl:nucleotide.out\r\n",
      "Alignment format       \t0\r\n",
      "Format alignment output\tquery,target,pident,qcov,tcov\r\n",
      "Translation table      \t1\r\n",
      "Gap open cost          \taa:11,nucl:5\r\n",
      "Gap extension cost     \taa:1,nucl:2\r\n",
      "Database output        \tfalse\r\n",
      "Preload mode           \t0\r\n",
      "Search type            \t0\r\n",
      "Threads                \t6\r\n",
      "Compressed             \t0\r\n",
      "Verbosity              \t3\r\n",
      "\r\n",
      "[=================================================================] 100.00% 25.97K 0s 44ms     \r\n",
      "Time for merging to align.tsv: 0h 0m 0s 11ms\r\n",
      "Time for processing: 0h 0m 0s 68ms\r\n"
     ]
    }
   ],
   "source": [
    "!cat $GF/*/all_nucleotide_families.fasta > $SIMILARITY/all_nucleotide_families.fasta\n",
    "!mmseqs createdb $SIMILARITY/all_nucleotide_families.fasta $SIMILARITY/sequenceDB\n",
    "!mmseqs search $SIMILARITY/sequenceDB $SIMILARITY/sequenceDB $SIMILARITY/alignDB /tmp/ --search-type 3 --threads $CPU\n",
    "!mmseqs convertalis $SIMILARITY/sequenceDB $SIMILARITY/sequenceDB $SIMILARITY/alignDB $SIMILARITY/align.tsv --format-output query,target,pident,qcov,tcov --threads $CPU"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Filter and parse alignment results"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "outputs": [],
   "source": [
    "def parse(file: Path, identity: float, coverage: float):\n",
    "    df = pd.read_csv(filepath_or_buffer=file, sep=\"\\t\", header=None,\n",
    "                     names=[\"GF_1\", \"GF_2\", \"identity\", \"qcov\", \"tcov\"])\n",
    "    df = df[df['GF_1'] != df['GF_2']]\n",
    "    df = df[df['identity'] > identity]\n",
    "    df = df[(df['qcov'] > coverage) & (df['tcov'] > coverage)]\n",
    "    filter_df = pd.concat([df[['GF_1', 'GF_2', 'identity']],\n",
    "                           df[['qcov', 'tcov']].min(axis=1).apply(lambda x: x*100)], axis=1)\n",
    "    filter_df.columns = ['GF_1', 'GF_2', 'identity', 'coverage']\n",
    "    return filter_df\n",
    "\n",
    "align_file = Path(f'{SIMILARITY}/align.tsv')\n",
    "filter_df = parse(align_file, identity=30, coverage=0.8)\n",
    "filter_df.to_csv(f'{SIMILARITY}/similarity.tsv', sep=\"\\t\", header=None, index=False)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Clean to relaunch"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "!rm -rf PPanGGOLiN\n",
    "!conda env remove -n pangraph"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "id": "0bb48de1",
   "metadata": {},
   "source": [
    "# References\n",
    "1. S. Sakr _qnd al_. “The future is big graphs: a community view on graph processing systems,” Commun. ACM, vol. 64, no.9, pp. 62–71, 2021.\n",
    "2. G. Gautreau _and al_. “PPanGGOLiN: Depicting microbial diversity via a partitioned pangenome graph,” vol. 16, no. 3, p. e1007732, publisher: Public Library of Science. [https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007732](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1007732)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
